---
title: "SRTpipeline - Embedding and spatial clustering for single sample"
output:
  html_document:
  theme: united
df_print: kable
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---
***

```{r setup, include=FALSE}
all_times <- list()  # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
      now <<- Sys.time()
    } else {
      res <- difftime(Sys.time(), now, units = "secs")
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  time_it = TRUE
)
```

# Overview
This tutorial demonstrates how to use `SRTpipeline` (>=0.1.0) to analyze multiple spatially-resolved transcriptomics (SRT) data.  We emphases how to use `DR-SC` model to achieve  joint embedding (dimension reduction) and spaital clustering based on the `SRTProject` object in the `SRTpipeline` package. This tutorial will cover the following tasks, which we believe will be common for many spatial analyses:

* Normalization
* Feature selection
* Joint embedding and spatial clustering
* DEG analysis
* Spatial trajectory inference
* Visualization

First, we load `SRTpipeline`.
```{r init}
library(SRTpipeline)
```

# Dataset
Here, we will be analyzing  a spatial transcriptomics dataset (SampleID: 151672) for human dorsolateral prefrontal cortex (DLPFC). 
There are 4015  spots and 33538 genes that were sequenced on the 10x Visium platform. Our preprocessed data can be downloaded [here](https://github.com/feiyoung/PRECAST/blob/main/vignettes_data/dlpfc_151672.rda?raw=true), and the raw data can be found [here](https://github.com/LieberInstitute/spatialLIBD).

Next, we  download the data to the current working path for the followed analysis by the following command:
```{r eval=FALSE}
githubURL <- "https://github.com/feiyoung/PRECAST/blob/main/vignettes_data/dlpfc_151672.rda?raw=true"
download.file(githubURL,"dlpfc_151672.rda",mode='wb')
```
Then load to R. The data is saved in a Seurat object named `dlpfc_151672`.
```{r  eval = FALSE}
load("dlpfc_151672.rda")
```




# Pre-processing workflow

## Create a SRTProject object
First, we show how to create a SRTProject object step by step.

Because the gene name is ensembl in this data, we change the genes' ensembl name to symbols for convenience of followed analyses. We can see most of ensembl names have been transferred to the symbol names and the genes not transferred are generally not in the variable gene set. Thhus, it does not matter for analyses that there are genes not be transferred.
```{r  eval = FALSE}
count_matrix <- dlpfc_151672[['RNA']]@counts
## Use eg.db database: this method is fast
symbol_name <- transferGeneNames(row.names(count_matrix), now_name = "ensembl", to_name="symbol", species="Human", Method='eg.db')
row.names(count_matrix) <- symbol_name
symbol_name[1:20]
```

```{r  eval = FALSE}
## create count matrix list: note each component has a name, i.e., `ID151672`.

cntList <- list(ID151672=count_matrix)

## create spatial coordinate matrix 
coordList <- list(cbind(dlpfc_151672$row, dlpfc_151672$col))

## create metadata list
meta.data <- dlpfc_151672@meta.data
metadataList <- list(meta.data)

## create meta data for each data batches. Here we only have one data batch.
sampleMetadata <- data.frame(species="Human", tissues=c('DLPFC'))
row.names(sampleMetadata) <- names(cntList)

## Name of this project
projectName <- "Brain151672"

rm(dlpfc_151672)
```


Next, we start creating a SRTProject object. We can print the basic information of this object, including  three parts. The first part have the class of this object, outputPath of data that require to be output, h5filePath that saves the memory-cusuming data (i.e., count, logcount, ...). The second part is about the datasets basic information, such as how many data batches(sample) and the data names, sample meta data (sampleColData) and meta data for each spot (cellMetaData). The last part is about downstream analyses information (that is empty) when this object created.
Quality control is also performed when creating the SRTProject.
SRTpipeline allows you to easily explore QC metrics and filter cells based on any user-defined criteria. A few QC metrics [commonly used](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4758103/) by the community include
* The number of unique genes detected in each cell. 
    + Low-quality cells or empty droplets will often have very few genes
    + Cell doublets or multiplets may exhibit an aberrantly high gene count
* Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)

In the example below, we visualize QC metrics, and use these to filter cells.
* filter out the spots with nonzero expression levels for less than 20 genes.
* filter out genes with non-zero expression levels for fewer than 20 spots 
    
```{r  eval = FALSE}
SRTProj <- CreateSRTProject(cntList, coordList, projectName =projectName, metadataList,
    sampleMetadata,  min.spots = 20, min.genes = 20)
SRTProj
```   


## Normalizing the data
After creating the SRTProject object, the next step is to normalize the data.  Normalized values are stored in the h5 file of h5file path.
```{r normalize, eval = FALSE}
SRTProj <- normalizeSRT(SRTProj)
```

## Feature selection
We next select a subset of genes that exhibit high spot-to-spot variation in the dataset (i.e, they are highly expressed in some spots, and lowly expressed in others). It has been  found that focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets [here](https://www.nature.com/articles/nmeth.2645).


### Choose spatially variable features
In the single_basic_tutorial, we introduced using `vst` to choose the highly vairiable genes. Here,  we introduce to apply the SPARK-X to choose the top spatially variable genes (SVGs).  By default, we return 2,000 genes per dataset.  These will be used in downstream analysis, like probabilistic embeddings and spatial clustering. In the SVGs selection implementation of `selectVariableFeatures()` function, we first  select 5,000 highly variable genes  to  speed up the computation, then use `SPARK-X` to choose top SVGs from these 5000 genes. The selected SVGs are saved in `SRTProj@geneMetaData$isVGs#combine` for followed analysse. In addition, the other columns of `SRTProj@geneMetaData` save the selected SVGs for each data batch. Because there is only one data batch, `SRTProj@geneMetaData$isSVGs#ID151672` is equal to `SRTProj@geneMetaData$isVGs#combine`. The column `rankSVGs#ID151672` retains the rank of SVGs for the selected top SVGs and ranking first means that it is the most spaitally variable gene.

```{r, eval = FALSE}
SRTProj <- selectVariableFeatures(SRTProj, type = 'SVGs', method='SPARK-X')
head(SRTProj@geneMetaData)
```

Users can access the top SVGs using 
```{r, eval = FALSE}
(svg4 <- topVGs(SRTProj, ntop = 4, type='SVGs'))

```
Visualize the expression of these SVGs on the spatial coordinates.
```{r, eval = FALSE}
EachExprSpaHeatMap(SRTProj, features=svg4, layout.dim = c(2,2))
```


### Choose highly variable features
We can also  choose the top highly variable genes (HVGs) using `selectVariableFeatures()` function. We found the performance of using SVGs or HVGs has no significant difference for the downstream analyses, such as dimension reduction and spatial clustering.
```{r, eval=FALSE}
# not run here
# SRTProj <- selectVariableFeatures(SRTProj, type = 'HVGs', method='vst')
```

##  Calculate the adjcence matrix
Next, we define the neighbors for each spot by the spatial coordinates, which is represented by a sparse adjacency matrix (AdjMat) saved in the h5file. We can check the saved sparse adjacency matrix using `h5ls()`.
```{r  eval = FALSE}
## Obtain adjacence matrix
SRTProj <- AddAdj(SRTProj, platform = "Visium")
h5ls(SRTProj@projectMetadata$h5filePath)
```

# Joint embedding and spatial clustering
We fit the `DR-SC` model by using the top 2000 SVGs. Users can also use the top HVGs but we found the performance of using SVGs or HVGs has no significant difference for the downstream analyses, such as dimension reduction and spatial clustering. Here, we specify the number of clusters as 5, while `DR.SC` can also select the number of clusters given a integer vector for argument $K$. 

After finishing joint dimension reduction and spatial clustering, the SRTProject `SRTProj` has two more downstream analyses information: the low-dimensional embeddings DR.SC and inferred cluster labels, and they are saved in `SRTProj@reductions$DR.SC` and `SRTProj@clusters`, respectively.

```{r, eval = FALSE}
### Given K
SRTProj <- Cluster_DRSC(SRTProj, K=5, q=15)
SRTProj

```


# Visualization
`SRTpipeline` package oﬀers a variety of visualization tools. 
To run tSNE in SRTpipeline we use the `AddTSNE()` function:
```{r, eval = FALSE}
SRTProj <- AddTSNE(SRTProj, n_comp = 2, reduction='DR.SC')
```
The `reduction` tells `AddTSNE()` function uses specified reduction  in the `SRTProj@reductions` slot to calculate tSNE embeddings.

To plot the two-dimensional tSNE results, we use the `EmbedPlot()` function and pass the name of the tSNE embedding we just generated (“tSNE”). We can tell SRTpipeline how to color the spots by using `item` argument which tells SRTpipeline which data to use to find the specified metadata column provided to name. The data are from two sources:
`SRTProj@clusters` or `SRTProj@cellMetaData`. In particular, `item='cluster'` means using `SRTProj@clusters` for coloring.

We visualized the inferred domain types on  embeddings  using two
components from tSNE. The tSNE plot showed the domain clusters were well segregated.
```{r, eval = FALSE}
cols_cluster <- chooseColors(n_colors = 5)
p_tsne2 <- EmbedPlot(SRTProj, item='cluster', plotEmbeddings = 'tSNE', cols=cols_cluster, legend.position='bottom')
p_tsne2
```

To run UMAP in SRTpipeline we use the `AddUMAP()` function. Frist, we evaluate the two-dimensional UMAPs.

```{r, eval = FALSE}
SRTProj <- AddUMAP(SRTProj, n_comp = 2, reduction='DR.SC')
p_umap2 <- EmbedPlot(SRTProj, item='cluster', plotEmbeddings = 'UMAP', cols=cols_cluster, legend.position='bottom')
p_umap2
```

Next, we merge all plots into one figure.
```{r, eval = FALSE}
drawFigs(list(p_tsne2, p_umap2), layout.dim = c(1,2), common.legend = TRUE, legend.position='bottom')
```

Calculate the 3-dimensional tSNEs based on the extracted features from DR-SC. 
```{r, eval = FALSE}
SRTProj <- AddTSNE(SRTProj, n_comp = 3, reduction = 'DR.SC')
```
We visualize the 3-dimensional tSNEs on the spatial coordinates using RGB colors. We can see there are apparent spatial pattern for the embeddings. Users can also plot the 3-dimensional  UMAP RGB plot similarly.
```{r, eval = FALSE}
EachRGBSpaHeatMap(SRTProj, plot_type ='tSNE', title_name = "tSNE RGB plot: ")
```
Then, we visualize the spatial distribution of cluster labels that shows the layer structure.
```{r, eval = FALSE}
EachClusterSpaHeatMap(SRTProj, cols=cols_cluster, legend.position='right',base_size=16)
# remove the border
# EachClusterSpaHeatMap(SRTProj, cols=cols_cluster, legend.position='right',base_size=16, border_col='white')
```

We plotted the heatmap of Pearson’s correlation coefcients of the DR.SC embeddings among
the detected domains shows the good separation of the estimated embeddings across
domains (high values in the diagnal) and the correlations between deeper layers were high, e.g., there were high
correlations between domain 2 and 3, while correlations among the separated layers were low, i.e., domain 1 and 4.

```{r, eval = FALSE}
CCHeatMap(SRTProj, reduction = 'DR.SC', grp_color=cols_cluster)
```


# DEG analysis
To do downstream analyses, we require to get the count data from h5file. The function `getGeneSpotData()` access the gene-by-spot count matrix and other data in `SRTProj` and then return a `SpatialExperiment` object,  including two assays: counts, logcounts; rowData: from `SRTProj@geneMetaData`; colData: from `SRTProj@cellMetaData`, `SRTProj@clusters` and sample_id; reducedDim: from `SRTProj@reductions`, `SRTProj@plotEmbeddings` and sptial coordinates `Coord`.
```{r, eval = FALSE}
spe <- getGeneSpotData(SRTProj)
spe
```

After fitting DR-SC model, we can perform differentially expression analysis using  the spatial cluster labels. First, we detect the DE genes for cluster 1, where the argument `use_cluster` tells `FindDEGs()` uses which colname name in the `colData(spe)` to do DEG analysis.
```{r, eval = FALSE}
dat_deg <- FindDEGs(spe, cluster1='1', use_cluster = 'clusters')
subset(dat_deg, avg_log2FC >0.25)
```
We perform differential expression analysis for all clusters by using `FindAllDEGs()` function, then the DE genes' information is saved in a `DataFrame` object `dat_degs`. 
```{r, eval = FALSE}
dat_degs <- FindAllDEGs(spe, use_cluster = 'clusters')
dat_degs
```

We identify the significant DE genes by two criteria: (a) adjustd p-value less than 0.01 and (b) average log fold change greater than 0.4.

```{r, eval = FALSE}
degs_sig <- subset(dat_degs, p_val_adj < 0.01 & avg_log2FC>0.4)
degs_sig
```

In the following, we perform gene set enrichment analysis for the DE genes of each  Domain identified by DR-SC model using R package `gprofiler2`. 
```{r, eval = FALSE}
library(gprofiler2)
termList <- list()
for(k in 1: 5){
  # k <- 1
  cat("k = ", k, '\n')
  dat_degs_sub <- subset(degs_sig,  cluster==k)
  if(nrow(dat_degs_sub)==0) next;
  que1 <- dat_degs_sub$gene
  gostres <- gost(query = que1, organism = "hsapiens", correction_method='fdr')
  termList[[k]] <- gostres
}
head(termList[[1]]$result)
```

 To understand the functions of the identified spatial domains by DR-SC model, we compare the top significant biological process (BP)  pathways in GO database for the DE genes from Domain 1 and 3. Here, we only show to visualize the significant BP pathways and users can explore the other databases such as KEGG and HPA.
 
The enrichment analysis shows that genes specific to Domain 1 were highly enriched for galactosylceramide metabolic process, galactosylceramide biosynthetic process and  galactolipid metabolic process, which may indicate the cells in Domain 1 play an important role to achieve these functions. The genes specific to Domain 4 were highly enriched for  negative regulation of response to external stimulus and negative regulation of response to extracellular stimulus. 
```{r, eval = FALSE}
## Most commonly used databases
source_set <- c("GO:BP","GO:CC", "GO:MF",   'KEGG', "HPA")
cols <- c("steelblue3", "goldenrod", "brown3", "#f98866", "#CE6DBD")
## Here, we show GO:BP
source1 <- "GO:BP"
ss <- which(source_set==source1)
ntop = 5
names(cols) <- source_set
pList_enrich <- list()
for(ii in 1: 5){
  ## ii <- 4
  message("ii=", ii)
  dat <- termList[[ii]]$result
  if(!is.null(dat)){
    dat1 <- subset(dat, term_size< 500)
    dat1 <- get_top_pathway(dat1, ntop=ntop, source_set = source1)
    dat1 <- dat1[complete.cases(dat1),]
    dat1$nlog10P <- -log10(dat1$p_value)
   
    pList_enrich[[ii]] <-  barPlot_enrich(dat1[order(dat1$nlog10P),], source='source', 'term_name',
                                          'nlog10P', cols=cols[source_set[ss]],base_size=14) + ylab("-log10(p-adj)")+
      xlab("Biological terms") + ggtitle(paste0("Domain", ii))
  }
  
}

drawFigs(pList_enrich[c(1,4)], common.legend = T, layout.dim = c(2,1), align = 'hv')

```

We take out the top DE genes for each cluster for visualization. 
```{r, eval = FALSE}

library(dplyr)
n <- 1
dat_degs %>% as.data.frame %>%
  group_by(cluster) %>%
  top_n(n = n, wt = avg_log2FC) -> topGene
topGene
```

Then we plot the volin plot of logcount expressions grouped by clusters, which suggests the marker gene has higher expression in the corresponding cluster.
```{r, eval = FALSE}
library(scater)
plotExpression(spe, features = topGene$gene, x='clusters', exprs_values = 'logcounts',
               colour_by = 'clusters')
```


# Trajectory inference
Next, we performed trajectory inference using the embeddings and domain labels jointly estimated by DR.SC model. The `EmbedPlot()` function can be used to visualize the inferred pseudotime on a specified embedding. If there is only one data batch, the function `EachEmbedPlot()` plays the same role as function `EmbedPlot()`.
```{r, eval = FALSE}
spe <- AddTrajectory(spe, reduction = 'DR.SC', name='PT')
p1 <- EmbedPlot(spe, reduction = 'DR.SC' ,colour_by='PT')
p2 <- EmbedPlot(spe, reduction = 'Coord' ,colour_by='PT')
drawFigs(list(p1, p2), layout.dim = c(1,2), common.legend = TRUE, 
         legend.position = 'right')

# save(spe, SRTProj, file=paste0(SRTProj@projectMetadata$outputPath,"/SRTProj.rds"))
# load("F:/Research paper/IntegrateDRcluster/AnalysisCode/SRTpipeline/vignettes/DLPFC151672/SRTProj.rds")
```


<details>
<summary>**Session Info**</summary>
```{r}
sessionInfo()
```
</details>
