---
title: "Fast and accurate multiple data integration"
output:
  html_document:
  theme: united
df_print: kable
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---
***

```{r setup, include=FALSE}
all_times <- list()  # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
      now <<- Sys.time()
    } else {
      res <- difftime(Sys.time(), now, units = "secs")
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  time_it = TRUE
)
```

# Overview
This tutorial demonstrates how to use `SRTpipeline` (>=0.1.0) to analyze multiple spatially-resolved transcriptomics (SRT) data.  We emphases how to use iSC.MEB model for spatial clustering and integration and its followed applications based on `SRTProject` object in the `SRTpipeline` package. This tutorial will cover the following tasks, which we believe will be common for many spatial analyses:

* Normalization
* Feature selection
* Spatial clustering
* Data integration
* Batch-corrected gene expression
* DEG analysis
* Spatial trajectory inference
* Visualization

First, we load `SRTpipeline`.
```{r init}
library(SRTpipeline)
set.seed(2023)
```

# Dataset
For this tutorial, we will introduce how to create a `SRTProject` object with multiple SRT samples using SRTpipeline that includes an introduction to common analytical workflows for multiple data batches.
Here, we will be taking spatial transcriptomics dataset for human dorsolateral prefrontal cortex (DLPFC) as an example.
There are three tissue slices from the various donors with 3500~4500  spots and 33538 genes that were sequenced on the 10x Visium platform. Our preprocessed data can be downloaded [here](https://github.com/feiyoung/iSC.MEB.Analysis/blob/main/data/DLPFC_data/), and the raw data can be found [here](https://github.com/LieberInstitute/spatialLIBD).



Next, we  download the data to the current working path for the followed analysis by the following command:
```{r eval=FALSE}
name_ID12 <- as.character(c(151507,  151669, 151673))

for(r in seq_along(name_ID12)){
  githubURL <- paste0("https://github.com/feiyoung/iSC.MEB.Analysis/blob/main/data/DLPFC_data/", name_ID12[r], ".rds?raw=true")
  download.file(githubURL, paste0("dlpfc_", name_ID12[r], ".rds"),mode='wb')
}                            

```


# Pre-processing workflow

Then load to R and prepare for creating the `SRTProject` object.
```{r  eval = FALSE}
dir_file <- "D:/LearnFiles/Research paper/ProPCA/iSC.MEB.Analysis/data/DLPFC_data/"
name_ID12 <- as.character(c(151507,  151669, 151673))
# dir_file <- './dlpfc_'
n_sample <- length(name_ID12)
library(Seurat)
## create count matrix list: note each component has a name, i.e., `ID151672`.
cntList <- list()
## create spatial coordinate matrix 
coordList <- list()
## create metadata list
metadataList <- list()

for(r in 1: n_sample){
  # r <- 1
  message('r = ', r)
  id <- name_ID12[r]
  dlpfc <- readRDS(paste0(dir_file,id, ".rds"))
  sp_count <- dlpfc@assays@data$counts
  meta.data <- data.frame(annotated_label=dlpfc$layer_guess_reordered,
                          row=dlpfc$row, col=dlpfc$col)
  row.names(meta.data) <- colnames(sp_count)
  cntList[[r]] <- sp_count
  metadataList[[r]] <- meta.data
  coordList[[r]] <- cbind(row=dlpfc$row, col=dlpfc$col)
}
names(cntList) <- name_ID12
## create meta data for each data batches. Here we only have one data batch.
sampleMetadata <- data.frame(species=rep("Human", n_sample),
                             tissues=rep('DLPFC', n_sample),
                             donor = rep(1:3, each=1))
row.names(sampleMetadata) <- names(cntList)
## Name of this project
projectName <- "DLPFC3"
```



We check the data by printing it. We can see there are 3500~4500 spots and 33538 genes for each data batch.
```{r  eval = FALSE}
sapply(cntList, dim)
```
Check whether each data batch has the same genes: Yes.
```{r  eval = FALSE}
geneNamesList <- lapply(cntList, row.names)
shared_genes <- Reduce(intersect, geneNamesList)
length(shared_genes)
```



## Create a SRTProject object
First, we show how to create a SRTProject object step by step.

Because the gene name is ensembl in this data, we change the genes' ensembl name to symbols for convenience of followed analyses.
```{r  eval = FALSE}
cntList <- lapply(cntList, function(x) x[shared_genes,])
## Use eg.db database: this method is fast
symbol_name <- transferGeneNames(shared_genes, now_name = "ensembl", to_name="symbol", species="Human", Method='eg.db')

for(r in 1:n_sample){
  row.names(cntList[[r]]) <- symbol_name
}
symbol_name[1:10]
```


Next, we start creating SRTProject object. We can print the basic information of this object, including  three parts. The first part have the class of this object, outputPath of data that require to output, h5filePath that save the memory-cusuming data (i.e., count, logcount, ...). The second part is about the datasets basic information, such as how many data batches(sample) and the data names, sample meta data (sampleColData) and meta data for each spot (cellMetaData). The last part is about downstream analyses information that is empty when this object created.
```{r  eval = FALSE}
SRTProj <- CreateSRTProject(cntList, coordList, projectName =projectName, metadataList,
    sampleMetadata, min.spots = 20, min.genes = 20,force=F)
SRTProj
```

<details>
  <summary>**What does data in a h5 file look like?**</summary>

```{r  eval = FALSE}
h5ls(SRTProj@projectMetadata$h5filePath)
```

</details>


## Normalizing the data
After removing unwanted cells and genes from the dataset, the next step is to normalize the data.  To save RAM memory, normalized values are stored in disk as a `h5file`.
```{r  eval = FALSE}
SRTProj <- normalizeSRT(SRTProj,normalization.method='LogNormalize')
```

# Feature selection
We next select a subset of genes that exhibit high spot-to-spot variation in the dataset (i.e, they are highly expressed in some spots, and lowly expressed in others). It has been  found that focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets [here](https://www.nature.com/articles/nmeth.2645).

Then we choose variable features. The default number of variable features is 2,000, and users can change it using argument `nfeatures`. The default type is highly variable genes (HVGs), but users can use spatially variable genes by seting `type='SVGs'`, then `SPARK-X` will be used to choose SVGs.

## Choose highly variable features
We can also directly use `FindVariableFeatures()` to choose the top highly variable genes (HVGs). We found the performance of using SVGs or HVGs has no significant difference for the downstream analyses, such as dimension reduction and spatial clustering.
```{r  eval = FALSE}
SRTProj <- selectVariableFeatures(SRTProj,nfeatures = 2000, type = "HVGs", method='vst')
SRTProj
```

## Choose spatially variable features
We first applied the SPARK-X to choose the top spatially variable genes (SVGs). SPARK-X is embeded in the `DR.SC` R package and is implement in the `FindSVGs()` function. By default, we return 2,000 genes per dataset.  These will be used in downstream analysis, like probabilistic embeddings and spatial clustering. To further speed up the computation, we first use `FindVariableFeatures()` to select 5,000 highly variable genes, then use SPARK-X to choose top SVGs in the  implementation of `FindSVGs()` function. The selected SVGs are saved in `dlpfc@assays$RNA@var.features`.

```{r  eval = FALSE}
#SRTProj <- selectVariableFeatures(SRTProj,nfeatures = 2000, type = "SVGs", method='SPARK-X')
```

Calculate the adjcence matrix
```{r  eval = FALSE}
## Obtain adjacence matrix
SRTProj <- AddAdj(SRTProj, platform = "Visium")

```
Visualize
```{r, eval = FALSE}
selectFeatures <- row.names(SRTProj@geneMetaData[which(SRTProj@geneMetaData$`isVGs#combine`),])[1:2]
EachExprSpaHeatMap(SRTProj, batch=1:2 ,features=selectFeatures, title_name=T)
```




# Dimension reduction
PCA is the most popular dimension reduction technique in single cell RNA sequencing (scRNA-seq) data and SRT　data because of its simplicity, computational efficiency, and relatively comparable performance. SRTpipeline provided three versions of PCA: standard PCA (PCA), approximated PCA (APCA),  and weighted PCA (WPCA) by the function `AddPCA` with default version as APCA for fastest computation. The 15-dimensional PCs are extracted by default, and users can use their own values.

After running `AddPCA`, we would see the output of `SRTProj` includes `PCA` in the `Low-dimensional embeddings` field.
```{r, eval = FALSE}
SRTProj <- AddPCA(SRTProj, Method='PCA')
#weighted PCA
# SRTProj <- AddPCA(SRTProj, Method='WPCA')
SRTProj
```
# Multiple sample integration analysis using iSC.MEB model 

iSC.MEB model achieves joint clustering and alignment by integration analysis for multiple samples.
Some SRT clustering methods use Markov random field to model  clusters of spots. These approaches work extremely well and are a standard practice in scRNA-seq. For this reason, SRTpipeline uses existing state-of-the-art clustering methods from SRT packages for clustering.

We have had the most success using the graph clustering approach implemented by Seurat. In ArchR, clustering is performed using the addClusters() function which permits additional clustering parameters to be passed to the `SC.MEB::SC.MEB()` function via.
In our hands, clustering using Seurat::FindClusters() is deterministic, meaning that the exact same input will always result in the exact same output.
```{r, eval = FALSE}
SRTProj <- Integrate_iSCMEB(SRTProj, K= 7, reduction = "PCA", init.start=5)
SRTProj
```


The estimated embeddings are saved in the slot`reductions$DR.SC` and the spatial clusters are saved in the slot `clusters$DR.SC`.
```{r, eval = FALSE}
SRTProj@reductions
table(SRTProj@clusters)
```

## SRT embeddings
To run tSNE in SRTpipeline we use the AddTSNE() function:
```{r, eval = FALSE}
SRTProj <- AddTSNE(SRTProj, n_comp = 2, reduction='aligned.iSC.MEB')
```
The `reduction` tells AddTSNE uses specified reduction in the `SRTProj@reductions` slot.

To plot the two-dimensional tSNE results, we use the `EmbedPlot()` function and pass the name of the UMAP embedding we just generated (“UMAP”). We can tell SRTpipeline how to color the spots by using a combination of cols which tells ArchR which matrix to use to find the specified metadata column provided to name.

iSC.MEB model can oﬀer good data visualization. We
visualized the inferred embeddings for biological eﬀects between domain types using two
components from tSNE. The tSNE plot showed the domain clusters were well segregated.
```{r, eval = FALSE}
p_tsne2_cluster <- EmbedPlot(SRTProj, item='cluster', plotEmbeddings = 'tSNE', cols=chooseColors(n_colors = 7), legend.position='bottom', pt_size=0.2)
p_tsne2_batch <- EmbedPlot(SRTProj, item='batch', plotEmbeddings = 'tSNE', cols=chooseColors(palettes_name = 'Light 13',n_colors = 12), legend.position='bottom', pt_size=0.2)
drawFigs(list(p_tsne2_cluster, p_tsne2_batch),layout.dim = c(1,2),
         legend.position = 'bottom')
```

Next, we summarized the inferred embeddings for biological eﬀects
between spatial domain types (the slot `reductions`) using three components from either tSNE or UMAP and visualized the resulting tSNE/UMAP components with red/green/blue (RGB) colors in the RGB plot.

The resulting RGB plots from iSC.MEB showed the laminar organization of the human cerebral cortex, and iSC.MEB provided smooth transitions across neighboring spots and spatial domains.
```{r, eval = FALSE}
SRTProj <- AddTSNE(SRTProj, n_comp = 3, reduction='aligned.iSC.MEB')
p_tsne3<- EachRGBSpaHeatMap(SRTProj, plot_type = "tSNE",layout.dim=c(1,3), pt_size=2 ,title_name="tSNE RGB plot: ")
p_tsne3
```

To run UMAP in SRTpipeline we use the AddUMAP() function. Frist, we evaluate the two-dimensional UMAPs.

```{r, eval = FALSE}
SRTProj <- AddUMAP(SRTProj, n_comp = 2, reduction='aligned.iSC.MEB')
p_umap2 <- EmbedPlot(SRTProj, item='cluster', plotEmbeddings = 'UMAP', cols=chooseColors(n_colors = 7), legend.position='right')
p_umap2
```

Then, we evaluate the three-dimensional UMAPs.
```{r, eval = FALSE}
SRTProj <- AddUMAP(SRTProj, n_comp = 3, reduction='aligned.iSC.MEB')
p_umap3<- EachRGBSpaHeatMap(SRTProj, plot_type = "UMAP", layout.dim=c(1,3), pt_size=2 ,title_name="UMAP RGB plot: ")

p_umap3
```


To save the plot, we can use `write_fig()` function.
```{r, eval = FALSE}
write_fig(p_umap3, filename = 'iSCMEB_p_umap3.png',width =14, height =11)
```

## Visualization
Except for the embedding plots, SRTpipeline also provides a variaty of visualization functions.
First, we visualize the spatial distribution of cluster labels that shows the layer structure for all 12 data batches.
```{r, eval = FALSE}
## choose colors to function chooseColors
cols <- chooseColors(n_colors = 7)
p12 <- EachClusterSpaHeatMap(SRTProj, cols=cols, legend.position='bottom',base_size=12,
                  pt_size=0.5, layout.dim=c(1,3), nrow.legend=1)
p12
# remove the border
# EachClusterSpaHeatMap(SRTProj, cols=cols, legend.position='right',base_size=16, border_col='white')
# write_fig(p12, filename = 'iSCMEB_p12.png',width =14, height =11)
```

By setting `combine =FALSE`, this function will return a list of ggplot2 objects, thus user can revise each plot.
```{r, eval = FALSE}
pList <- EachClusterSpaHeatMap(SRTProj, cols=cols, legend.position='bottom',base_size=12,
                  pt_size=0.5, layout.dim=c(1,3), nrow.legend=1, combine=FALSE)
## Calculate the ARIs and NMIs for each data batch
annotateList <- vec2list(SRTProj@cellMetaData$annotated_label, nvec=SRTProj@sampleMetaData$NumOfSpots)
clusterList <- vec2list(SRTProj@clusters, nvec=SRTProj@sampleMetaData$NumOfSpots)
ARI_vec <- sapply(1:length(annotateList), function(r) cluster_metric(annotateList[[r]], clusterList[[r]]))
ARI_vec
```

We plotted the heatmap of Pearson’s correlation coefcients of the estimated embeddings among
the detected domains shows the good separation of the estimated embeddings across
domains and the correlations between deeper layers were high, e.g., there were high
correlations between domain 2 and 3, while correlations among the separated layers were low, i.e., domain 1 and 4.

```{r, eval = FALSE}
p_cc <- CCHeatMap(SRTProj, reduction = 'aligned.iSC.MEB', grp_color=cols)
p_cc
```

After adding the quantities for data visualization, the `SRTProject` object will have more information in the downstream analyses information.
Now, we print this `SRTProject` object to check it. We observed two components added in the slot `plotEmbeddings` (Embeddings for plotting): `tSNE`, `tSNE3`, `UMAP` and `UMAP3`.
```{r, eval = FALSE}
SRTProj
```




# Combined DEG analysis
After obtain the spatial cluster labels using a clustering model, we can perform differentially expression analysis. 
```{r, eval = FALSE}
speInt <- getIntegratedData(SRTProj, Method='iSC.MEB', species="Human")

```

First, we detect the DE genes for cluster 3.
```{r, eval = FALSE}
dat_deg <- FindDEGs(speInt, cluster1='3')
subset(dat_deg, avg_log2FC > 0.5)
```
We perform differential expression analysis for all clusters by using `FindAllMarkers()` function, then the DE genes' information is saved in a data.frame object `dat_degs`.
```{r, eval = FALSE}
dat_degs <- FindAllDEGs(speInt)
dat_degs
```
We identify the significant DE genes by two criteria: (a) adjustd p-value less than 0.01 and (b) average log fold change greater than 0.4.

```{r, eval = FALSE}
degs_sig <- subset(dat_degs, p_val_adj < 0.01 & avg_log2FC>0.25)
degs_sig
```

In the following, we perform gene set enrichment analysis for the DE genes of each  Domain identified by iSC.MEB model using R package `gprofiler2`.
```{r, eval = FALSE}
library(gprofiler2)
termList <- list()
for(k in 1: 7){
  # k <- 1
  cat("k = ", k, '\n')
  dat_degs_sub <- subset(degs_sig,  cluster==k)

  que1 <- dat_degs_sub$gene
  gostres <- gost(query = que1, organism = "hsapiens", correction_method='fdr')
  termList[[k]] <- gostres
}
head(termList[[1]]$result)
```

To understand the functions of the identified spatial domains by iSC.MEB model, we compare the top significant biological process (BP)  pathways in GO database for the DE genes from Domain 1 and 2. Here, we only show to visualize the significant BP pathways and users can explore the other databases such as KEGG and HPA.


```{r, eval = FALSE}
## Most commonly used databases
source_set <- c("GO:BP","GO:CC", "GO:MF",   'KEGG', "HPA")
cols <- c("steelblue3", "goldenrod", "brown3", "#f98866", "#CE6DBD")
## Here, we show GO:BP
source1 <- "GO:BP"
ss <- which(source_set==source1)
ntop = 5
names(cols) <- source_set
pList_enrich <- list()
for(ii in 1: 5){
  ## ii <- 3
  message("ii=", ii)
  gostres2 <- termList[[ii]]
  if(!is.null(gostres2$result)){
    dat1 <- subset(gostres2$result, term_size< 500)
  dat1 <- get_top_pathway(dat1, ntop=ntop, source_set = source1)
  dat1 <- dat1[complete.cases(dat1),]
  dat1$nlog10P <- -log10(dat1$p_value)

  pList_enrich[[ii]] <-  barPlot_enrich(dat1[order(dat1$nlog10P),], source='source', 'term_name',
                                        'nlog10P', cols=cols[source_set[ss]],base_size=14) + ylab("-log10(p-adj)")+
    xlab("Biological terms") + ggtitle(paste0("Domain", ii))
  }
  
}

drawFigs(pList_enrich[c(1,2)], layout.dim = c(2,1), common.legend = T, align='hv')

```

We take out the top DE genes for each cluster for visualization.
```{r, eval = FALSE}

library(dplyr)
n <- 5
dat_degs %>% as.data.frame %>%
  group_by(cluster) %>%
  top_n(n = n, wt = avg_log2FC) -> topGene
topGene
```
We visualize the DE genes for each cluster group by gene-by-cell heatmap using the `GCHeatMap()` function.
```{r, eval = FALSE}

cols_cluster <- chooseColors(n_colors = 7)

p1 <- GCHeatMap(speInt, features=topGene$gene, grp_color=cols_cluster ,y_text_size=9)
p1
```


# Trajectory inference
Next, we performed trajectory inference using the embeddings and domain labels estimated by iSC.MEB model.
```{r, eval = FALSE}
speInt <- AddTrajectory(speInt, reduction= 'aligned.iSC.MEB')

EmbedPlot(speInt,  plotEmbeddings = 'tSNE', colour_by='PT')
```

Visualize the inferred pseudotime on the spatial coordinates for each data batch.
```{r, eval = FALSE}
p_spa <- EachEmbedPlot(speInt, reduction = 'Coord', colour_by='PT', layout.dim = c(1, 3), add_border_box = T,
                       no_axis = TRUE)
p_spa 

# save(SRTProj, file=paste0(SRTProj@projectMetadata$outputPath,"/SRTProj.rds"))
# load("F://Research paper//IntegrateDRcluster//AnalysisCode//SRTpipeline//vignettes//DLPFC12V2//SRTProj.rds")

```



# Other downstream analyses


<details>
  <summary>**Session Info**</summary>
```{r}
sessionInfo()
```
</details>
