<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Seurat - Guided Clustering Tutorial • SRTtools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Seurat - Guided Clustering Tutorial">
<meta property="og:description" content="SRTpipeline">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SRTtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../articles/Installation.html">
    <span class="fas fa-book"></span>
     
    Install
  </a>
</li>
<li>
  <a href="../articles/get_started.html">
    <span class="fas fa-book"></span>
     
    Get Started
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fas fa-book"></span>
     
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/dlpfc_tutorial.html">DLPFC tutorial</a>
    </li>
    <li>
      <a href="../articles/DR_SC_introduction.html">Spatial DR and clustering introduction</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/faq.html">
    <span class="fa fa-question-circle-o"></span>
     
    FAQ
  </a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions
  </a>
</li>
<li>
  <a href="https://github.com/feiyoung/SRTtools/" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
    Archive
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Seurat - Guided Clustering Tutorial</h1>
            
            <h4 data-toc-skip class="date">Compiled: October 11, 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/feiyoung/SRTpipeline/blob/HEAD/vignettes/DR_SC_introduction.Rmd" class="external-link"><code>vignettes/DR_SC_introduction.Rmd</code></a></small>
      <div class="hidden name"><code>DR_SC_introduction.Rmd</code></div>

    </div>

    
    
<hr>
<div class="section level2">
<h2 id="setup-the-seurat-object">Setup the Seurat Object<a class="anchor" aria-label="anchor" href="#setup-the-seurat-object"></a>
</h2>
<p>For this tutorial, we will be analyzing the a dataset of Peripheral Blood Mononuclear Cells (PBMC) freely available from 10X Genomics. There are 2,700 single cells that were sequenced on the Illumina NextSeq 500. The raw data can be found <a href="https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz" class="external-link">here</a>.</p>
<p>We start by reading in the data. The <code><a href="https://satijalab.org/seurat/reference/Read10X.html" class="external-link">Read10X()</a></code> function reads in the output of the <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger" class="external-link">cellranger</a> pipeline from 10X, returning a unique molecular identified (UMI) count matrix. The values in this matrix represent the number of molecules for each feature (i.e. gene; row) that are detected in each cell (column).</p>
<p>We next use the count matrix to create a <code>Seurat</code> object. The object serves as a container that contains both data (like the count matrix) and analysis (like PCA, or clustering results) for a single-cell dataset. For a technical discussion of the <code>Seurat</code> object structure, check out our <a href="https://github.com/satijalab/seurat/wiki" class="external-link">GitHub Wiki</a>. For example, the count matrix is stored in <code>pbmc[["RNA"]]@counts</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></code></pre></div>
<details><p><summary><strong>What does data in a count matrix look like?</strong></summary></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R">  <span class="co"># Lets examine a few genes in the first thirty cells</span></code></pre></div>
<p>The <code>.</code> values in the matrix represent 0s (no molecules detected). Since most values in an scRNA-seq matrix are 0, Seurat uses a sparse-matrix representation whenever possible. This results in significant memory and speed savings for Drop-seq/inDrop/10x data.</p>
</details><details><p><summary><strong>Session Info</strong></summary></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R">  <span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.1.2 (2021-11-01)</span>
  <span class="co">## Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
  <span class="co">## Running under: Windows 10 x64 (build 22000)</span>
  <span class="co">## </span>
  <span class="co">## Matrix products: default</span>
  <span class="co">## </span>
  <span class="co">## locale:</span>
  <span class="co">## [1] LC_COLLATE=English_Singapore.1252  LC_CTYPE=English_Singapore.1252   </span>
  <span class="co">## [3] LC_MONETARY=English_Singapore.1252 LC_NUMERIC=C                      </span>
  <span class="co">## [5] LC_TIME=English_Singapore.1252    </span>
  <span class="co">## system code page: 936</span>
  <span class="co">## </span>
  <span class="co">## attached base packages:</span>
  <span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
  <span class="co">## </span>
  <span class="co">## other attached packages:</span>
  <span class="co">## [1] patchwork_1.1.1    sp_1.5-0           SeuratObject_4.1.0 Seurat_4.1.1      </span>
  <span class="co">## [5] dplyr_1.0.9       </span>
  <span class="co">## </span>
  <span class="co">## loaded via a namespace (and not attached):</span>
  <span class="co">##   [1] Rtsne_0.16            colorspace_2.0-3      deldir_1.0-6         </span>
  <span class="co">##   [4] ellipsis_0.3.2        ggridges_0.5.3        rprojroot_2.0.3      </span>
  <span class="co">##   [7] fs_1.5.2              spatstat.data_2.2-0   rstudioapi_0.13      </span>
  <span class="co">##  [10] leiden_0.4.2          listenv_0.8.0         ggrepel_0.9.1        </span>
  <span class="co">##  [13] fansi_1.0.3           codetools_0.2-18      splines_4.1.2        </span>
  <span class="co">##  [16] cachem_1.0.6          knitr_1.37            polyclip_1.10-0      </span>
  <span class="co">##  [19] jsonlite_1.8.0        ica_1.0-2             cluster_2.1.2        </span>
  <span class="co">##  [22] png_0.1-7             rgeos_0.5-9           uwot_0.1.11          </span>
  <span class="co">##  [25] spatstat.sparse_2.1-1 shiny_1.7.1           sctransform_0.3.3    </span>
  <span class="co">##  [28] compiler_4.1.2        httr_1.4.3            assertthat_0.2.1     </span>
  <span class="co">##  [31] Matrix_1.4-0          fastmap_1.1.0         lazyeval_0.2.2       </span>
  <span class="co">##  [34] cli_3.2.0             later_1.3.0           formatR_1.11         </span>
  <span class="co">##  [37] htmltools_0.5.2       tools_4.1.2           igraph_1.3.2         </span>
  <span class="co">##  [40] gtable_0.3.0          glue_1.6.2            RANN_2.6.1           </span>
  <span class="co">##  [43] reshape2_1.4.4        Rcpp_1.0.8.3          scattermore_0.8      </span>
  <span class="co">##  [46] jquerylib_0.1.4       pkgdown_2.0.2         vctrs_0.4.1          </span>
  <span class="co">##  [49] nlme_3.1-155          progressr_0.10.1      lmtest_0.9-40        </span>
  <span class="co">##  [52] spatstat.random_2.2-0 xfun_0.29             stringr_1.4.0        </span>
  <span class="co">##  [55] globals_0.15.0        mime_0.12             miniUI_0.1.1.1       </span>
  <span class="co">##  [58] lifecycle_1.0.1       irlba_2.3.5           goftest_1.2-3        </span>
  <span class="co">##  [61] future_1.26.1         MASS_7.3-55           zoo_1.8-10           </span>
  <span class="co">##  [64] scales_1.2.0          spatstat.core_2.4-4   spatstat.utils_2.3-1 </span>
  <span class="co">##  [67] ragg_1.2.2            promises_1.2.0.1      parallel_4.1.2       </span>
  <span class="co">##  [70] RColorBrewer_1.1-3    yaml_2.3.5            gridExtra_2.3        </span>
  <span class="co">##  [73] memoise_2.0.1         reticulate_1.25       pbapply_1.5-0        </span>
  <span class="co">##  [76] ggplot2_3.3.6         sass_0.4.1            rpart_4.1.16         </span>
  <span class="co">##  [79] stringi_1.7.6         desc_1.4.0            rlang_1.0.2          </span>
  <span class="co">##  [82] pkgconfig_2.0.3       systemfonts_1.0.4     matrixStats_0.62.0   </span>
  <span class="co">##  [85] evaluate_0.15         lattice_0.20-45       tensor_1.5           </span>
  <span class="co">##  [88] ROCR_1.0-11           purrr_0.3.4           htmlwidgets_1.5.4    </span>
  <span class="co">##  [91] cowplot_1.1.1         tidyselect_1.1.2      parallelly_1.32.0    </span>
  <span class="co">##  [94] RcppAnnoy_0.0.19      plyr_1.8.7            magrittr_2.0.3       </span>
  <span class="co">##  [97] R6_2.5.1              generics_0.1.2        DBI_1.1.2            </span>
  <span class="co">## [100] mgcv_1.8-39           pillar_1.7.0          fitdistrplus_1.1-8   </span>
  <span class="co">## [103] abind_1.4-5           survival_3.2-13       tibble_3.1.7         </span>
  <span class="co">## [106] future.apply_1.9.0    crayon_1.5.1          KernSmooth_2.23-20   </span>
  <span class="co">## [109] utf8_1.2.2            spatstat.geom_2.4-0   plotly_4.10.0        </span>
  <span class="co">## [112] rmarkdown_2.11        grid_4.1.2            data.table_1.14.2    </span>
  <span class="co">## [115] digest_0.6.29         xtable_1.8-4          tidyr_1.2.0          </span>
  <span class="co">## [118] httpuv_1.6.5          textshaping_0.3.6     munsell_0.5.0        </span>
  <span class="co">## [121] viridisLite_0.4.0     bslib_0.3.1</span></code></pre>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
