<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Fast and accurate multiple data integration • SRTpipeline</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Fast and accurate multiple data integration">
<meta property="og:description" content="SRTpipeline">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SRTpipeline</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../articles/Installation.html">
    <span class="fas fa-book"></span>
     
    Install
  </a>
</li>
<li>
  <a href="../articles/get_started.html">
    <span class="fas fa-book"></span>
     
    Get Started
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fas fa-book"></span>
     
    Single Data Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/single_basic_tutorial.html">Basic tutorial</a>
    </li>
    <li>
      <a href="../articles/dlpfc_tutorial.html">Embedding and clustering introduction</a>
    </li>
    <li>
      <a href="../articles/single_Embryo_SC_tutorial.html">SRT data clustering</a>
    </li>
    <li>
      <a href="../articles/single_Embryo_DRSC_tutorial.html">SRT data embedding and clustering</a>
    </li>
    <li>
      <a href="../articles/single_MOB_Anno_tutorial.html">SRT data embedding and annotation</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fas fa-book"></span>
     
    Multiple Data Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/multibatch_basic_tutorial.html">Basic tutorial</a>
    </li>
    <li>
      <a href="../articles/integration_introduction.html">Basic tutorial v2</a>
    </li>
    <li>
      <a href="../articles/multibatch_brain12_tutorial.html">SRT data clustering and integration</a>
    </li>
    <li>
      <a href="../articles/multibatch_HCC4_tutorial.html">SRT data embedding, clustering and integration example 1</a>
    </li>
    <li>
      <a href="../articles/multibatch_liver8_tutorial.html">SRT data embedding, clustering and integration example 2</a>
    </li>
    <li>
      <a href="../articles/multibatch_MOB16_tutorial.html">SRT data embedding, clustering and integration example 3</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/faq.html">
    <span class="fa fa-question-circle-o"></span>
     
    FAQ
  </a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Fast and accurate multiple data integration</h1>
            
            <h4 data-toc-skip class="date">Compiled: 二月 21, 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/feiyoung/SRTpipeline/blob/HEAD/vignettes/integration_large_datasets.Rmd"><code>vignettes/integration_large_datasets.Rmd</code></a></small>
      <div class="hidden name"><code>integration_large_datasets.Rmd</code></div>

    </div>

    
    
<hr>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This tutorial demonstrates how to use <code>SRTpipeline</code> (&gt;=0.1.0) to analyze multiple spatially-resolved transcriptomics (SRT) data. For very large datasets (such as the number of spots greater than 1000k), the <code>PRECAST</code> model can sometimes be prohibitively computationally expensive since this model jointly performs dimension reduction, spatial clustering and embedding alignment. As an alternative, <code>iSC.MEB</code> performs spatial clustering and embedding alignment on the principal component scores (PCs) space rather than the gene expression space, which can greatly increase the computational efficiency. In this workflow, based on <code>iSC.MEB</code> model, we employ two tips that can further improve its efficiency and runtimes:</p>
<ul>
<li>Approximate PCA to obtain PCs</li>
<li>Choose the number of clusters using prior information or other method</li>
</ul>
<p>The main one of efficiency improvements are gained in <code><a href="../reference/AddPCA.html">AddPCA()</a></code> by setting the <code>Method='APCA'</code>. This means one uses an approximate PCA to obtain the PCs.</p>
<p>First, we load <code>SRTpipeline</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feiyoung/SRTpipeline">SRTpipeline</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2023</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dataset">Dataset<a class="anchor" aria-label="anchor" href="#dataset"></a>
</h2>
<p>For this tutorial, we will introduce how to create a <code>SRTProject</code> object with multiple SRT samples using SRTpipeline that includes an introduction to common analytical workflows for multiple data batches. Here, we will be taking spatial transcriptomics dataset for human dorsolateral prefrontal cortex (DLPFC) as an example. There are 12 tissue slices with 3500~4500 spots and 33538 genes that were sequenced on the 10x Visium platform. Our preprocessed data can be downloaded <a href="https://github.com/feiyoung/DR-SC.Analysis/blob/main/data/DLPFC_data/" class="external-link">here</a>, and the raw data can be found <a href="https://github.com/LieberInstitute/spatialLIBD" class="external-link">here</a>.</p>
<p>Next, we download the data to the current working path for the followed analysis by the following command:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">name_ID12</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">151507</span>, <span class="fl">151508</span>, <span class="fl">151509</span>, <span class="fl">151510</span>, <span class="fl">151669</span>, <span class="fl">151670</span>, <span class="fl">151671</span>, <span class="fl">151672</span>, <span class="fl">151673</span>,
    <span class="fl">151674</span>, <span class="fl">151675</span>, <span class="fl">151676</span><span class="op">)</span><span class="op">)</span>

<span class="kw">for</span> <span class="op">(</span><span class="va">r</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">name_ID12</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">githubURL</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"https://github.com/feiyoung/DR-SC.Analysis/blob/main/data/DLPFC_data/"</span>,
        <span class="va">name_ID12</span><span class="op">[</span><span class="va">r</span><span class="op">]</span>, <span class="st">".rds?raw=true"</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">githubURL</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"dlpfc_"</span>, <span class="va">name_ID12</span><span class="op">[</span><span class="va">r</span><span class="op">]</span>, <span class="st">".rds"</span><span class="op">)</span>, mode <span class="op">=</span> <span class="st">"wb"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="pre-processing-workflow">Pre-processing workflow<a class="anchor" aria-label="anchor" href="#pre-processing-workflow"></a>
</h2>
<p>Then load to R and prepare for creating the <code>SRTProject</code> object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dir_file</span> <span class="op">&lt;-</span> <span class="st">"D:/LearnFiles/Research paper/ProPCA/DR-SC.Analysis/data/DLPFC_data/"</span>
<span class="va">name_ID12</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">151507</span>, <span class="fl">151508</span>, <span class="fl">151509</span>, <span class="fl">151510</span>, <span class="fl">151669</span>, <span class="fl">151670</span>, <span class="fl">151671</span>, <span class="fl">151672</span>, <span class="fl">151673</span>,
    <span class="fl">151674</span>, <span class="fl">151675</span>, <span class="fl">151676</span><span class="op">)</span><span class="op">)</span>
<span class="co"># dir_file &lt;- './dlpfc_'</span>
<span class="va">n_sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">name_ID12</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>
<span class="co">## create count matrix list: note each component has a name, i.e., `ID151672`.</span>
<span class="va">cntList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">## create spatial coordinate matrix</span>
<span class="va">coordList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">## create metadata list</span>
<span class="va">metadataList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>

<span class="kw">for</span> <span class="op">(</span><span class="va">r</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">n_sample</span><span class="op">)</span> <span class="op">{</span>
    <span class="co"># r &lt;- 1</span>
    <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"r = "</span>, <span class="va">r</span><span class="op">)</span>
    <span class="va">id</span> <span class="op">&lt;-</span> <span class="va">name_ID12</span><span class="op">[</span><span class="va">r</span><span class="op">]</span>
    <span class="va">dlpfc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">dir_file</span>, <span class="va">id</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span>
    <span class="va">sp_count</span> <span class="op">&lt;-</span> <span class="va">dlpfc</span><span class="op">@</span><span class="va">assays</span><span class="op">@</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span>
    <span class="va">meta.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>annotated_label <span class="op">=</span> <span class="va">dlpfc</span><span class="op">$</span><span class="va">layer_guess_reordered</span>, row <span class="op">=</span> <span class="va">dlpfc</span><span class="op">$</span><span class="va">row</span>, col <span class="op">=</span> <span class="va">dlpfc</span><span class="op">$</span><span class="va">col</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">meta.data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sp_count</span><span class="op">)</span>
    <span class="va">cntList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">sp_count</span>
    <span class="va">metadataList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">meta.data</span>
    <span class="va">coordList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>row <span class="op">=</span> <span class="va">dlpfc</span><span class="op">$</span><span class="va">row</span>, col <span class="op">=</span> <span class="va">dlpfc</span><span class="op">$</span><span class="va">col</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cntList</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">name_ID12</span>
<span class="co">## create meta data for each data batches. Here we only have one data batch.</span>
<span class="va">sampleMetadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>species <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Human"</span>, <span class="va">n_sample</span><span class="op">)</span>, tissues <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"DLPFC"</span>, <span class="va">n_sample</span><span class="op">)</span>,
    donor <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, each <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">sampleMetadata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cntList</span><span class="op">)</span>
<span class="co">## Name of this project</span>
<span class="va">projectName</span> <span class="op">&lt;-</span> <span class="st">"DLPFC12"</span></code></pre></div>
<p>We check the data by printing it. We can see there are 3500~4500 spots and 33538 genes for each data batch.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">cntList</span>, <span class="va">dim</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##      151507 151508 151509 151510 151669 151670 151671 151672 151673 151674</span>
<span class="co">## [1,]  33538  33538  33538  33538  33538  33538  33538  33538  33538  33538</span>
<span class="co">## [2,]   4226   4384   4789   4634   3661   3498   4110   4015   3639   3673</span>
<span class="co">##      151675 151676</span>
<span class="co">## [1,]  33538  33538</span>
<span class="co">## [2,]   3592   3460</span></code></pre>
<p>Check whether each data batch has the same genes: Yes.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">geneNamesList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">cntList</span>, <span class="va">row.names</span><span class="op">)</span>
<span class="va">shared_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="va">intersect</span>, <span class="va">geneNamesList</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">shared_genes</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 33538</span></code></pre>
<div class="section level3">
<h3 id="create-a-srtproject-object">Create a SRTProject object<a class="anchor" aria-label="anchor" href="#create-a-srtproject-object"></a>
</h3>
<p>Because the gene name is ensembl in this data, we change the genes’ ensembl name to symbols for convenience of followed analyses.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cntList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">cntList</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">[</span><span class="va">shared_genes</span>, <span class="op">]</span><span class="op">)</span>
<span class="co">## Use eg.db database: this method is fast</span>
<span class="va">symbol_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/transferGeneNames.html">transferGeneNames</a></span><span class="op">(</span><span class="va">shared_genes</span>, now_name <span class="op">=</span> <span class="st">"ensembl"</span>, to_name <span class="op">=</span> <span class="st">"symbol"</span>, species <span class="op">=</span> <span class="st">"Human"</span>,
    Method <span class="op">=</span> <span class="st">"eg.db"</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">r</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">n_sample</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">cntList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">symbol_name</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "F:/Research paper/IntegrateDRcluster/AnalysisCode/SRTpipeline/vignettes"</span></code></pre>
<p>Next, we start creating SRTProject object.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CreateSRTProject.html">CreateSRTProject</a></span><span class="op">(</span><span class="va">cntList</span>, <span class="va">coordList</span>, projectName <span class="op">=</span> <span class="va">projectName</span>, <span class="va">metadataList</span>, <span class="va">sampleMetadata</span>,
    min.spots <span class="op">=</span> <span class="fl">20</span>, min.genes <span class="op">=</span> <span class="fl">20</span>, force <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>
<span class="va">SRTProj</span></code></pre></div>
<pre><code><span class="co">## class: SRTProject </span>
<span class="co">## outputPath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\DLPFC12 </span>
<span class="co">## h5filePath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\DLPFC12/DLPFC12.h5 </span>
<span class="co">## ---------Datasets basic information-----------------</span>
<span class="co">## samples(12): 151507 151508 ... 151675 151676</span>
<span class="co">## sampleColData names(4): species tissues donor NumOfSpots</span>
<span class="co">## cellMetaData names(4): annotated_label row col batch</span>
<span class="co">## numberOfSpots(12): 4226 4384 ... 3592 3460</span>
<span class="co">## ---------Downstream analyses information-----------------</span>
<span class="co">## Low-dimensional embeddings(0):</span>
<span class="co">## Inferred cluster labels: No</span>
<span class="co">## Embedding for plotting(0):</span></code></pre>
</div>
<div class="section level3">
<h3 id="standard-workflow">Standard workflow<a class="anchor" aria-label="anchor" href="#standard-workflow"></a>
</h3>
<p>After removing unwanted cells and genes from the dataset, the next step is to normalize the data. To save RAM memory, normalized values are stored in disk as a <code>h5file</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Normalizing the data</span>
<span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeSRT.html">normalizeSRT</a></span><span class="op">(</span><span class="va">SRTProj</span>, normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span><span class="op">)</span>
<span class="co">## Choose highly variable features</span>
<span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectVariableFeatures.html">selectVariableFeatures</a></span><span class="op">(</span><span class="va">SRTProj</span>, nfeatures <span class="op">=</span> <span class="fl">2000</span>, type <span class="op">=</span> <span class="st">"HVGs"</span>, method <span class="op">=</span> <span class="st">"vst"</span><span class="op">)</span>
<span class="co">## Calculate the adjcence matrix</span>
<span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddAdj.html">AddAdj</a></span><span class="op">(</span><span class="va">SRTProj</span>, platform <span class="op">=</span> <span class="st">"Visium"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="approximate-pca">Approximate PCA<a class="anchor" aria-label="anchor" href="#approximate-pca"></a>
</h3>
<p>After running <code>AddPCA</code>, we would see the output of <code>SRTProj</code> includes <code>PCA</code> in the <code>Low-dimensional embeddings</code> field.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## APCA</span>
<span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddPCA.html">AddPCA</a></span><span class="op">(</span><span class="va">SRTProj</span>, n_comp <span class="op">=</span> <span class="fl">15</span>, Method <span class="op">=</span> <span class="st">"APCA"</span><span class="op">)</span>
<span class="va">SRTProj</span></code></pre></div>
<pre><code><span class="co">## class: SRTProject </span>
<span class="co">## outputPath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\DLPFC12 </span>
<span class="co">## h5filePath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\DLPFC12/DLPFC12.h5 </span>
<span class="co">## ---------Datasets basic information-----------------</span>
<span class="co">## samples(12): 151507 151508 ... 151675 151676</span>
<span class="co">## sampleColData names(4): species tissues donor NumOfSpots</span>
<span class="co">## cellMetaData names(4): annotated_label row col batch</span>
<span class="co">## numberOfSpots(12): 4226 4384 ... 3592 3460</span>
<span class="co">## ---------Downstream analyses information-----------------</span>
<span class="co">## Variable features:  2000 </span>
<span class="co">## Low-dimensional embeddings(1): PCA</span>
<span class="co">## Inferred cluster labels: No</span>
<span class="co">## Embedding for plotting(0):</span></code></pre>
</div>
<div class="section level3">
<h3 id="fast-integration-using-a-prior-k">Fast integration using a prior K<a class="anchor" aria-label="anchor" href="#fast-integration-using-a-prior-k"></a>
</h3>
<p>Using a proir <span class="math inline">\(K=7\)</span>, iSC.MEB model runs faster.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Integrate_iSCMEB.html">Integrate_iSCMEB</a></span><span class="op">(</span><span class="va">SRTProj</span>, K <span class="op">=</span> <span class="fl">7</span>, reduction <span class="op">=</span> <span class="st">"PCA"</span>, init.start <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></code></pre></div>
<pre><code>## fitting ...
## 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |======================================================================| 100%
## fitting ...
## 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |======================================================================| 100%
## fitting ...
## 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |======================================================================| 100%
## fitting ...
## 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |======================================================================| 100%
## fitting ...
## 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |======================================================================| 100%
## Finish variable initialization 
## K = 7, iter = 2, loglik= -546826.389236, dloglik=0.999745 
## K = 7, iter = 3, loglik= -494662.582570, dloglik=0.095394 
## K = 7, iter = 4, loglik= -463152.836846, dloglik=0.063699 
## K = 7, iter = 5, loglik= -439208.156319, dloglik=0.051699 
## K = 7, iter = 6, loglik= -420801.614243, dloglik=0.041908 
## K = 7, iter = 7, loglik= -406818.642275, dloglik=0.033229 
## K = 7, iter = 8, loglik= -395768.450994, dloglik=0.027162 
## K = 7, iter = 9, loglik= -387641.029581, dloglik=0.020536 
## K = 7, iter = 10, loglik= -381362.532925, dloglik=0.016197 
## K = 7, iter = 11, loglik= -376469.004059, dloglik=0.012832 
## K = 7, iter = 12, loglik= -372658.415320, dloglik=0.010122 
## K = 7, iter = 13, loglik= -369713.676881, dloglik=0.007902 
## K = 7, iter = 14, loglik= -367339.261189, dloglik=0.006422 
## K = 7, iter = 15, loglik= -365608.261517, dloglik=0.004712 
## K = 7, iter = 16, loglik= -364192.290049, dloglik=0.003873 
## K = 7, iter = 17, loglik= -362931.468458, dloglik=0.003462 
## K = 7, iter = 18, loglik= -361939.279290, dloglik=0.002734 
## K = 7, iter = 19, loglik= -361067.606699, dloglik=0.002408 
## K = 7, iter = 20, loglik= -360267.527303, dloglik=0.002216 
## K = 7, iter = 21, loglik= -359582.895453, dloglik=0.001900 
## K = 7, iter = 22, loglik= -358861.690606, dloglik=0.002006 
## K = 7, iter = 23, loglik= -358294.965926, dloglik=0.001579 
## K = 7, iter = 24, loglik= -357773.724866, dloglik=0.001455 
## K = 7, iter = 25, loglik= -357246.946241, dloglik=0.001472</code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">toc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">time_use</span> <span class="op">&lt;-</span> <span class="va">toc</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span> <span class="op">-</span> <span class="va">tic</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>  <span class="co"># 267 seconds</span>

<span class="co">## or use Louvain clustering to find the number of clusters K SRTProj2 &lt;-</span>
<span class="co">## Cluster_Louvain(SRTProj, resolution = 0.4) K &lt;- length(unique(SRTProj2@clusters))</span>
<span class="va">SRTProj</span></code></pre></div>
<pre><code><span class="co">## class: SRTProject </span>
<span class="co">## outputPath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\DLPFC12 </span>
<span class="co">## h5filePath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\DLPFC12/DLPFC12.h5 </span>
<span class="co">## ---------Datasets basic information-----------------</span>
<span class="co">## samples(12): 151507 151508 ... 151675 151676</span>
<span class="co">## sampleColData names(4): species tissues donor NumOfSpots</span>
<span class="co">## cellMetaData names(4): annotated_label row col batch</span>
<span class="co">## numberOfSpots(12): 4226 4384 ... 3592 3460</span>
<span class="co">## ---------Downstream analyses information-----------------</span>
<span class="co">## Variable features:  2000 </span>
<span class="co">## Low-dimensional embeddings(3): PCA microEnv.iSC.MEB aligned.iSC.MEB</span>
<span class="co">## Inferred cluster labels: Yes</span>
<span class="co">## Embedding for plotting(0):</span></code></pre>
</div>
<div class="section level3">
<h3 id="srt-embeddings">SRT embeddings<a class="anchor" aria-label="anchor" href="#srt-embeddings"></a>
</h3>
<p>To run tSNE in SRTpipeline we use the AddTSNE() function:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddTSNE.html">AddTSNE</a></span><span class="op">(</span><span class="va">SRTProj</span>, n_comp <span class="op">=</span> <span class="fl">2</span>, reduction <span class="op">=</span> <span class="st">"aligned.iSC.MEB"</span><span class="op">)</span>
<span class="co">## plot</span>
<span class="va">p_tsne2_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EmbedPlot.html">EmbedPlot</a></span><span class="op">(</span><span class="va">SRTProj</span>, item <span class="op">=</span> <span class="st">"cluster"</span>, plotEmbeddings <span class="op">=</span> <span class="st">"tSNE"</span>, cols <span class="op">=</span> <span class="fu"><a href="../reference/chooseColors.html">chooseColors</a></span><span class="op">(</span>n_colors <span class="op">=</span> <span class="fl">7</span><span class="op">)</span>,
    legend.position <span class="op">=</span> <span class="st">"bottom"</span>, pt_size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>
<span class="va">p_tsne2_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EmbedPlot.html">EmbedPlot</a></span><span class="op">(</span><span class="va">SRTProj</span>, item <span class="op">=</span> <span class="st">"batch"</span>, plotEmbeddings <span class="op">=</span> <span class="st">"tSNE"</span>, cols <span class="op">=</span> <span class="fu"><a href="../reference/chooseColors.html">chooseColors</a></span><span class="op">(</span>palettes_name <span class="op">=</span> <span class="st">"Light 13"</span>,
    n_colors <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>, legend.position <span class="op">=</span> <span class="st">"bottom"</span>, pt_size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>
<span class="fu"><a href="../reference/drawFigs.html">drawFigs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p_tsne2_cluster</span>, <span class="va">p_tsne2_batch</span><span class="op">)</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></code></pre></div>
<p><img src="integration_large_datasets_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.1.2 (2021-11-01)</span>
<span class="co">## Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">## Running under: Windows 10 x64 (build 22621)</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">## [1] LC_COLLATE=Chinese (Simplified)_China.936 </span>
<span class="co">## [2] LC_CTYPE=Chinese (Simplified)_China.936   </span>
<span class="co">## [3] LC_MONETARY=Chinese (Simplified)_China.936</span>
<span class="co">## [4] LC_NUMERIC=C                              </span>
<span class="co">## [5] LC_TIME=Chinese (Simplified)_China.936    </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">## [8] methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] colorspace_2.0-3     iSC.MEB_1.0.1        ggplot2_3.3.6       </span>
<span class="co">##  [4] gtools_3.9.2.2       irlba_2.3.5          DR.SC_3.1           </span>
<span class="co">##  [7] spatstat.geom_2.4-0  spatstat.data_3.0-0  Matrix_1.4-0        </span>
<span class="co">## [10] hdf5r_1.3.5          ff_4.0.7             bit_4.0.4           </span>
<span class="co">## [13] rhdf5_2.38.0         org.Hs.eg.db_3.14.0  AnnotationDbi_1.56.2</span>
<span class="co">## [16] IRanges_2.28.0       S4Vectors_0.32.3     Biobase_2.54.0      </span>
<span class="co">## [19] BiocGenerics_0.40.0  sp_1.5-0             SeuratObject_4.1.0  </span>
<span class="co">## [22] Seurat_4.1.1         SRTpipeline_0.1.1   </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] scattermore_0.8             ggthemes_4.2.4             </span>
<span class="co">##   [3] princurve_2.1.6             GiRaF_1.0.1                </span>
<span class="co">##   [5] ragg_1.2.2                  tidyr_1.2.0                </span>
<span class="co">##   [7] bit64_4.0.5                 knitr_1.37                 </span>
<span class="co">##   [9] DelayedArray_0.20.0         data.table_1.14.2          </span>
<span class="co">##  [11] rpart_4.1.16                TrajectoryUtils_1.2.0      </span>
<span class="co">##  [13] KEGGREST_1.34.0             RCurl_1.98-1.6             </span>
<span class="co">##  [15] generics_0.1.2              ScaledMatrix_1.2.0         </span>
<span class="co">##  [17] cowplot_1.1.1               RSQLite_2.2.10             </span>
<span class="co">##  [19] RANN_2.6.1                  future_1.26.1              </span>
<span class="co">##  [21] httpuv_1.6.5                SummarizedExperiment_1.24.0</span>
<span class="co">##  [23] assertthat_0.2.1            viridis_0.6.2              </span>
<span class="co">##  [25] xfun_0.29                   jquerylib_0.1.4            </span>
<span class="co">##  [27] evaluate_0.15               promises_1.2.0.1           </span>
<span class="co">##  [29] fansi_1.0.3                 igraph_1.3.5               </span>
<span class="co">##  [31] DBI_1.1.2                   htmlwidgets_1.5.4          </span>
<span class="co">##  [33] purrr_0.3.4                 ellipsis_0.3.2             </span>
<span class="co">##  [35] dplyr_1.0.9                 ggpubr_0.4.0               </span>
<span class="co">##  [37] backports_1.4.1             deldir_1.0-6               </span>
<span class="co">##  [39] sparseMatrixStats_1.6.0     MatrixGenerics_1.6.0       </span>
<span class="co">##  [41] vctrs_0.4.1                 SingleCellExperiment_1.16.0</span>
<span class="co">##  [43] ROCR_1.0-11                 abind_1.4-5                </span>
<span class="co">##  [45] cachem_1.0.6                withr_2.5.0                </span>
<span class="co">##  [47] PRECAST_1.4                 progressr_0.10.1           </span>
<span class="co">##  [49] sctransform_0.3.3           mclust_5.4.10              </span>
<span class="co">##  [51] goftest_1.2-3               cluster_2.1.2              </span>
<span class="co">##  [53] lazyeval_0.2.2              crayon_1.5.1               </span>
<span class="co">##  [55] SpatialAnno_1.0.0           pkgconfig_2.0.3            </span>
<span class="co">##  [57] labeling_0.4.2              GenomeInfoDb_1.30.1        </span>
<span class="co">##  [59] nlme_3.1-155                vipor_0.4.5                </span>
<span class="co">##  [61] rlang_1.0.2                 globals_0.15.0             </span>
<span class="co">##  [63] lifecycle_1.0.1             miniUI_0.1.1.1             </span>
<span class="co">##  [65] rsvd_1.0.5                  rprojroot_2.0.3            </span>
<span class="co">##  [67] polyclip_1.10-0             matrixStats_0.62.0         </span>
<span class="co">##  [69] lmtest_0.9-40               SC.MEB_1.1                 </span>
<span class="co">##  [71] carData_3.0-5               Rhdf5lib_1.16.0            </span>
<span class="co">##  [73] zoo_1.8-10                  beeswarm_0.4.0             </span>
<span class="co">##  [75] ggridges_0.5.3              png_0.1-7                  </span>
<span class="co">##  [77] viridisLite_0.4.0           bitops_1.0-7               </span>
<span class="co">##  [79] KernSmooth_2.23-20          rhdf5filters_1.6.0         </span>
<span class="co">##  [81] Biostrings_2.62.0           blob_1.2.2                 </span>
<span class="co">##  [83] DelayedMatrixStats_1.16.0   stringr_1.4.0              </span>
<span class="co">##  [85] slingshot_2.2.0             parallelly_1.32.0          </span>
<span class="co">##  [87] spatstat.random_2.2-0       rstatix_0.7.0              </span>
<span class="co">##  [89] ggsignif_0.6.3              beachmat_2.10.0            </span>
<span class="co">##  [91] scales_1.2.0                memoise_2.0.1              </span>
<span class="co">##  [93] magrittr_2.0.3              plyr_1.8.7                 </span>
<span class="co">##  [95] ica_1.0-2                   zlibbioc_1.40.0            </span>
<span class="co">##  [97] compiler_4.1.2              RColorBrewer_1.1-3         </span>
<span class="co">##  [99] fitdistrplus_1.1-8          cli_3.2.0                  </span>
<span class="co">## [101] XVector_0.34.0              listenv_0.8.0              </span>
<span class="co">## [103] patchwork_1.1.1             pbapply_1.5-0              </span>
<span class="co">## [105] formatR_1.11                MASS_7.3-55                </span>
<span class="co">## [107] mgcv_1.8-39                 tidyselect_1.1.2           </span>
<span class="co">## [109] stringi_1.7.6               textshaping_0.3.6          </span>
<span class="co">## [111] highr_0.9                   yaml_2.3.6                 </span>
<span class="co">## [113] BiocSingular_1.10.0         ggrepel_0.9.1              </span>
<span class="co">## [115] grid_4.1.2                  sass_0.4.1                 </span>
<span class="co">## [117] tools_4.1.2                 future.apply_1.9.0         </span>
<span class="co">## [119] rstudioapi_0.13             gridExtra_2.3              </span>
<span class="co">## [121] farver_2.1.0                Rtsne_0.16                 </span>
<span class="co">## [123] digest_0.6.29               rgeos_0.5-9                </span>
<span class="co">## [125] shiny_1.7.1                 Rcpp_1.0.10                </span>
<span class="co">## [127] GenomicRanges_1.46.1        car_3.0-12                 </span>
<span class="co">## [129] broom_0.7.12                scuttle_1.4.0              </span>
<span class="co">## [131] later_1.3.0                 RcppAnnoy_0.0.19           </span>
<span class="co">## [133] httr_1.4.3                  fs_1.5.2                   </span>
<span class="co">## [135] tensor_1.5                  reticulate_1.25            </span>
<span class="co">## [137] splines_4.1.2               uwot_0.1.11                </span>
<span class="co">## [139] spatstat.utils_3.0-1        pkgdown_2.0.6              </span>
<span class="co">## [141] scater_1.25.1               plotly_4.10.0              </span>
<span class="co">## [143] systemfonts_1.0.4           xtable_1.8-4               </span>
<span class="co">## [145] jsonlite_1.8.0              R6_2.5.1                   </span>
<span class="co">## [147] pillar_1.7.0                htmltools_0.5.2            </span>
<span class="co">## [149] mime_0.12                   glue_1.6.2                 </span>
<span class="co">## [151] fastmap_1.1.0               BiocParallel_1.28.3        </span>
<span class="co">## [153] BiocNeighbors_1.12.0        codetools_0.2-18           </span>
<span class="co">## [155] utf8_1.2.2                  lattice_0.20-45            </span>
<span class="co">## [157] bslib_0.3.1                 spatstat.sparse_2.1-1      </span>
<span class="co">## [159] tibble_3.1.7                ggbeeswarm_0.6.0           </span>
<span class="co">## [161] leiden_0.4.2                survival_3.2-13            </span>
<span class="co">## [163] CompQuadForm_1.4.3          rmarkdown_2.11             </span>
<span class="co">## [165] desc_1.4.0                  munsell_0.5.0              </span>
<span class="co">## [167] GenomeInfoDbData_1.2.7      reshape2_1.4.4             </span>
<span class="co">## [169] gtable_0.3.0                spatstat.core_2.4-4</span></code></pre>

</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Wei Liu, Yi Yang, Xu Liao, Xiao Zhang, Jin Liu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
