<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A basic overview of multiple data analysis • SRTpipeline</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="A basic overview of multiple data analysis">
<meta property="og:description" content="SRTpipeline">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SRTpipeline</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../articles/Installation.html">
    <span class="fas fa-book"></span>
     
    Install
  </a>
</li>
<li>
  <a href="../articles/get_started.html">
    <span class="fas fa-book"></span>
     
    Get Started
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fas fa-book"></span>
     
    Single Data Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/single_basic_tutorial.html">Basic tutorial</a>
    </li>
    <li>
      <a href="../articles/dlpfc_tutorial.html">Embedding and clustering introduction</a>
    </li>
    <li>
      <a href="../articles/single_Embryo_SC_tutorial.html">SRT data clustering</a>
    </li>
    <li>
      <a href="../articles/single_Embryo_DRSC_tutorial.html">SRT data embedding and clustering</a>
    </li>
    <li>
      <a href="../articles/single_MOB_Anno_tutorial.html">SRT data embedding and annotation</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fas fa-book"></span>
     
    Multiple Data Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/multibatch_basic_tutorial.html">Basic tutorial</a>
    </li>
    <li>
      <a href="../articles/integration_introduction.html">Basic tutorial v2</a>
    </li>
    <li>
      <a href="../articles/multibatch_brain12_tutorial.html">SRT data clustering and integration</a>
    </li>
    <li>
      <a href="../articles/multibatch_HCC4_tutorial.html">SRT data embedding, clustering and integration example 1</a>
    </li>
    <li>
      <a href="../articles/multibatch_liver8_tutorial.html">SRT data embedding, clustering and integration example 2</a>
    </li>
    <li>
      <a href="../articles/multibatch_MOB16_tutorial.html">SRT data embedding, clustering and integration example 3</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/faq.html">
    <span class="fa fa-question-circle-o"></span>
     
    FAQ
  </a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>A basic overview of multiple data analysis</h1>
            
            <h4 data-toc-skip class="date">Compiled: 二月 14, 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/feiyoung/SRTpipeline/blob/HEAD/vignettes/multibatch_basic_tutorial.Rmd"><code>vignettes/multibatch_basic_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>multibatch_basic_tutorial.Rmd</code></div>

    </div>

    
    
<hr>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This tutorial demonstrates how to use <code>SRTpipeline</code> (&gt;=0.1.0) to analyze multiple spatially-resolved transcriptomics (SRT) data. We emphases the workflow based on the <code>SRTProject</code> object in the <code>SRTpipeline</code> package. This tutorial will cover the following tasks, which we believe will be common for many SRT data analyses:</p>
<ul>
<li>Normalization</li>
<li>Feature selection</li>
<li>Spatial clustering</li>
<li>Data integration</li>
<li>Batch-corrected gene expression</li>
<li>DEG analysis</li>
<li>Spatial trajectory inference</li>
<li>Visualization</li>
</ul>
<p>First, we load <code>SRTpipeline</code> package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feiyoung/SRTpipeline">SRTpipeline</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2023</span><span class="op">)</span>  <span class="co"># set a random seed for reproducibility</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dataset">Dataset<a class="anchor" aria-label="anchor" href="#dataset"></a>
</h2>
<p>For this tutorial, we will introduce how to create a <code>SRTProject</code> object with multiple SRT samples using SRTpipeline that includes an introduction to common analytical workflows for multiple data batches. Here, we will be taking spatial transcriptomics dataset for mouse liver as an example. There are eight tissue slices with 450~700 spots and 8500~9500 genes that were sequenced on the ST platform. Our preprocessed data can be downloaded <a href="https://github.com/feiyoung/PRECAST/tree/main/vignettes_data/" class="external-link">here</a>, and the raw data can be found <a href="https://zenodo.org/record/4399655" class="external-link">here</a>.</p>
<p>Next, we download the data to the current working path for the followed analysis by the following command:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seuList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"seuList_liver8.RDS"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="pre-processing-workflow">Pre-processing workflow<a class="anchor" aria-label="anchor" href="#pre-processing-workflow"></a>
</h2>
<div class="section level3">
<h3 id="prepare-srtproject">Prepare SRTProject<a class="anchor" aria-label="anchor" href="#prepare-srtproject"></a>
</h3>
<p>Then load to R.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">n_sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">seuList</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>
<span class="co">## create count matrix list</span>
<span class="va">cntList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">## create spatial coordinate matrix</span>
<span class="va">coordList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">## create metadata list</span>
<span class="va">metadataList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>

<span class="kw">for</span> <span class="op">(</span><span class="va">r</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">n_sample</span><span class="op">)</span> <span class="op">{</span>
    <span class="co"># r &lt;- 1</span>
    <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"r = "</span>, <span class="va">r</span><span class="op">)</span>
    <span class="va">seu</span> <span class="op">&lt;-</span> <span class="va">seuList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span>
    <span class="va">sp_count</span> <span class="op">&lt;-</span> <span class="va">seu</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span>
    <span class="va">meta.data</span> <span class="op">&lt;-</span> <span class="va">seu</span><span class="op">@</span><span class="va">meta.data</span>
    <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">meta.data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sp_count</span><span class="op">)</span>
    <span class="va">cntList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">sp_count</span>
    <span class="va">metadataList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">meta.data</span>
    <span class="va">coordList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>row <span class="op">=</span> <span class="va">seu</span><span class="op">$</span><span class="va">row</span>, col <span class="op">=</span> <span class="va">seu</span><span class="op">$</span><span class="va">col</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cntList</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Liver"</span>, <span class="fl">1</span><span class="op">:</span><span class="va">n_sample</span><span class="op">)</span>  <span class="co"># Note each component in count matrix list has a name.</span>
<span class="co">## create meta data for each data batches.</span>
<span class="va">sampleMetadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>species <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Mouse"</span>, <span class="va">n_sample</span><span class="op">)</span>, tissues <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Liver"</span>, <span class="va">n_sample</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">sampleMetadata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cntList</span><span class="op">)</span>
<span class="co">## Name of this project</span>
<span class="va">projectName</span> <span class="op">&lt;-</span> <span class="st">"Liver8"</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-a-srtproject-object">Create a SRTProject object<a class="anchor" aria-label="anchor" href="#create-a-srtproject-object"></a>
</h3>
<p>We start creating SRTProject object. Filter out the genes less than 20 spots with nonzero read counts and the spots less than 20 genes with nonzero read counts.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CreateSRTProject.html">CreateSRTProject</a></span><span class="op">(</span><span class="va">cntList</span>, <span class="va">coordList</span>, projectName <span class="op">=</span> <span class="va">projectName</span>, <span class="va">metadataList</span>, <span class="va">sampleMetadata</span>,
    min.spots <span class="op">=</span> <span class="fl">20</span>, min.genes <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>
<span class="va">SRTProj</span></code></pre></div>
<pre><code><span class="co">## class: SRTProject </span>
<span class="co">## outputPath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\Liver8 </span>
<span class="co">## h5filePath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\Liver8/Liver8.h5 </span>
<span class="co">## ---------Datasets basic information-----------------</span>
<span class="co">## samples(8): Liver1 Liver2 ... Liver7 Liver8</span>
<span class="co">## sampleColData names(3): species tissues NumOfSpots</span>
<span class="co">## cellMetaData names(13): orig.ident nCount_RNA ... sample_id batch</span>
<span class="co">## numberOfSpots(8): 675 684 ... 629 590</span>
<span class="co">## ---------Downstream analyses information-----------------</span>
<span class="co">## Low-dimensional embeddings(0):</span>
<span class="co">## Inferred cluster labels: No</span>
<span class="co">## Embedding for plotting(0):</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="normalizing-the-data">Normalizing the data<a class="anchor" aria-label="anchor" href="#normalizing-the-data"></a>
</h2>
<p>After removing unwanted cells and genes from the dataset, the next step is to normalize the data. To save RAM memory, normalized values are stored in disk as a <code>h5file</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeSRT.html">normalizeSRT</a></span><span class="op">(</span><span class="va">SRTProj</span>, normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="feature-selection">Feature selection<a class="anchor" aria-label="anchor" href="#feature-selection"></a>
</h2>
<p>We next select a subset of genes that exhibit high spot-to-spot variation in the dataset (i.e, they are highly expressed in some spots, and lowly expressed in others). It has been found that focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets <a href="https://www.nature.com/articles/nmeth.2645" class="external-link">here</a>.</p>
<p>Then we choose variable features. The default number of variable features is 2,000, and users can change it using argument <code>nfeatures</code>. The default type is highly variable genes (HVGs), but users can use spatially variable genes by seting <code>type='SVGs'</code>, then <code>SPARK-X</code> will be used to choose SVGs.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectVariableFeatures.html">selectVariableFeatures</a></span><span class="op">(</span><span class="va">SRTProj</span>, nfeatures <span class="op">=</span> <span class="fl">2000</span>, type <span class="op">=</span> <span class="st">"HVGs"</span>, method <span class="op">=</span> <span class="st">"vst"</span><span class="op">)</span>
<span class="va">SRTProj</span></code></pre></div>
<pre><code><span class="co">## class: SRTProject </span>
<span class="co">## outputPath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\Liver8 </span>
<span class="co">## h5filePath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\Liver8/Liver8.h5 </span>
<span class="co">## ---------Datasets basic information-----------------</span>
<span class="co">## samples(8): Liver1 Liver2 ... Liver7 Liver8</span>
<span class="co">## sampleColData names(3): species tissues NumOfSpots</span>
<span class="co">## cellMetaData names(13): orig.ident nCount_RNA ... sample_id batch</span>
<span class="co">## numberOfSpots(8): 675 684 ... 629 590</span>
<span class="co">## ---------Downstream analyses information-----------------</span>
<span class="co">## Variable features:  2000 </span>
<span class="co">## Low-dimensional embeddings(0):</span>
<span class="co">## Inferred cluster labels: No</span>
<span class="co">## Embedding for plotting(0):</span></code></pre>
<p>Calculate the adjcence matrix</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Obtain adjacency matrix</span>
<span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddAdj.html">AddAdj</a></span><span class="op">(</span><span class="va">SRTProj</span>, platform <span class="op">=</span> <span class="st">"ST"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dimension-reduction">Dimension reduction<a class="anchor" aria-label="anchor" href="#dimension-reduction"></a>
</h2>
<p>PCA is the most popular dimension reduction technique in single cell RNA sequencing (scRNA-seq) data and SRT　data because of its simplicity, computational efficiency, and relatively comparable performance. SRTpipeline provided three versions of PCA: standard PCA (PCA), approximated PCA (APCA), and weighted PCA (WPCA) by the function <code>AddPCA</code> with default version as APCA for fastest computation. The 15-dimensional PCs are extracted by default, and users can use their own values.</p>
<p>After running <code>AddPCA</code>, we would see the output of <code>SRTProj</code> includes <code>PCA</code> in the <code>Low-dimensional embeddings</code> field.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## APCA</span>
<span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddPCA.html">AddPCA</a></span><span class="op">(</span><span class="va">SRTProj</span>, n_comp <span class="op">=</span> <span class="fl">15</span><span class="op">)</span>

<span class="co"># accurate PCA SRTProj &lt;- AddPCA(SRTProj, Method='PCA')</span>

<span class="co"># weighted PCA SRTProj &lt;- AddPCA(SRTProj, Method='WPCA')</span>
<span class="va">SRTProj</span></code></pre></div>
<pre><code><span class="co">## class: SRTProject </span>
<span class="co">## outputPath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\Liver8 </span>
<span class="co">## h5filePath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\Liver8/Liver8.h5 </span>
<span class="co">## ---------Datasets basic information-----------------</span>
<span class="co">## samples(8): Liver1 Liver2 ... Liver7 Liver8</span>
<span class="co">## sampleColData names(3): species tissues NumOfSpots</span>
<span class="co">## cellMetaData names(13): orig.ident nCount_RNA ... sample_id batch</span>
<span class="co">## numberOfSpots(8): 675 684 ... 629 590</span>
<span class="co">## ---------Downstream analyses information-----------------</span>
<span class="co">## Variable features:  2000 </span>
<span class="co">## Low-dimensional embeddings(1): PCA</span>
<span class="co">## Inferred cluster labels: No</span>
<span class="co">## Embedding for plotting(0):</span></code></pre>
</div>
<div class="section level2">
<h2 id="data-clustering-and-integration-analysis-using-isc-meb-model">Data clustering and integration analysis using iSC.MEB model<a class="anchor" aria-label="anchor" href="#data-clustering-and-integration-analysis-using-isc-meb-model"></a>
</h2>
<p><code>iSC.MEB</code> model achieves joint clustering and alignment by integration analysis for multiple samples. Some SRT clustering methods use Markov random field to model clusters of spots. These approaches work extremely well when the spots have strong spatial dependence. <code>iSC.MEB</code> extracts the micro-environment related embeddings and aligned embeddings. The micro-environment related and aligned embeddings are saved in the slot <code>reductions$microEnv.iSC.MEB</code> and <code>reductions$aligned.iSC.MEB</code>, and the spatial clusters are saved in the slot <code>clusters</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Integrate_iSCMEB.html">Integrate_iSCMEB</a></span><span class="op">(</span><span class="va">SRTProj</span>, K <span class="op">=</span> <span class="fl">6</span>, reduction <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></code></pre></div>
<pre><code>## fitting ...
## 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |======================================================================| 100%
## Finish variable initialization 
## K = 6, iter = 2, loglik= -39897.400172, dloglik=0.999981 
## K = 6, iter = 3, loglik= -36353.590917, dloglik=0.088823 
## K = 6, iter = 4, loglik= -35341.453489, dloglik=0.027841 
## K = 6, iter = 5, loglik= -34875.854831, dloglik=0.013174 
## K = 6, iter = 6, loglik= -34573.520344, dloglik=0.008669 
## K = 6, iter = 7, loglik= -34346.587951, dloglik=0.006564 
## K = 6, iter = 8, loglik= -34159.814691, dloglik=0.005438 
## K = 6, iter = 9, loglik= -33996.841377, dloglik=0.004771 
## K = 6, iter = 10, loglik= -33848.331818, dloglik=0.004368 
## K = 6, iter = 11, loglik= -33710.517314, dloglik=0.004072 
## K = 6, iter = 12, loglik= -33582.502133, dloglik=0.003797 
## K = 6, iter = 13, loglik= -33465.382937, dloglik=0.003488 
## K = 6, iter = 14, loglik= -33361.001693, dloglik=0.003119 
## K = 6, iter = 15, loglik= -33268.844201, dloglik=0.002762 
## K = 6, iter = 16, loglik= -33186.458205, dloglik=0.002476 
## K = 6, iter = 17, loglik= -33111.504371, dloglik=0.002259 
## K = 6, iter = 18, loglik= -33042.335693, dloglik=0.002089 
## K = 6, iter = 19, loglik= -32977.845549, dloglik=0.001952 
## K = 6, iter = 20, loglik= -32917.269207, dloglik=0.001837 
## K = 6, iter = 21, loglik= -32860.061645, dloglik=0.001738 
## K = 6, iter = 22, loglik= -32805.830764, dloglik=0.001650 
## K = 6, iter = 23, loglik= -32754.294333, dloglik=0.001571 
## K = 6, iter = 24, loglik= -32705.245790, dloglik=0.001497 
## K = 6, iter = 25, loglik= -32658.526356, dloglik=0.001428</code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span></code></pre></div>
<pre><code><span class="co">## class: SRTProject </span>
<span class="co">## outputPath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\Liver8 </span>
<span class="co">## h5filePath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\Liver8/Liver8.h5 </span>
<span class="co">## ---------Datasets basic information-----------------</span>
<span class="co">## samples(8): Liver1 Liver2 ... Liver7 Liver8</span>
<span class="co">## sampleColData names(3): species tissues NumOfSpots</span>
<span class="co">## cellMetaData names(13): orig.ident nCount_RNA ... sample_id batch</span>
<span class="co">## numberOfSpots(8): 675 684 ... 629 590</span>
<span class="co">## ---------Downstream analyses information-----------------</span>
<span class="co">## Variable features:  2000 </span>
<span class="co">## Low-dimensional embeddings(3): PCA microEnv.iSC.MEB aligned.iSC.MEB</span>
<span class="co">## Inferred cluster labels: Yes</span>
<span class="co">## Embedding for plotting(0):</span></code></pre>
</div>
<div class="section level2">
<h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h2>
<div class="section level3">
<h3 id="srt-embeddings">SRT embeddings<a class="anchor" aria-label="anchor" href="#srt-embeddings"></a>
</h3>
<p>To check the performance of integration, we visualize the inferred clusters and data batches on the two-dimensional tSNEs. First, we use <code><a href="../reference/AddTSNE.html">AddTSNE()</a></code> function to calculate the two-dimensional tSNEs based on the <code>aligned.iSC.MEB</code> embeddings.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddTSNE.html">AddTSNE</a></span><span class="op">(</span><span class="va">SRTProj</span>, n_comp <span class="op">=</span> <span class="fl">2</span>, reduction <span class="op">=</span> <span class="st">"aligned.iSC.MEB"</span><span class="op">)</span></code></pre></div>
<p>We visualized the inferred clusters and data batches. The tSNE plot showed the domain clusters were well segregated and data batches were well mixed.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cols_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chooseColors.html">chooseColors</a></span><span class="op">(</span>n_colors <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>
<span class="va">cols_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chooseColors.html">chooseColors</a></span><span class="op">(</span>palettes_name <span class="op">=</span> <span class="st">"Light 13"</span>, n_colors <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
<span class="va">p_tsne2_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EmbedPlot.html">EmbedPlot</a></span><span class="op">(</span><span class="va">SRTProj</span>, item <span class="op">=</span> <span class="st">"cluster"</span>, plotEmbeddings <span class="op">=</span> <span class="st">"tSNE"</span>, cols <span class="op">=</span> <span class="va">cols_cluster</span>,
    legend.position <span class="op">=</span> <span class="st">"right"</span>, pt_size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>
<span class="va">p_tsne2_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EmbedPlot.html">EmbedPlot</a></span><span class="op">(</span><span class="va">SRTProj</span>, item <span class="op">=</span> <span class="st">"batch"</span>, plotEmbeddings <span class="op">=</span> <span class="st">"tSNE"</span>, cols <span class="op">=</span> <span class="va">cols_batch</span>,
    legend.position <span class="op">=</span> <span class="st">"right"</span>, pt_size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>
<span class="fu"><a href="../reference/drawFigs.html">drawFigs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p_tsne2_cluster</span>, <span class="va">p_tsne2_batch</span><span class="op">)</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></code></pre></div>
<p><img src="multibatch_basic_tutorial_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="spatial-heatmap-to-show-the-clustering-performance">Spatial heatmap to show the clustering performance<a class="anchor" aria-label="anchor" href="#spatial-heatmap-to-show-the-clustering-performance"></a>
</h3>
<p>Except for the embedding plots, SRTpipeline also provides a variaty of visualization functions. First, we visualize the spatial distribution of cluster labels that shows the tissue structure for all data batches.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p12</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EachClusterSpaHeatMap.html">EachClusterSpaHeatMap</a></span><span class="op">(</span><span class="va">SRTProj</span>, cols <span class="op">=</span> <span class="va">cols_cluster</span>, legend.position <span class="op">=</span> <span class="st">"bottom"</span>, base_size <span class="op">=</span> <span class="fl">12</span>,
    pt_size <span class="op">=</span> <span class="fl">2</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span>, nrow.legend <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">p12</span></code></pre></div>
<p><img src="multibatch_basic_tutorial_files/figure-html/unnamed-chunk-11-1.png" width="700"> By setting <code>combine =FALSE</code>, this function will return a list of ggplot2 objects, thus user can revise each plot. In addtion, we can also plot some of data batches that are interested.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EachClusterSpaHeatMap.html">EachClusterSpaHeatMap</a></span><span class="op">(</span><span class="va">SRTProj</span>, cols <span class="op">=</span> <span class="va">cols_cluster</span>, legend.position <span class="op">=</span> <span class="st">"bottom"</span>, base_size <span class="op">=</span> <span class="fl">12</span>,
    pt_size <span class="op">=</span> <span class="fl">0.5</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span>, nrow.legend <span class="op">=</span> <span class="fl">1</span>, combine <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="../reference/EachClusterSpaHeatMap.html">EachClusterSpaHeatMap</a></span><span class="op">(</span><span class="va">SRTProj</span>, batch <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, title_name <span class="op">=</span> <span class="st">"iSC.MEB: "</span>, cols <span class="op">=</span> <span class="va">cols_cluster</span>, legend.position <span class="op">=</span> <span class="st">"right"</span>,
    base_size <span class="op">=</span> <span class="fl">12</span>, pt_size <span class="op">=</span> <span class="fl">2</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="multibatch_basic_tutorial_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="rgb-plot-to-show-the-embedding-performance">RGB plot to show the embedding performance<a class="anchor" aria-label="anchor" href="#rgb-plot-to-show-the-embedding-performance"></a>
</h3>
<p>Next, we summarized the inferred embeddings for biological eﬀects between spatial domain types (the slot <code>reductions</code>) using three components from either tSNE or UMAP and visualized the resulting tSNE/UMAP components with red/green/blue (RGB) colors in the RGB plot.</p>
<p>The resulting RGB plots from <code>iSC.MEB</code> showed the structure organization of the mouse liver tissue, and <code>iSC.MEB</code> provided fast transitions across portal vein and central vein domains.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddTSNE.html">AddTSNE</a></span><span class="op">(</span><span class="va">SRTProj</span>, n_comp <span class="op">=</span> <span class="fl">3</span>, reduction <span class="op">=</span> <span class="st">"aligned.iSC.MEB"</span><span class="op">)</span>
<span class="va">p_tsne3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EachRGBSpaHeatMap.html">EachRGBSpaHeatMap</a></span><span class="op">(</span><span class="va">SRTProj</span>, plot_type <span class="op">=</span> <span class="st">"tSNE"</span>, pt_size <span class="op">=</span> <span class="fl">1</span>, title_name <span class="op">=</span> <span class="st">""</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,
    <span class="fl">4</span><span class="op">)</span><span class="op">)</span>
<span class="va">p_tsne3</span></code></pre></div>
<p><img src="multibatch_basic_tutorial_files/figure-html/unnamed-chunk-13-1.png" width="700"> To run UMAP in SRTpipeline we use the <code><a href="../reference/AddUMAP.html">AddUMAP()</a></code> function. Frist, we evaluate the two-dimensional UMAPs.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddUMAP.html">AddUMAP</a></span><span class="op">(</span><span class="va">SRTProj</span>, n_comp <span class="op">=</span> <span class="fl">2</span>, reduction <span class="op">=</span> <span class="st">"aligned.iSC.MEB"</span><span class="op">)</span>
<span class="va">p_umap2_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EmbedPlot.html">EmbedPlot</a></span><span class="op">(</span><span class="va">SRTProj</span>, item <span class="op">=</span> <span class="st">"cluster"</span>, plotEmbeddings <span class="op">=</span> <span class="st">"UMAP"</span>, cols <span class="op">=</span> <span class="va">cols_cluster</span>,
    legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span>
<span class="va">p_umap2_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EmbedPlot.html">EmbedPlot</a></span><span class="op">(</span><span class="va">SRTProj</span>, item <span class="op">=</span> <span class="st">"batch"</span>, plotEmbeddings <span class="op">=</span> <span class="st">"UMAP"</span>, cols <span class="op">=</span> <span class="va">cols_batch</span>,
    legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span>
<span class="fu"><a href="../reference/drawFigs.html">drawFigs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p_umap2_cluster</span>, <span class="va">p_umap2_batch</span><span class="op">)</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></code></pre></div>
<p><img src="multibatch_basic_tutorial_files/figure-html/unnamed-chunk-14-1.png" width="700"> Then, we evaluate the three-dimensional UMAPs.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddUMAP.html">AddUMAP</a></span><span class="op">(</span><span class="va">SRTProj</span>, n_comp <span class="op">=</span> <span class="fl">3</span>, reduction <span class="op">=</span> <span class="st">"aligned.iSC.MEB"</span><span class="op">)</span>
<span class="va">p_umap3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EachRGBSpaHeatMap.html">EachRGBSpaHeatMap</a></span><span class="op">(</span><span class="va">SRTProj</span>, plot_type <span class="op">=</span> <span class="st">"UMAP"</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span>, pt_size <span class="op">=</span> <span class="fl">1.2</span>, title_name <span class="op">=</span> <span class="st">"UMAP: "</span><span class="op">)</span>
<span class="va">p_umap3</span></code></pre></div>
<p><img src="multibatch_basic_tutorial_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>To save the plot, we can use <code><a href="../reference/write_fig.html">write_fig()</a></code> function.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/write_fig.html">write_fig</a></span><span class="op">(</span><span class="va">p_umap3</span>, filename <span class="op">=</span> <span class="st">"iSCMEB_p_umap3.png"</span>, width <span class="op">=</span> <span class="fl">14</span>, height <span class="op">=</span> <span class="fl">11</span><span class="op">)</span></code></pre></div>
<p>Moreover, we can visualize the microenvironment effects on the spatial coordinates. Then, we evaluate the three-dimensional UMAPs based on <code>microEnv.iSC.MEB</code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># save the previous calculated UMAP3</span>
<span class="va">umap3_cluster</span> <span class="op">&lt;-</span> <span class="va">SRTProj</span><span class="op">@</span><span class="va">plotEmbeddings</span><span class="op">$</span><span class="va">UMAP3</span>
<span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddUMAP.html">AddUMAP</a></span><span class="op">(</span><span class="va">SRTProj</span>, n_comp <span class="op">=</span> <span class="fl">3</span>, reduction <span class="op">=</span> <span class="st">"microEnv.iSC.MEB"</span><span class="op">)</span>
<span class="va">p_umap3_micro</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EachRGBSpaHeatMap.html">EachRGBSpaHeatMap</a></span><span class="op">(</span><span class="va">SRTProj</span>, plot_type <span class="op">=</span> <span class="st">"UMAP"</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span>, pt_size <span class="op">=</span> <span class="fl">1.2</span>,
    title_name <span class="op">=</span> <span class="st">"mircoEnv: "</span><span class="op">)</span>
<span class="va">p_umap3_micro</span></code></pre></div>
<p><img src="multibatch_basic_tutorial_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="cell-by-cell-heatmap-to-show-the-relation-of-spatial-domains">Cell-by-cell heatmap to show the relation of spatial domains<a class="anchor" aria-label="anchor" href="#cell-by-cell-heatmap-to-show-the-relation-of-spatial-domains"></a>
</h3>
<p>We plotted the heatmap of Pearson’s correlation coefcients of the estimated embeddings among the detected domains shows the good separation of the estimated embeddings across domains and the high correlations between domains with similarity.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p_cc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CCHeatMap.html">CCHeatMap</a></span><span class="op">(</span><span class="va">SRTProj</span>, reduction <span class="op">=</span> <span class="st">"aligned.iSC.MEB"</span>, grp_color <span class="op">=</span> <span class="va">cols_cluster</span><span class="op">)</span>
<span class="va">p_cc</span></code></pre></div>
<p><img src="multibatch_basic_tutorial_files/figure-html/unnamed-chunk-18-1.png" width="700"> After adding the quantities for data visualization, the <code>SRTProject</code> object will have more information in the downstream analyses information. Now, we print this <code>SRTProject</code> object to check it. We observed two components added in the slot <code>plotEmbeddings</code> (Embeddings for plotting): <code>tSNE</code>, <code>tSNE3</code>, <code>UMAP</code> and <code>UMAP3</code>.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span></code></pre></div>
<pre><code><span class="co">## class: SRTProject </span>
<span class="co">## outputPath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\Liver8 </span>
<span class="co">## h5filePath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\Liver8/Liver8.h5 </span>
<span class="co">## ---------Datasets basic information-----------------</span>
<span class="co">## samples(8): Liver1 Liver2 ... Liver7 Liver8</span>
<span class="co">## sampleColData names(3): species tissues NumOfSpots</span>
<span class="co">## cellMetaData names(13): orig.ident nCount_RNA ... sample_id batch</span>
<span class="co">## numberOfSpots(8): 675 684 ... 629 590</span>
<span class="co">## ---------Downstream analyses information-----------------</span>
<span class="co">## Variable features:  2000 </span>
<span class="co">## Low-dimensional embeddings(3): PCA microEnv.iSC.MEB aligned.iSC.MEB</span>
<span class="co">## Inferred cluster labels: Yes</span>
<span class="co">## Embedding for plotting(4): tSNE tSNE3 UMAP UMAP3</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="combined-deg-analysis">Combined DEG analysis<a class="anchor" aria-label="anchor" href="#combined-deg-analysis"></a>
</h2>
<p>After fitting the integration model <code>iSC.MEB</code>, we could obatin the batch corrected log-normalized gene expression data by using function <code><a href="../reference/getIntegratedData.html">getIntegratedData()</a></code>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">speInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getIntegratedData.html">getIntegratedData</a></span><span class="op">(</span><span class="va">SRTProj</span>, Method <span class="op">=</span> <span class="st">"iSC.MEB"</span>, species <span class="op">=</span> <span class="st">"Mouse"</span><span class="op">)</span></code></pre></div>
<p>After obtain the spatial cluster labels using a clustering model (i.e., <code>iSC.MEB</code>), we can perform differentially expression analysis. First, we can detect the DE genes using function <code>FindDEGs</code> for only one cluster, i.e., cluster 3.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dat_deg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindDEGs.html">FindDEGs</a></span><span class="op">(</span><span class="va">speInt</span>, cluster1 <span class="op">=</span> <span class="st">"3"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">dat_deg</span>, <span class="va">avg_log2FC</span> <span class="op">&gt;</span> <span class="fl">0.1</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 1 row and 5 columns</span>
<span class="co">##            p_val avg_log2FC     pct.1     pct.2   p_val_adj</span>
<span class="co">##        &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;</span>
<span class="co">## A1bg 3.04919e-42   0.252504     0.979     0.877 6.09838e-39</span></code></pre>
<p>We perform differential expression analysis for all clusters by using <code><a href="../reference/FindAllDEGs.html">FindAllDEGs()</a></code> function, then the DE genes’ information is saved in a data.frame object <code>dat_degs</code>.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dat_degs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindAllDEGs.html">FindAllDEGs</a></span><span class="op">(</span><span class="va">speInt</span><span class="op">)</span>
<span class="va">dat_degs</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 137 rows and 7 columns</span>
<span class="co">##                p_val avg_log2FC     pct.1     pct.2    p_val_adj  cluster</span>
<span class="co">##            &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;    &lt;numeric&gt; &lt;factor&gt;</span>
<span class="co">## Cyp2f2   1.30039e-98  -0.250081     0.987     0.973  2.60078e-95        1</span>
<span class="co">## Cyp3a44  1.63897e-63   0.314311     0.928     0.779  3.27793e-60        1</span>
<span class="co">## Sds      1.77854e-48  -0.355872     0.802     0.827  3.55708e-45        1</span>
<span class="co">## Aldh1b1  5.89245e-22  -0.264130     0.672     0.727  1.17849e-18        1</span>
<span class="co">## Cyp2e1  1.15125e-274  -0.423678     0.993     0.998 2.30251e-271        2</span>
<span class="co">## ...              ...        ...       ...       ...          ...      ...</span>
<span class="co">## Rnf13    1.20494e-16  -0.250187     0.820     0.943  2.40988e-13        5</span>
<span class="co">## Dpt      1.41514e-14   0.295305     0.639     0.545  2.83028e-11        5</span>
<span class="co">## Vim      9.67010e-09   0.326605     0.671     0.647  1.93402e-05        5</span>
<span class="co">## Col1a1   6.84510e-06   0.302746     0.599     0.572  1.36902e-02        5</span>
<span class="co">## Col1a2   7.85131e-03   0.327749     0.601     0.612  1.00000e+00        5</span>
<span class="co">##                gene</span>
<span class="co">##         &lt;character&gt;</span>
<span class="co">## Cyp2f2       Cyp2f2</span>
<span class="co">## Cyp3a44     Cyp3a44</span>
<span class="co">## Sds             Sds</span>
<span class="co">## Aldh1b1     Aldh1b1</span>
<span class="co">## Cyp2e1       Cyp2e1</span>
<span class="co">## ...             ...</span>
<span class="co">## Rnf13         Rnf13</span>
<span class="co">## Dpt             Dpt</span>
<span class="co">## Vim             Vim</span>
<span class="co">## Col1a1       Col1a1</span>
<span class="co">## Col1a2       Col1a2</span></code></pre>
<p>We identify the significant DE genes by two criteria: (a) adjustd p-value less than 0.01 and (b) average log fold change greater than 0.15.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">degs_sig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">dat_degs</span>, <span class="va">p_val_adj</span> <span class="op">&lt;</span> <span class="fl">0.01</span> <span class="op">&amp;</span> <span class="va">avg_log2FC</span> <span class="op">&gt;</span> <span class="fl">0.15</span><span class="op">)</span>
<span class="va">degs_sig</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 69 rows and 7 columns</span>
<span class="co">##                 p_val avg_log2FC     pct.1     pct.2    p_val_adj  cluster</span>
<span class="co">##             &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;    &lt;numeric&gt; &lt;factor&gt;</span>
<span class="co">## Cyp3a44   1.63897e-63   0.314311     0.928     0.779  3.27793e-60        1</span>
<span class="co">## Cyp2f2.1 3.97931e-223   0.387100     1.000     0.972 7.95862e-220        2</span>
<span class="co">## Sds.1    1.08390e-213   0.659946     0.984     0.791 2.16781e-210        2</span>
<span class="co">## Hal      4.06491e-196   0.427699     0.999     0.933 8.12982e-193        2</span>
<span class="co">## Ctsc     2.60875e-186   0.508951     0.993     0.848 5.21749e-183        2</span>
<span class="co">## ...               ...        ...       ...       ...          ...      ...</span>
<span class="co">## Slc13a3   9.57091e-26   0.291170     0.659     0.559  1.91418e-22        4</span>
<span class="co">## Gsn       2.01150e-36   0.567004     0.748     0.611  4.02300e-33        5</span>
<span class="co">## Col3a1    2.37755e-27   0.491134     0.771     0.699  4.75510e-24        5</span>
<span class="co">## Dpt       1.41514e-14   0.295305     0.639     0.545  2.83028e-11        5</span>
<span class="co">## Vim       9.67010e-09   0.326605     0.671     0.647  1.93402e-05        5</span>
<span class="co">##                 gene</span>
<span class="co">##          &lt;character&gt;</span>
<span class="co">## Cyp3a44      Cyp3a44</span>
<span class="co">## Cyp2f2.1      Cyp2f2</span>
<span class="co">## Sds.1            Sds</span>
<span class="co">## Hal              Hal</span>
<span class="co">## Ctsc            Ctsc</span>
<span class="co">## ...              ...</span>
<span class="co">## Slc13a3      Slc13a3</span>
<span class="co">## Gsn              Gsn</span>
<span class="co">## Col3a1        Col3a1</span>
<span class="co">## Dpt              Dpt</span>
<span class="co">## Vim              Vim</span></code></pre>
<p>In the following, we perform gene set enrichment analysis for the DE genes of each spatial Domain identified by <code>iSC.MEB</code> model using R package <code>gprofiler2</code>.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">gprofiler2</span><span class="op">)</span>
<span class="va">termList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span> <span class="op">{</span>
    <span class="co"># k &lt;- 1</span>
    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">degs_sig</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="va">k</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">{</span>
        <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"k = "</span>, <span class="va">k</span>, <span class="st">"\n"</span><span class="op">)</span>
        <span class="va">dat_degs_sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">degs_sig</span>, <span class="va">cluster</span> <span class="op">==</span> <span class="va">k</span><span class="op">)</span>

        <span class="va">que1</span> <span class="op">&lt;-</span> <span class="va">dat_degs_sub</span><span class="op">$</span><span class="va">gene</span>
        <span class="va">gostres</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gprofiler2/man/gost.html" class="external-link">gost</a></span><span class="op">(</span>query <span class="op">=</span> <span class="va">que1</span>, organism <span class="op">=</span> <span class="st">"hsapiens"</span>, correction_method <span class="op">=</span> <span class="st">"fdr"</span><span class="op">)</span>
        <span class="va">termList</span><span class="op">[[</span><span class="va">k</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">gostres</span>
    <span class="op">}</span>

<span class="op">}</span></code></pre></div>
<pre><code><span class="co">## k =  1 </span>
<span class="co">## k =  2 </span>
<span class="co">## k =  3 </span>
<span class="co">## k =  4 </span>
<span class="co">## k =  5</span></code></pre>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">termList</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">result</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## NULL</span></code></pre>
<p>To understand the functions of the identified spatial domains by <code>iSC.MEB</code> model, we compare the top significant biological process (BP) pathways in GO database for the DE genes from Domain 1 and 4. Here, we only show the significant BP pathways and users can explore the other databases such as KEGG and HPA.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Most commonly used databases</span>
<span class="va">source_set</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GO:BP"</span>, <span class="st">"GO:CC"</span>, <span class="st">"GO:MF"</span>, <span class="st">"KEGG"</span>, <span class="st">"HPA"</span><span class="op">)</span>
<span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"steelblue3"</span>, <span class="st">"goldenrod"</span>, <span class="st">"brown3"</span>, <span class="st">"#f98866"</span>, <span class="st">"#CE6DBD"</span><span class="op">)</span>
<span class="co">## Here, we show GO:BP</span>
<span class="va">source1</span> <span class="op">&lt;-</span> <span class="st">"GO:BP"</span>
<span class="va">ss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">source_set</span> <span class="op">==</span> <span class="va">source1</span><span class="op">)</span>
<span class="va">ntop</span> <span class="op">=</span> <span class="fl">5</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cols</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">source_set</span>
<span class="va">pList_enrich</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">ii</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span> <span class="op">{</span>
    <span class="co">## ii &lt;- 3</span>
    <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"ii="</span>, <span class="va">ii</span><span class="op">)</span>
    <span class="va">gostres2</span> <span class="op">&lt;-</span> <span class="va">termList</span><span class="op">[[</span><span class="va">ii</span><span class="op">]</span><span class="op">]</span>
    <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">gostres2</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
        <span class="va">dat1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">gostres2</span><span class="op">$</span><span class="va">result</span>, <span class="va">term_size</span> <span class="op">&lt;</span> <span class="fl">500</span><span class="op">)</span>
        <span class="va">dat1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_top_pathway.html">get_top_pathway</a></span><span class="op">(</span><span class="va">dat1</span>, ntop <span class="op">=</span> <span class="va">ntop</span>, source_set <span class="op">=</span> <span class="va">source1</span><span class="op">)</span>
        <span class="va">dat1</span> <span class="op">&lt;-</span> <span class="va">dat1</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/stats/complete.cases.html" class="external-link">complete.cases</a></span><span class="op">(</span><span class="va">dat1</span><span class="op">)</span>, <span class="op">]</span>
        <span class="va">dat1</span><span class="op">$</span><span class="va">nlog10P</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">dat1</span><span class="op">$</span><span class="va">p_value</span><span class="op">)</span>

        <span class="va">pList_enrich</span><span class="op">[[</span><span class="va">ii</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/barPlot_enrich.html">barPlot_enrich</a></span><span class="op">(</span><span class="va">dat1</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">dat1</span><span class="op">$</span><span class="va">nlog10P</span><span class="op">)</span>, <span class="op">]</span>, source <span class="op">=</span> <span class="st">"source"</span>, <span class="st">"term_name"</span>,
            <span class="st">"nlog10P"</span>, cols <span class="op">=</span> <span class="va">cols</span><span class="op">[</span><span class="va">source_set</span><span class="op">[</span><span class="va">ss</span><span class="op">]</span><span class="op">]</span>, base_size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ylab</span><span class="op">(</span><span class="st">"-log10(p-adj)"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">xlab</span><span class="op">(</span><span class="st">"Biological terms"</span><span class="op">)</span> <span class="op">+</span>
            <span class="fu">ggtitle</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Domain"</span>, <span class="va">ii</span><span class="op">)</span><span class="op">)</span>
    <span class="op">}</span>

<span class="op">}</span>
<span class="fu"><a href="../reference/drawFigs.html">drawFigs</a></span><span class="op">(</span><span class="va">pList_enrich</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">4</span><span class="op">)</span><span class="op">]</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span>, common.legend <span class="op">=</span> <span class="cn">T</span>, align <span class="op">=</span> <span class="st">"hv"</span><span class="op">)</span></code></pre></div>
<p><img src="multibatch_basic_tutorial_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<p>We take out the top DE genes for each cluster for visualization.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>
<span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">5</span>
<span class="va">dat_degs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="va">as.data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html" class="external-link">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n</span>, wt <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">topGene</span>
<span class="va">topGene</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 22 x 7</span></span>
<span class="co">## <span style="color: #949494;"># Groups:   cluster [5]</span></span>
<span class="co">##        p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene   </span>
<span class="co">##        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> 1.30<span style="color: #949494;">e</span><span style="color: #BB0000;">- 98</span>     -<span style="color: #BB0000;">0.250</span> 0.987 0.973 2.60<span style="color: #949494;">e</span><span style="color: #BB0000;">- 95</span> 1       Cyp2f2 </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> 1.64<span style="color: #949494;">e</span><span style="color: #BB0000;">- 63</span>      0.314 0.928 0.779 3.28<span style="color: #949494;">e</span><span style="color: #BB0000;">- 60</span> 1       Cyp3a44</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> 1.78<span style="color: #949494;">e</span><span style="color: #BB0000;">- 48</span>     -<span style="color: #BB0000;">0.356</span> 0.802 0.827 3.56<span style="color: #949494;">e</span><span style="color: #BB0000;">- 45</span> 1       Sds    </span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> 5.89<span style="color: #949494;">e</span><span style="color: #BB0000;">- 22</span>     -<span style="color: #BB0000;">0.264</span> 0.672 0.727 1.18<span style="color: #949494;">e</span><span style="color: #BB0000;">- 18</span> 1       Aldh1b1</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> 1.08<span style="color: #949494;">e</span><span style="color: #BB0000;">-213</span>      0.660 0.984 0.791 2.17<span style="color: #949494;">e</span><span style="color: #BB0000;">-210</span> 2       Sds    </span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> 2.61<span style="color: #949494;">e</span><span style="color: #BB0000;">-186</span>      0.509 0.993 0.848 5.22<span style="color: #949494;">e</span><span style="color: #BB0000;">-183</span> 2       Ctsc   </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> 1.24<span style="color: #949494;">e</span><span style="color: #BB0000;">-166</span>      0.629 0.925 0.677 2.48<span style="color: #949494;">e</span><span style="color: #BB0000;">-163</span> 2       Aldh1b1</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> 2.60<span style="color: #949494;">e</span><span style="color: #BB0000;">-107</span>      0.475 0.919 0.728 5.21<span style="color: #949494;">e</span><span style="color: #BB0000;">-104</span> 2       Etnppl </span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> 1.27<span style="color: #949494;">e</span><span style="color: #BB0000;">- 72</span>      0.524 0.8   0.599 2.54<span style="color: #949494;">e</span><span style="color: #BB0000;">- 69</span> 2       Spp1   </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> 3.05<span style="color: #949494;">e</span><span style="color: #BB0000;">- 42</span>      0.253 0.979 0.877 6.10<span style="color: #949494;">e</span><span style="color: #BB0000;">- 39</span> 3       A1bg   </span>
<span class="co">## <span style="color: #949494;"># ... with 12 more rows</span></span></code></pre>
<p>We visualize the DE genes for each cluster group by gene-by-cell heatmap using the <code><a href="../reference/GCHeatMap.html">GCHeatMap()</a></code> function.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GCHeatMap.html">GCHeatMap</a></span><span class="op">(</span><span class="va">speInt</span>, features <span class="op">=</span> <span class="va">topGene</span><span class="op">$</span><span class="va">gene</span>, grp_color <span class="op">=</span> <span class="va">cols_cluster</span>, y_text_size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>
<span class="va">p1</span></code></pre></div>
<p><img src="multibatch_basic_tutorial_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="trajectory-inference">Trajectory inference<a class="anchor" aria-label="anchor" href="#trajectory-inference"></a>
</h2>
<p>Next, we performed trajectory inference using the aligned embeddings and domain labels estimated by iSC.MEB model.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">speInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddTrajectory.html">AddTrajectory</a></span><span class="op">(</span><span class="va">speInt</span>, reduction <span class="op">=</span> <span class="st">"aligned.iSC.MEB"</span><span class="op">)</span>
<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EmbedPlot.html">EmbedPlot</a></span><span class="op">(</span><span class="va">speInt</span>, reduction <span class="op">=</span> <span class="st">"aligned.iSC.MEB"</span>, colour_by <span class="op">=</span> <span class="st">"PT"</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EmbedPlot.html">EmbedPlot</a></span><span class="op">(</span><span class="va">speInt</span>, reduction <span class="op">=</span> <span class="st">"tSNE"</span>, colour_by <span class="op">=</span> <span class="st">"PT"</span><span class="op">)</span>
<span class="fu"><a href="../reference/drawFigs.html">drawFigs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span><span class="op">)</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, common.legend <span class="op">=</span> <span class="cn">TRUE</span>, legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></code></pre></div>
<p><img src="multibatch_basic_tutorial_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
<p>Visualize the inferred pseudotime on the spatial coordinates for each data batch.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p_spa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EachEmbedPlot.html">EachEmbedPlot</a></span><span class="op">(</span><span class="va">speInt</span>, reduction <span class="op">=</span> <span class="st">"Coord"</span>, colour_by <span class="op">=</span> <span class="st">"PT"</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span>
<span class="va">p_spa</span></code></pre></div>
<p><img src="multibatch_basic_tutorial_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># save(SRTProj, file=paste0(SRTProj@projectMetadata$outputPath,'/SRTProj.rds'))</span>
<span class="co"># load('F:/Research</span>
<span class="co"># paper/IntegrateDRcluster/AnalysisCode/SRTpipeline/vignettes/Liver8/SRTProj.rds')</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="other-downstream-analyses">Other downstream analyses<a class="anchor" aria-label="anchor" href="#other-downstream-analyses"></a>
</h2>
<details><p><summary><strong>Session Info</strong></summary></p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.1.2 (2021-11-01)</span>
<span class="co">## Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">## Running under: Windows 10 x64 (build 22621)</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">## [1] LC_COLLATE=Chinese (Simplified)_China.936 </span>
<span class="co">## [2] LC_CTYPE=Chinese (Simplified)_China.936   </span>
<span class="co">## [3] LC_MONETARY=Chinese (Simplified)_China.936</span>
<span class="co">## [4] LC_NUMERIC=C                              </span>
<span class="co">## [5] LC_TIME=Chinese (Simplified)_China.936    </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">## [8] methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] scater_1.25.1               scuttle_1.4.0              </span>
<span class="co">##  [3] slingshot_2.2.0             TrajectoryUtils_1.2.0      </span>
<span class="co">##  [5] princurve_2.1.6             gprofiler2_0.2.1           </span>
<span class="co">##  [7] bigalgebra_1.1.0            bigmemory_4.5.36           </span>
<span class="co">##  [9] SpatialExperiment_1.4.0     SingleCellExperiment_1.16.0</span>
<span class="co">## [11] SummarizedExperiment_1.24.0 Biobase_2.54.0             </span>
<span class="co">## [13] GenomicRanges_1.46.1        GenomeInfoDb_1.30.1        </span>
<span class="co">## [15] IRanges_2.28.0              MatrixGenerics_1.6.0       </span>
<span class="co">## [17] matrixStats_0.62.0          dplyr_1.0.9                </span>
<span class="co">## [19] colorspace_2.0-3            iSC.MEB_1.0.1              </span>
<span class="co">## [21] ggplot2_3.3.6               gtools_3.9.2.2             </span>
<span class="co">## [23] irlba_2.3.5                 DR.SC_3.1                  </span>
<span class="co">## [25] spatstat.geom_2.4-0         spatstat.data_3.0-0        </span>
<span class="co">## [27] Matrix_1.4-0                hdf5r_1.3.5                </span>
<span class="co">## [29] ff_4.0.7                    bit_4.0.4                  </span>
<span class="co">## [31] S4Vectors_0.32.3            BiocGenerics_0.40.0        </span>
<span class="co">## [33] rhdf5_2.38.0                sp_1.5-0                   </span>
<span class="co">## [35] SeuratObject_4.1.0          Seurat_4.1.1               </span>
<span class="co">## [37] SRTpipeline_0.1.1          </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] scattermore_0.8           ggthemes_4.2.4           </span>
<span class="co">##   [3] R.methodsS3_1.8.1         GiRaF_1.0.1              </span>
<span class="co">##   [5] ragg_1.2.2                tidyr_1.2.0              </span>
<span class="co">##   [7] bit64_4.0.5               knitr_1.37               </span>
<span class="co">##   [9] DelayedArray_0.20.0       R.utils_2.11.0           </span>
<span class="co">##  [11] data.table_1.14.2         rpart_4.1.16             </span>
<span class="co">##  [13] RCurl_1.98-1.6            generics_0.1.2           </span>
<span class="co">##  [15] ScaledMatrix_1.2.0        cowplot_1.1.1            </span>
<span class="co">##  [17] RANN_2.6.1                future_1.26.1            </span>
<span class="co">##  [19] httpuv_1.6.5              assertthat_0.2.1         </span>
<span class="co">##  [21] viridis_0.6.2             xfun_0.29                </span>
<span class="co">##  [23] jquerylib_0.1.4           evaluate_0.15            </span>
<span class="co">##  [25] promises_1.2.0.1          fansi_1.0.3              </span>
<span class="co">##  [27] igraph_1.3.5              DBI_1.1.2                </span>
<span class="co">##  [29] htmlwidgets_1.5.4         purrr_0.3.4              </span>
<span class="co">##  [31] ellipsis_0.3.2            RSpectra_0.16-1          </span>
<span class="co">##  [33] ggpubr_0.4.0              backports_1.4.1          </span>
<span class="co">##  [35] deldir_1.0-6              sparseMatrixStats_1.6.0  </span>
<span class="co">##  [37] vctrs_0.4.1               ROCR_1.0-11              </span>
<span class="co">##  [39] abind_1.4-5               cachem_1.0.6             </span>
<span class="co">##  [41] withr_2.5.0               PRECAST_1.4              </span>
<span class="co">##  [43] progressr_0.10.1          sctransform_0.3.3        </span>
<span class="co">##  [45] mclust_5.4.10             goftest_1.2-3            </span>
<span class="co">##  [47] cluster_2.1.2             lazyeval_0.2.2           </span>
<span class="co">##  [49] crayon_1.5.1              edgeR_3.36.0             </span>
<span class="co">##  [51] pkgconfig_2.0.3           labeling_0.4.2           </span>
<span class="co">##  [53] nlme_3.1-155              vipor_0.4.5              </span>
<span class="co">##  [55] rlang_1.0.2               globals_0.15.0           </span>
<span class="co">##  [57] lifecycle_1.0.1           miniUI_0.1.1.1           </span>
<span class="co">##  [59] bigmemory.sri_0.1.3       rsvd_1.0.5               </span>
<span class="co">##  [61] rprojroot_2.0.3           polyclip_1.10-0          </span>
<span class="co">##  [63] lmtest_0.9-40             carData_3.0-5            </span>
<span class="co">##  [65] Rhdf5lib_1.16.0           zoo_1.8-10               </span>
<span class="co">##  [67] beeswarm_0.4.0            ggridges_0.5.3           </span>
<span class="co">##  [69] rjson_0.2.21              png_0.1-7                </span>
<span class="co">##  [71] viridisLite_0.4.0         bitops_1.0-7             </span>
<span class="co">##  [73] R.oo_1.24.0               KernSmooth_2.23-20       </span>
<span class="co">##  [75] rhdf5filters_1.6.0        DelayedMatrixStats_1.16.0</span>
<span class="co">##  [77] stringr_1.4.0             parallelly_1.32.0        </span>
<span class="co">##  [79] spatstat.random_2.2-0     rstatix_0.7.0            </span>
<span class="co">##  [81] ggsignif_0.6.3            beachmat_2.10.0          </span>
<span class="co">##  [83] scales_1.2.0              memoise_2.0.1            </span>
<span class="co">##  [85] magrittr_2.0.3            plyr_1.8.7               </span>
<span class="co">##  [87] ica_1.0-2                 zlibbioc_1.40.0          </span>
<span class="co">##  [89] compiler_4.1.2            dqrng_0.3.0              </span>
<span class="co">##  [91] RColorBrewer_1.1-3        fitdistrplus_1.1-8       </span>
<span class="co">##  [93] cli_3.2.0                 XVector_0.34.0           </span>
<span class="co">##  [95] listenv_0.8.0             patchwork_1.1.1          </span>
<span class="co">##  [97] pbapply_1.5-0             formatR_1.11             </span>
<span class="co">##  [99] MASS_7.3-55               mgcv_1.8-39              </span>
<span class="co">## [101] tidyselect_1.1.2          stringi_1.7.6            </span>
<span class="co">## [103] textshaping_0.3.6         highr_0.9                </span>
<span class="co">## [105] yaml_2.3.6                BiocSingular_1.10.0      </span>
<span class="co">## [107] locfit_1.5-9.4            ggrepel_0.9.1            </span>
<span class="co">## [109] grid_4.1.2                sass_0.4.1               </span>
<span class="co">## [111] tools_4.1.2               future.apply_1.9.0       </span>
<span class="co">## [113] rstudioapi_0.13           gridExtra_2.3            </span>
<span class="co">## [115] farver_2.1.0              Rtsne_0.16               </span>
<span class="co">## [117] DropletUtils_1.14.2       digest_0.6.29            </span>
<span class="co">## [119] rgeos_0.5-9               shiny_1.7.1              </span>
<span class="co">## [121] Rcpp_1.0.10               car_3.0-12               </span>
<span class="co">## [123] broom_0.7.12              later_1.3.0              </span>
<span class="co">## [125] RcppAnnoy_0.0.19          httr_1.4.3               </span>
<span class="co">## [127] fs_1.5.2                  tensor_1.5               </span>
<span class="co">## [129] reticulate_1.25           splines_4.1.2            </span>
<span class="co">## [131] uwot_0.1.11               spatstat.utils_3.0-1     </span>
<span class="co">## [133] pkgdown_2.0.6             plotly_4.10.0            </span>
<span class="co">## [135] systemfonts_1.0.4         xtable_1.8-4             </span>
<span class="co">## [137] jsonlite_1.8.0            R6_2.5.1                 </span>
<span class="co">## [139] pillar_1.7.0              htmltools_0.5.2          </span>
<span class="co">## [141] mime_0.12                 glue_1.6.2               </span>
<span class="co">## [143] fastmap_1.1.0             BiocParallel_1.28.3      </span>
<span class="co">## [145] BiocNeighbors_1.12.0      codetools_0.2-18         </span>
<span class="co">## [147] utf8_1.2.2                lattice_0.20-45          </span>
<span class="co">## [149] bslib_0.3.1               spatstat.sparse_2.1-1    </span>
<span class="co">## [151] tibble_3.1.7              ggbeeswarm_0.6.0         </span>
<span class="co">## [153] leiden_0.4.2              magick_2.7.3             </span>
<span class="co">## [155] survival_3.2-13           limma_3.50.1             </span>
<span class="co">## [157] CompQuadForm_1.4.3        rmarkdown_2.11           </span>
<span class="co">## [159] desc_1.4.0                munsell_0.5.0            </span>
<span class="co">## [161] GenomeInfoDbData_1.2.7    HDF5Array_1.22.1         </span>
<span class="co">## [163] reshape2_1.4.4            gtable_0.3.0             </span>
<span class="co">## [165] spatstat.core_2.4-4</span></code></pre>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Wei Liu, Yi Yang, Xu Liao, Xiao Zhang, Jin Liu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
