<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A basic overview of single data analysis • SRTpipeline</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="A basic overview of single data analysis">
<meta property="og:description" content="SRTpipeline">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SRTpipeline</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../articles/Installation.html">
    <span class="fas fa-book"></span>
     
    Install
  </a>
</li>
<li>
  <a href="../articles/get_started.html">
    <span class="fas fa-book"></span>
     
    Get Started
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fas fa-book"></span>
     
    Single Data Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/single_basic_tutorial.html">Basic tutorial</a>
    </li>
    <li>
      <a href="../articles/dlpfc_tutorial.html">Embedding and clustering introduction</a>
    </li>
    <li>
      <a href="../articles/single_Embryo_SC_tutorial.html">SRT data clustering</a>
    </li>
    <li>
      <a href="../articles/single_Embryo_DRSC_tutorial.html">SRT data embedding and clustering</a>
    </li>
    <li>
      <a href="../articles/single_MOB_Anno_tutorial.html">SRT data embedding and annotation</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fas fa-book"></span>
     
    Multiple Data Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/multibatch_basic_tutorial.html">Basic tutorial</a>
    </li>
    <li>
      <a href="../articles/integration_introduction.html">Basic tutorial v2</a>
    </li>
    <li>
      <a href="../articles/multibatch_brain12_tutorial.html">SRT data clustering and integration</a>
    </li>
    <li>
      <a href="../articles/multibatch_HCC4_tutorial.html">SRT data embedding, clustering and integration example 1</a>
    </li>
    <li>
      <a href="../articles/multibatch_liver8_tutorial.html">SRT data embedding, clustering and integration example 2</a>
    </li>
    <li>
      <a href="../articles/multibatch_MOB16_tutorial.html">SRT data embedding, clustering and integration example 3</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/faq.html">
    <span class="fa fa-question-circle-o"></span>
     
    FAQ
  </a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>A basic overview of single data analysis</h1>
            
            <h4 data-toc-skip class="date">Compiled: 二月 13, 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/feiyoung/SRTpipeline/blob/HEAD/vignettes/single_basic_tutorial.Rmd"><code>vignettes/single_basic_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>single_basic_tutorial.Rmd</code></div>

    </div>

    
    
<hr>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This tutorial demonstrates how to use <code>SRTpipeline</code> (&gt;=0.1.0) to analyze a spatially-resolved transcriptomics (SRT) data. This tutorial will cover the following tasks, which we believe will be common for many SRT data analyses:</p>
<ul>
<li>Normalization</li>
<li>Feature selection, including detecting spatially-variable features</li>
<li>Dimensional reduction</li>
<li>Clustering</li>
<li>Interactive visualization</li>
<li>Integration with single-cell RNA-seq data</li>
<li>Working with multiple slices</li>
</ul>
</div>
<div class="section level2">
<h2 id="prepare-srtproject">Prepare SRTProject<a class="anchor" aria-label="anchor" href="#prepare-srtproject"></a>
</h2>
<p>For this tutorial, we will introduce how to create a <code>SRTProject</code> object with single SRT sample using SRTpipeline that includes an introduction to common analytical workflows for a single data batch. Here, we will be taking a spatial transcriptomics dataset (SampleID: 151672) for human dorsolateral prefrontal cortex (DLPFC) as an example. There are 4015 spots and 33538 genes that were sequenced on the 10x Visium platform. Our preprocessed data can be downloaded <a href="https://github.com/feiyoung/PRECAST/blob/main/vignettes_data/dlpfc_151672.rda?raw=true" class="external-link">here</a>, and the raw data can be found <a href="https://github.com/LieberInstitute/spatialLIBD" class="external-link">here</a>.</p>
<p>First, we loaded the required packages with the command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feiyoung/SRTpipeline">SRTpipeline</a></span><span class="op">)</span></code></pre></div>
<p>Next, we download the data to the current working path for the followed analysis by the following command:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">githubURL</span> <span class="op">&lt;-</span> <span class="st">"https://github.com/feiyoung/PRECAST/blob/main/vignettes_data/dlpfc_151672.rda?raw=true"</span>
<span class="co"># download.file(githubURL,'dlpfc_151672.rda',mode='wb')</span></code></pre></div>
<p>Then load to R. The data is saved in a Seurat object named <code>dlpfc_151672</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"dlpfc_151672.rda"</span><span class="op">)</span></code></pre></div>
<details><p><summary><strong>How to read the raw sptial transcriptomics data?</strong></summary> In addition, if users’ data are in the raw form from spaceranger. We start by reading in the data. The <code>read10XVisium()</code> function reads in the output of the <a href="https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/what-is-space-ranger" class="external-link">spaceranger</a> pipeline from 10X, returning a unique molecular identified (UMI) count matrix and a spatial coordinates file. The values in count matrix represent the number of molecules for each feature (i.e. gene; row) that are detected in each spot (column). The spatial information of all spots is represented by the spatial coordinates file.</p>
We next use the count matrix and sptial coordinates to create a <code>SRTProject</code> object. The object serves as a container that contains both data (like the count matrix, spatial coordinates) and analysis (like PCA, or clustering results) for a spatial transcriptomics dataset. For a technical discussion of the <code>SRTProject</code> object structure, check the ‘SRTProject-class’.
</details><details><p><summary><strong>What does data in a count matrix look like?</strong></summary></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Let's examine a few genes in the first five spots</span>
<span class="va">count</span> <span class="op">&lt;-</span> <span class="va">dlpfc_151672</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">counts</span>
<span class="va">count</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## 5 x 5 sparse Matrix of class "dgCMatrix"</span>
<span class="co">##                 AAACAAGTATCTCCCA-1 AAACACCAATAACTGC-1 AAACAGAGCGACTCCT-1</span>
<span class="co">## ENSG00000243485                  .                  .                  .</span>
<span class="co">## ENSG00000237613                  .                  .                  .</span>
<span class="co">## ENSG00000186092                  .                  .                  .</span>
<span class="co">## ENSG00000238009                  .                  .                  .</span>
<span class="co">## ENSG00000239945                  .                  .                  .</span>
<span class="co">##                 AAACAGCTTTCAGAAG-1 AAACAGGGTCTATATT-1</span>
<span class="co">## ENSG00000243485                  .                  .</span>
<span class="co">## ENSG00000237613                  .                  .</span>
<span class="co">## ENSG00000186092                  .                  .</span>
<span class="co">## ENSG00000238009                  .                  .</span>
<span class="co">## ENSG00000239945                  .                  .</span></code></pre>
<p>The <code>.</code> values in the matrix represent 0s (no molecules detected). Since most values in an count matrix are 0, a sparse-matrix representation is necessary to save memory and speed savings for data.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meta_data</span> <span class="op">&lt;-</span> <span class="va">dlpfc_151672</span><span class="op">@</span><span class="va">meta.data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">meta_data</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##                       orig.ident nCount_RNA nFeature_RNA            barcode</span>
<span class="co">## AAACAAGTATCTCCCA-1 SeuratProject       5002         2286 AAACAAGTATCTCCCA-1</span>
<span class="co">## AAACACCAATAACTGC-1 SeuratProject       5826         2638 AAACACCAATAACTGC-1</span>
<span class="co">## AAACAGAGCGACTCCT-1 SeuratProject       3813         1872 AAACAGAGCGACTCCT-1</span>
<span class="co">## AAACAGCTTTCAGAAG-1 SeuratProject       3379         1630 AAACAGCTTTCAGAAG-1</span>
<span class="co">## AAACAGGGTCTATATT-1 SeuratProject       6623         2885 AAACAGGGTCTATATT-1</span>
<span class="co">## AAACATTTCCCGGATT-1 SeuratProject       3704         1974 AAACATTTCCCGGATT-1</span>
<span class="co">##                    sample_name</span>
<span class="co">## AAACAAGTATCTCCCA-1      151672</span>
<span class="co">## AAACACCAATAACTGC-1      151672</span>
<span class="co">## AAACAGAGCGACTCCT-1      151672</span>
<span class="co">## AAACAGCTTTCAGAAG-1      151672</span>
<span class="co">## AAACAGGGTCTATATT-1      151672</span>
<span class="co">## AAACATTTCCCGGATT-1      151672</span></code></pre>
</details><p>We check the data by printing it. We can see there are 4015 spots and 33538 genes.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dlpfc_151672</span></code></pre></div>
<pre><code><span class="co">## An object of class Seurat </span>
<span class="co">## 33538 features across 4015 samples within 1 assay </span>
<span class="co">## Active assay: RNA (33538 features, 0 variable features)</span></code></pre>
<p>Check the meta data: each row contains the meta information of a spot with row name such as <code>AAACAAGTATCTCCCA-1</code>, which consists of the barcode, to identify each spot in RNA sequencing, separator <code>-</code> and a number <code>1</code>; note that the columns, <code>row</code>, <code>col</code>, <code>imagerow</code> and <code>imagecol</code>, save the spatial information of each spot, in wihch <code>row</code> and <code>col</code> are the location of a spot on the slide and will be used in spatial analysis, <code>imagerow</code> and <code>imagecol</code> are the location of a spot on the H&amp;E image; there is the manual annotations <code>layer_guess_reordered</code> based on the cytoarchitecture in the original study (Maynard et al., 2021).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dlpfc_151672</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##                       orig.ident nCount_RNA nFeature_RNA            barcode</span>
<span class="co">## AAACAAGTATCTCCCA-1 SeuratProject       5002         2286 AAACAAGTATCTCCCA-1</span>
<span class="co">## AAACACCAATAACTGC-1 SeuratProject       5826         2638 AAACACCAATAACTGC-1</span>
<span class="co">## AAACAGAGCGACTCCT-1 SeuratProject       3813         1872 AAACAGAGCGACTCCT-1</span>
<span class="co">## AAACAGCTTTCAGAAG-1 SeuratProject       3379         1630 AAACAGCTTTCAGAAG-1</span>
<span class="co">## AAACAGGGTCTATATT-1 SeuratProject       6623         2885 AAACAGGGTCTATATT-1</span>
<span class="co">## AAACATTTCCCGGATT-1 SeuratProject       3704         1974 AAACATTTCCCGGATT-1</span>
<span class="co">##                    sample_name tissue row col imagerow imagecol Cluster height</span>
<span class="co">## AAACAAGTATCTCCCA-1      151672      1  50 102 391.6742 459.7210       2    600</span>
<span class="co">## AAACACCAATAACTGC-1      151672      1  59  19 439.5590 202.4753       3    600</span>
<span class="co">## AAACAGAGCGACTCCT-1      151672      1  14  94 197.5698 435.4186       5    600</span>
<span class="co">## AAACAGCTTTCAGAAG-1      151672      1  43   9 353.2403 171.7372       4    600</span>
<span class="co">## AAACAGGGTCTATATT-1      151672      1  47  13 374.8425 184.0684       2    600</span>
<span class="co">## AAACATTTCCCGGATT-1      151672      1  61  97 450.9451 444.1044       5    600</span>
<span class="co">##                    width sum_umi sum_gene subject position replicate</span>
<span class="co">## AAACAAGTATCTCCCA-1   600    5002     2286  Br5595      300         2</span>
<span class="co">## AAACACCAATAACTGC-1   600    5826     2638  Br5595      300         2</span>
<span class="co">## AAACAGAGCGACTCCT-1   600    3813     1872  Br5595      300         2</span>
<span class="co">## AAACAGCTTTCAGAAG-1   600    3379     1630  Br5595      300         2</span>
<span class="co">## AAACAGGGTCTATATT-1   600    6623     2885  Br5595      300         2</span>
<span class="co">## AAACATTTCCCGGATT-1   600    3704     1974  Br5595      300         2</span>
<span class="co">##                    subject_position discard                       key</span>
<span class="co">## AAACAAGTATCTCCCA-1    Br5595_pos300   FALSE 151672_AAACAAGTATCTCCCA-1</span>
<span class="co">## AAACACCAATAACTGC-1    Br5595_pos300   FALSE 151672_AAACACCAATAACTGC-1</span>
<span class="co">## AAACAGAGCGACTCCT-1    Br5595_pos300   FALSE 151672_AAACAGAGCGACTCCT-1</span>
<span class="co">## AAACAGCTTTCAGAAG-1    Br5595_pos300   FALSE 151672_AAACAGCTTTCAGAAG-1</span>
<span class="co">## AAACAGGGTCTATATT-1    Br5595_pos300   FALSE 151672_AAACAGGGTCTATATT-1</span>
<span class="co">## AAACATTTCCCGGATT-1    Br5595_pos300   FALSE 151672_AAACATTTCCCGGATT-1</span>
<span class="co">##                    cell_count SNN_k50_k4 SNN_k50_k5 SNN_k50_k6 SNN_k50_k7</span>
<span class="co">## AAACAAGTATCTCCCA-1          2          1          1          2          2</span>
<span class="co">## AAACACCAATAACTGC-1          3          1          1          2          2</span>
<span class="co">## AAACAGAGCGACTCCT-1          1          1          1          2          2</span>
<span class="co">## AAACAGCTTTCAGAAG-1          3          1          1          2          2</span>
<span class="co">## AAACAGGGTCTATATT-1          5          1          1          2          2</span>
<span class="co">## AAACATTTCCCGGATT-1          2          1          1          2          2</span>
<span class="co">##                    SNN_k50_k8 SNN_k50_k9 SNN_k50_k10 SNN_k50_k11 SNN_k50_k12</span>
<span class="co">## AAACAAGTATCTCCCA-1          2          2           2           1           1</span>
<span class="co">## AAACACCAATAACTGC-1          2          2           2           1           1</span>
<span class="co">## AAACAGAGCGACTCCT-1          2          2           2           1           1</span>
<span class="co">## AAACAGCTTTCAGAAG-1          2          2           2           1           1</span>
<span class="co">## AAACAGGGTCTATATT-1          2          2           2           1           1</span>
<span class="co">## AAACATTTCCCGGATT-1          2          2           2           1           1</span>
<span class="co">##                    SNN_k50_k13 SNN_k50_k14 SNN_k50_k15 SNN_k50_k16 SNN_k50_k17</span>
<span class="co">## AAACAAGTATCTCCCA-1           1           1           4           4           4</span>
<span class="co">## AAACACCAATAACTGC-1           1           1           4           4           4</span>
<span class="co">## AAACAGAGCGACTCCT-1           1           1           5           5           5</span>
<span class="co">## AAACAGCTTTCAGAAG-1           1           5           6           6           6</span>
<span class="co">## AAACAGGGTCTATATT-1           1           1           4           4           4</span>
<span class="co">## AAACATTTCCCGGATT-1           1           5           6           6           6</span>
<span class="co">##                    SNN_k50_k18 SNN_k50_k19 SNN_k50_k20 SNN_k50_k21 SNN_k50_k22</span>
<span class="co">## AAACAAGTATCTCCCA-1           3           3           3           3           2</span>
<span class="co">## AAACACCAATAACTGC-1           3           3           3           3           2</span>
<span class="co">## AAACAGAGCGACTCCT-1           5           5           5           5           4</span>
<span class="co">## AAACAGCTTTCAGAAG-1           7           7           7           7           8</span>
<span class="co">## AAACAGGGTCTATATT-1           3           3           3           3           2</span>
<span class="co">## AAACATTTCCCGGATT-1           7           7           7           7           8</span>
<span class="co">##                    SNN_k50_k23 SNN_k50_k24 SNN_k50_k25 SNN_k50_k26 SNN_k50_k27</span>
<span class="co">## AAACAAGTATCTCCCA-1           2           2           1           1           1</span>
<span class="co">## AAACACCAATAACTGC-1           2           2           1           1           1</span>
<span class="co">## AAACAGAGCGACTCCT-1           4           4           3           3           3</span>
<span class="co">## AAACAGCTTTCAGAAG-1           8           8           8           7           7</span>
<span class="co">## AAACAGGGTCTATATT-1           2           2           1           1           1</span>
<span class="co">## AAACATTTCCCGGATT-1           8           8           8           7           7</span>
<span class="co">##                    SNN_k50_k28 GraphBased Maynard Martinowich Layer layer_guess</span>
<span class="co">## AAACAAGTATCTCCCA-1           1          2       5           4    NA      Layer5</span>
<span class="co">## AAACACCAATAACTGC-1           1          3       6           5    NA      Layer6</span>
<span class="co">## AAACAGAGCGACTCCT-1           2          5     2_3         2_3    NA      Layer3</span>
<span class="co">## AAACAGCTTTCAGAAG-1           6          4       4           4    NA      Layer5</span>
<span class="co">## AAACAGGGTCTATATT-1           1          2       5           4    NA      Layer5</span>
<span class="co">## AAACATTTCCCGGATT-1           6          5     2_3         2_3    NA      Layer6</span>
<span class="co">##                    layer_guess_reordered layer_guess_reordered_short expr_chrM</span>
<span class="co">## AAACAAGTATCTCCCA-1                Layer5                          L5      1072</span>
<span class="co">## AAACACCAATAACTGC-1                Layer6                          L6      1090</span>
<span class="co">## AAACAGAGCGACTCCT-1                Layer3                          L3       734</span>
<span class="co">## AAACAGCTTTCAGAAG-1                Layer5                          L5       824</span>
<span class="co">## AAACAGGGTCTATATT-1                Layer5                          L5      1107</span>
<span class="co">## AAACATTTCCCGGATT-1                Layer6                          L6       606</span>
<span class="co">##                    expr_chrM_ratio SpatialDE_PCA SpatialDE_pool_PCA HVG_PCA</span>
<span class="co">## AAACAAGTATCTCCCA-1       0.2143143             1                  2       3</span>
<span class="co">## AAACACCAATAACTGC-1       0.1870923             1                  1       3</span>
<span class="co">## AAACAGAGCGACTCCT-1       0.1924993             2                  4       4</span>
<span class="co">## AAACAGCTTTCAGAAG-1       0.2438591             3                  2       3</span>
<span class="co">## AAACAGGGTCTATATT-1       0.1671448             1                  2       2</span>
<span class="co">## AAACATTTCCCGGATT-1       0.1636069             3                  3       3</span>
<span class="co">##                    pseudobulk_PCA markers_PCA SpatialDE_UMAP</span>
<span class="co">## AAACAAGTATCTCCCA-1              1           3              1</span>
<span class="co">## AAACACCAATAACTGC-1              1           2              2</span>
<span class="co">## AAACAGAGCGACTCCT-1              2           4              1</span>
<span class="co">## AAACAGCTTTCAGAAG-1              1           3              1</span>
<span class="co">## AAACAGGGTCTATATT-1              1           3              1</span>
<span class="co">## AAACATTTCCCGGATT-1              1           1              2</span>
<span class="co">##                    SpatialDE_pool_UMAP HVG_UMAP pseudobulk_UMAP markers_UMAP</span>
<span class="co">## AAACAAGTATCTCCCA-1                   2        3               1            1</span>
<span class="co">## AAACACCAATAACTGC-1                   4        1               3            1</span>
<span class="co">## AAACAGAGCGACTCCT-1                   3        2               5            1</span>
<span class="co">## AAACAGCTTTCAGAAG-1                   3        7               4            1</span>
<span class="co">## AAACAGGGTCTATATT-1                   3        3               4            1</span>
<span class="co">## AAACATTTCCCGGATT-1                   4        1               1            2</span>
<span class="co">##                    SpatialDE_PCA_spatial SpatialDE_pool_PCA_spatial</span>
<span class="co">## AAACAAGTATCTCCCA-1                     1                          1</span>
<span class="co">## AAACACCAATAACTGC-1                     4                          2</span>
<span class="co">## AAACAGAGCGACTCCT-1                     2                          4</span>
<span class="co">## AAACAGCTTTCAGAAG-1                     3                          1</span>
<span class="co">## AAACAGGGTCTATATT-1                     1                          1</span>
<span class="co">## AAACATTTCCCGGATT-1                     4                          6</span>
<span class="co">##                    HVG_PCA_spatial pseudobulk_PCA_spatial markers_PCA_spatial</span>
<span class="co">## AAACAAGTATCTCCCA-1               3                      4                   3</span>
<span class="co">## AAACACCAATAACTGC-1               3                      1                   1</span>
<span class="co">## AAACAGAGCGACTCCT-1               5                      5                   2</span>
<span class="co">## AAACAGCTTTCAGAAG-1               4                      4                   1</span>
<span class="co">## AAACAGGGTCTATATT-1               1                      4                   3</span>
<span class="co">## AAACATTTCCCGGATT-1               3                      1                   1</span>
<span class="co">##                    SpatialDE_UMAP_spatial SpatialDE_pool_UMAP_spatial</span>
<span class="co">## AAACAAGTATCTCCCA-1                      1                           4</span>
<span class="co">## AAACACCAATAACTGC-1                      2                           3</span>
<span class="co">## AAACAGAGCGACTCCT-1                      1                           1</span>
<span class="co">## AAACAGCTTTCAGAAG-1                      1                           1</span>
<span class="co">## AAACAGGGTCTATATT-1                      1                           1</span>
<span class="co">## AAACATTTCCCGGATT-1                      2                           6</span>
<span class="co">##                    HVG_UMAP_spatial pseudobulk_UMAP_spatial</span>
<span class="co">## AAACAAGTATCTCCCA-1                5                       5</span>
<span class="co">## AAACACCAATAACTGC-1                1                       4</span>
<span class="co">## AAACAGAGCGACTCCT-1                3                       6</span>
<span class="co">## AAACAGCTTTCAGAAG-1                8                       5</span>
<span class="co">## AAACAGGGTCTATATT-1                2                       5</span>
<span class="co">## AAACATTTCCCGGATT-1                1                       1</span>
<span class="co">##                    markers_UMAP_spatial sizeFactor x_gmm     r_gmm.V1</span>
<span class="co">## AAACAAGTATCTCCCA-1                    3  1.3828966     7 3.047079e-07</span>
<span class="co">## AAACACCAATAACTGC-1                    1  1.6659761     7 3.869298e-04</span>
<span class="co">## AAACAGAGCGACTCCT-1                    1  1.0756848     6 4.537852e-09</span>
<span class="co">## AAACAGCTTTCAGAAG-1                    1  0.8474877     7 3.547920e-07</span>
<span class="co">## AAACAGGGTCTATATT-1                    1  1.8339177     7 2.863789e-06</span>
<span class="co">## AAACATTTCCCGGATT-1                    1  1.1119238     7 2.046383e-03</span>
<span class="co">##                        r_gmm.V2     r_gmm.V3    r_gmm.V4     r_gmm.V5</span>
<span class="co">## AAACAAGTATCTCCCA-1 1.306795e-04 2.091002e-25 0.005526178 8.700836e-03</span>
<span class="co">## AAACACCAATAACTGC-1 1.381066e-04 2.954016e-20 0.001390498 7.885426e-06</span>
<span class="co">## AAACAGAGCGACTCCT-1 8.074682e-02 7.691888e-24 0.352418690 5.246635e-04</span>
<span class="co">## AAACAGCTTTCAGAAG-1 5.011087e-03 1.279334e-23 0.002123237 5.140198e-03</span>
<span class="co">## AAACAGGGTCTATATT-1 7.863625e-06 3.956421e-23 0.000455874 4.871899e-06</span>
<span class="co">## AAACATTTCCCGGATT-1 4.154461e-03 8.955671e-17 0.002881995 6.060503e-04</span>
<span class="co">##                       r_gmm.V6  r_gmm.V7 x_icm    r_icm.V1    r_icm.V2</span>
<span class="co">## AAACAAGTATCTCCCA-1 0.164158155 0.8214838     7 0.006404860 0.006404860</span>
<span class="co">## AAACACCAATAACTGC-1 0.009975489 0.9881011     7 0.002442427 0.002442427</span>
<span class="co">## AAACAGAGCGACTCCT-1 0.464721898 0.1015879     6 0.014928224 0.110305483</span>
<span class="co">## AAACAGCTTTCAGAAG-1 0.396250473 0.5914747     7 0.042450129 0.015616530</span>
<span class="co">## AAACAGGGTCTATATT-1 0.008336799 0.9911917     7 0.002442427 0.002442427</span>
<span class="co">## AAACATTTCCCGGATT-1 0.011206997 0.9791041     1 0.950565486 0.006404860</span>
<span class="co">##                       r_icm.V3    r_icm.V4    r_icm.V5    r_icm.V6   r_icm.V7</span>
<span class="co">## AAACAAGTATCTCCCA-1 0.017410214 0.006404860 0.006404860 0.006404860 0.95056549</span>
<span class="co">## AAACACCAATAACTGC-1 0.002442427 0.002442427 0.002442427 0.002442427 0.98534544</span>
<span class="co">## AAACAGAGCGACTCCT-1 0.014928224 0.014928224 0.014928224 0.815053399 0.01492822</span>
<span class="co">## AAACAGCTTTCAGAAG-1 0.015616530 0.042450129 0.015616530 0.015616530 0.85263362</span>
<span class="co">## AAACAGGGTCTATATT-1 0.002442427 0.002442427 0.002442427 0.002442427 0.98534544</span>
<span class="co">## AAACATTTCCCGGATT-1 0.006404860 0.006404860 0.006404860 0.006404860 0.01741021</span>
<span class="co">##                    x_gc     r_gc.V1     r_gc.V2     r_gc.V3     r_gc.V4</span>
<span class="co">## AAACAAGTATCTCCCA-1    7 0.006404860 0.006404860 0.017410214 0.006404860</span>
<span class="co">## AAACACCAATAACTGC-1    7 0.002442427 0.002442427 0.002442427 0.002442427</span>
<span class="co">## AAACAGAGCGACTCCT-1    2 0.002442427 0.985345437 0.002442427 0.002442427</span>
<span class="co">## AAACAGCTTTCAGAAG-1    7 0.002442427 0.002442427 0.002442427 0.002442427</span>
<span class="co">## AAACAGGGTCTATATT-1    7 0.002442427 0.002442427 0.002442427 0.002442427</span>
<span class="co">## AAACATTTCCCGGATT-1    1 0.950565486 0.006404860 0.006404860 0.006404860</span>
<span class="co">##                        r_gc.V5     r_gc.V6     r_gc.V7</span>
<span class="co">## AAACAAGTATCTCCCA-1 0.006404860 0.006404860 0.950565486</span>
<span class="co">## AAACACCAATAACTGC-1 0.002442427 0.002442427 0.985345437</span>
<span class="co">## AAACAGAGCGACTCCT-1 0.002442427 0.002442427 0.002442427</span>
<span class="co">## AAACAGCTTTCAGAAG-1 0.002442427 0.002442427 0.985345437</span>
<span class="co">## AAACAGGGTCTATATT-1 0.002442427 0.002442427 0.985345437</span>
<span class="co">## AAACATTTCCCGGATT-1 0.006404860 0.006404860 0.017410214</span></code></pre>
<p>Check the manual annotations: this slice is annotated as five layers: Layer 3-6 and White matter.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">dlpfc_151672</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">layer_guess_reordered</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">## Layer1 Layer2 Layer3 Layer4 Layer5 Layer6     WM </span>
<span class="co">##      0      0   1575    304    728    882    399</span></code></pre>
</div>
<div class="section level2">
<h2 id="pre-processing-workflow">Pre-processing workflow<a class="anchor" aria-label="anchor" href="#pre-processing-workflow"></a>
</h2>
<div class="section level3">
<h3 id="create-a-srtproject-object">Create a SRTProject object<a class="anchor" aria-label="anchor" href="#create-a-srtproject-object"></a>
</h3>
<p>First, we show how to create a SRTProject object step by step.</p>
<p>Because the gene name is Ensembl format in this data, we change the genes’ Ensembl name to symbols for convenience of followed analyses.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">count_matrix</span> <span class="op">&lt;-</span> <span class="va">dlpfc_151672</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">counts</span>
<span class="co">## Use eg.db database: this method is fast</span>
<span class="va">symbol_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/transferGeneNames.html">transferGeneNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">count_matrix</span><span class="op">)</span>, now_name <span class="op">=</span> <span class="st">"ensembl"</span>, to_name <span class="op">=</span> <span class="st">"symbol"</span>,
    species <span class="op">=</span> <span class="st">"Human"</span>, Method <span class="op">=</span> <span class="st">"eg.db"</span><span class="op">)</span>
<span class="va">symbol_name</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">##  [1] "ENSG00000243485" "Fam138a"         "Or4f5"           "Loc100996442"   </span>
<span class="co">##  [5] "ENSG00000239945" "ENSG00000239906" "ENSG00000241599" "ENSG00000236601"</span>
<span class="co">##  [9] "Or4f29"          "ENSG00000235146"</span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">count_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">symbol_name</span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## create count matrix list: note each component has a name, i.e., `ID151672`.</span>

<span class="va">cntList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ID151672 <span class="op">=</span> <span class="va">count_matrix</span><span class="op">)</span>

<span class="co">## create spatial coordinate matrix</span>
<span class="va">coordList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">dlpfc_151672</span><span class="op">$</span><span class="va">row</span>, <span class="va">dlpfc_151672</span><span class="op">$</span><span class="va">col</span><span class="op">)</span><span class="op">)</span>

<span class="co">## create metadata list</span>
<span class="va">meta.data</span> <span class="op">&lt;-</span> <span class="va">dlpfc_151672</span><span class="op">@</span><span class="va">meta.data</span>
<span class="va">metadataList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">meta.data</span><span class="op">)</span>

<span class="co">## create meta data for each data batches. Here we only have one data batch.</span>
<span class="va">sampleMetadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">"Human"</span>, tissues <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DLPFC"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">sampleMetadata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cntList</span><span class="op">)</span>

<span class="co">## Name of this project</span>
<span class="va">projectName</span> <span class="op">&lt;-</span> <span class="st">"DLPFC151672"</span>

<span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">dlpfc_151672</span><span class="op">)</span></code></pre></div>
<p>Next, we start creating <code>SRTProject</code> object. We can print the basic information of this object, including three parts. The first part have the class of this object, outputPath of data that require to output, h5filePath that save the memory-cusuming data (i.e., count, logcount, …). The second part is about the datasets basic information, such as how many data batches(sample) and the data names, sample meta data (sampleColData) and meta data for each spot (cellMetaData). The last part is about downstream analyses information (that is empty) when this object created.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CreateSRTProject.html">CreateSRTProject</a></span><span class="op">(</span><span class="va">cntList</span>, <span class="va">coordList</span>, projectName <span class="op">=</span> <span class="va">projectName</span>, <span class="va">metadataList</span>, <span class="va">sampleMetadata</span>,
    force <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>
<span class="va">SRTProj</span></code></pre></div>
<pre><code><span class="co">## class: SRTProject </span>
<span class="co">## outputPath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\DLPFC151672 </span>
<span class="co">## h5filePath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\DLPFC151672/DLPFC151672.h5 </span>
<span class="co">## ---------Datasets basic information-----------------</span>
<span class="co">## samples(1): ID151672</span>
<span class="co">## sampleColData names(3): species tissues NumOfSpots</span>
<span class="co">## cellMetaData names(101): orig.ident nCount_RNA ... r_gc.V7 batch</span>
<span class="co">## numberOfSpots(1): 4015</span>
<span class="co">## ---------Downstream analyses information-----------------</span>
<span class="co">## Low-dimensional embeddings(0):</span>
<span class="co">## Inferred cluster labels: No</span>
<span class="co">## Embedding for plotting(0):</span></code></pre>
<details><p><summary><strong>What does data in a h5 file look like?</strong></summary></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">h5ls</span><span class="op">(</span><span class="va">SRTProj</span><span class="op">@</span><span class="va">projectMetadata</span><span class="op">$</span><span class="va">h5filePath</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##             group           name       otype   dclass      dim</span>
<span class="co">## 0               /  SpatialCoords H5I_DATASET  INTEGER 4015 x 2</span>
<span class="co">## 1               /   cellMetaData H5I_DATASET COMPOUND     4015</span>
<span class="co">## 2               /          count   H5I_GROUP                  </span>
<span class="co">## 3          /count        count_1   H5I_GROUP                  </span>
<span class="co">## 4  /count/count_1            Dim H5I_DATASET  INTEGER        2</span>
<span class="co">## 5  /count/count_1      cellNames H5I_DATASET   STRING     4015</span>
<span class="co">## 6  /count/count_1           data H5I_DATASET    FLOAT  7060060</span>
<span class="co">## 7  /count/count_1      geneNames H5I_DATASET   STRING    33538</span>
<span class="co">## 8  /count/count_1        indices H5I_DATASET  INTEGER  7060060</span>
<span class="co">## 9  /count/count_1         indptr H5I_DATASET  INTEGER     4016</span>
<span class="co">## 10              / sampleMetadata H5I_DATASET COMPOUND        1</span>
<span class="co">## 11              /    samplenames H5I_DATASET   STRING        1</span></code></pre>
</details>
</div>
<div class="section level3">
<h3 id="normalizing-the-data">Normalizing the data<a class="anchor" aria-label="anchor" href="#normalizing-the-data"></a>
</h3>
<p>After removing unwanted cells and genes from the dataset, the next step is to normalize the data. To save RAM memory, normalized values are stored in disk as a <code>h5file</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeSRT.html">normalizeSRT</a></span><span class="op">(</span><span class="va">SRTProj</span><span class="op">)</span></code></pre></div>
<div class="section level4">
<h4 id="feature-selection">Feature selection<a class="anchor" aria-label="anchor" href="#feature-selection"></a>
</h4>
<p>We next select a subset of genes that exhibit high spot-to-spot variation in the dataset (i.e, they are highly expressed in some spots, and lowly expressed in others). It has been found that focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets <a href="https://www.nature.com/articles/nmeth.2645" class="external-link">here</a>.</p>
<p>Then we choose variable features. The default number of variable features is 2,000, and users can change it using argument <code>nfeatures</code>. The default type is highly variable genes (HVGs), but users can use spatially variable genes by seting <code>type='SVGs'</code>, then <code>SPARK-X</code> will be used to choose SVGs. #### Choose highly variable features We can also directly use <code><a href="../reference/selectVariableFeatures.html">selectVariableFeatures()</a></code> function to choose the top highly variable genes (HVGs). We found the performance of using SVGs or HVGs has no significant difference for the downstream analyses, such as dimension reduction and spatial clustering.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectVariableFeatures.html">selectVariableFeatures</a></span><span class="op">(</span><span class="va">SRTProj</span><span class="op">)</span></code></pre></div>
<div class="section level5">
<h5 id="choose-spatially-variable-features">Choose spatially variable features<a class="anchor" aria-label="anchor" href="#choose-spatially-variable-features"></a>
</h5>
<p>Additionally, <code><a href="../reference/selectVariableFeatures.html">selectVariableFeatures()</a></code> function can also choose the top spatially variable genes (SVGs) based on SPARK-X. SPARK-X is embeded in the <code>DR.SC</code> R package and is implemented in the <code>FindSVGs()</code> function. By default, we return 2,000 genes per dataset. These will be used in downstream analysis, like probabilistic embeddings and spatial clustering. To further speed up the computation, we first use <code>FindVariableFeatures()</code> to select 5,000 highly variable genes, then use SPARK-X to choose top SVGs in the implementation of <code>FindSVGs()</code> function.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectVariableFeatures.html">selectVariableFeatures</a></span><span class="op">(</span><span class="va">SRTProj</span>, type <span class="op">=</span> <span class="st">"SVGs"</span>, method <span class="op">=</span> <span class="st">"spark-x"</span><span class="op">)</span></code></pre></div>
<p>Calculate the adjcence matrix</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Obtain adjacence matrix</span>
<span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddAdj.html">AddAdj</a></span><span class="op">(</span><span class="va">SRTProj</span>, platform <span class="op">=</span> <span class="st">"Visium"</span>, force <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></code></pre></div>
<p>Visualize</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">selectFeatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">SRTProj</span><span class="op">@</span><span class="va">geneMetaData</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">SRTProj</span><span class="op">@</span><span class="va">geneMetaData</span><span class="op">$</span><span class="va">`isVGs#combine`</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="fu"><a href="../reference/EachExprSpaHeatMap.html">EachExprSpaHeatMap</a></span><span class="op">(</span><span class="va">SRTProj</span>, features <span class="op">=</span> <span class="va">selectFeatures</span>, title_name <span class="op">=</span> <span class="cn">T</span>, quantVec <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.9</span>, <span class="fl">0.9</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="single_basic_tutorial_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="dimension-reduction">Dimension reduction<a class="anchor" aria-label="anchor" href="#dimension-reduction"></a>
</h2>
<p>PCA is the most popular dimension reduction technique in single cell RNA sequencing (scRNA-seq) data and SRT　data because of its simplicity, computational efficiency, and relatively comparable performance. SRTpipeline provided three versions of PCA: standard PCA (PCA), approximated PCA (APCA), and weighted PCA (WPCA) by the function <code>AddPCA</code> with default version as APCA for fastest computation. The 15-dimensional PCs are extracted by default, and users can use their own values.</p>
<p>After running <code>AddPCA</code>, we would see the output of <code>SRTProj</code> includes <code>PCA</code> in the <code>Low-dimensional embeddings</code> field.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## APCA</span>
<span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddPCA.html">AddPCA</a></span><span class="op">(</span><span class="va">SRTProj</span>, n_comp <span class="op">=</span> <span class="fl">15</span><span class="op">)</span>
<span class="co"># accurate PCA</span>

<span class="co"># SRTProj &lt;- AddPCA(SRTProj, Method='PCA')</span>

<span class="co"># weighted PCA</span>

<span class="co"># SRTProj &lt;- AddPCA(SRTProj, Method='WPCA')</span>
<span class="va">SRTProj</span></code></pre></div>
<pre><code><span class="co">## class: SRTProject </span>
<span class="co">## outputPath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\DLPFC151672 </span>
<span class="co">## h5filePath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\DLPFC151672/DLPFC151672.h5 </span>
<span class="co">## ---------Datasets basic information-----------------</span>
<span class="co">## samples(1): ID151672</span>
<span class="co">## sampleColData names(3): species tissues NumOfSpots</span>
<span class="co">## cellMetaData names(101): orig.ident nCount_RNA ... r_gc.V7 batch</span>
<span class="co">## numberOfSpots(1): 4015</span>
<span class="co">## ---------Downstream analyses information-----------------</span>
<span class="co">## Variable features:  2000 </span>
<span class="co">## Low-dimensional embeddings(1): PCA</span>
<span class="co">## Inferred cluster labels: No</span>
<span class="co">## Embedding for plotting(0):</span></code></pre>
</div>
<div class="section level2">
<h2 id="clustering-using-sc-meb">Clustering using SC-MEB<a class="anchor" aria-label="anchor" href="#clustering-using-sc-meb"></a>
</h2>
<p>Some SRT clustering methods use Markov random field to model clusters of spots, such as SC-MEB and DR-SC. These approaches work extremely well for data with spatial smoothness and are a standard practice in SRT data. For this reason, SRTpipeline uses existing state-of-the-art clustering methods from SRT packages for clustering.</p>
<p>We have had the most success using the Markov random field model implemented by SC-MEB. In SRTpipeline, SC-MEB clustering is performed using the <code><a href="../reference/Cluster_SCMEB.html">Cluster_SCMEB()</a></code> function.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Cluster_SCMEB.html">Cluster_SCMEB</a></span><span class="op">(</span><span class="va">SRTProj</span>, K <span class="op">=</span> <span class="fl">5</span>, reduction <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## diff Energy = 10.050948</span>
<span class="co">## diff Energy = 8.206142</span>
<span class="co">## diff Energy = 0.930653</span>
<span class="co">## diff Energy = 6.674451</span>
<span class="co">## diff Energy = 4.819096</span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span></code></pre></div>
<pre><code><span class="co">## class: SRTProject </span>
<span class="co">## outputPath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\DLPFC151672 </span>
<span class="co">## h5filePath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\DLPFC151672/DLPFC151672.h5 </span>
<span class="co">## ---------Datasets basic information-----------------</span>
<span class="co">## samples(1): ID151672</span>
<span class="co">## sampleColData names(3): species tissues NumOfSpots</span>
<span class="co">## cellMetaData names(101): orig.ident nCount_RNA ... r_gc.V7 batch</span>
<span class="co">## numberOfSpots(1): 4015</span>
<span class="co">## ---------Downstream analyses information-----------------</span>
<span class="co">## Variable features:  2000 </span>
<span class="co">## Low-dimensional embeddings(1): PCA</span>
<span class="co">## Inferred cluster labels: Yes</span>
<span class="co">## Embedding for plotting(0):</span></code></pre>
<div class="section level3">
<h3 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h3>
<p>To run tSNE in SRTpipeline we use the AddTSNE() function:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddTSNE.html">AddTSNE</a></span><span class="op">(</span><span class="va">SRTProj</span>, n_comp <span class="op">=</span> <span class="fl">2</span>, reduction <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></code></pre></div>
<p>The <code>reduction</code> tells <code><a href="../reference/AddTSNE.html">AddTSNE()</a></code> function uses specified reduction in the <code>SRTProj@reductions</code> slot.</p>
<p>To plot the two-dimensional tSNE results, we use the <code><a href="../reference/EmbedPlot.html">EmbedPlot()</a></code> function and pass the name of the tSNE embedding we just generated (“tSNE”). We can tell SRTpipeline how to color the spots by using <code>item</code> argument which tells SRTpipeline which matrix to use to find the specified metadata column provided to name.</p>
<p><code>SRTpipeline</code> package oﬀers a variety of visualization tools. We visualized the inferred domain types on embeddings using two components from tSNE. The tSNE plot showed the domain clusters were well segregated.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p_pca_tsne2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EmbedPlot.html">EmbedPlot</a></span><span class="op">(</span><span class="va">SRTProj</span>, item <span class="op">=</span> <span class="st">"cluster"</span>, plotEmbeddings <span class="op">=</span> <span class="st">"tSNE"</span>, cols <span class="op">=</span> <span class="fu"><a href="../reference/chooseColors.html">chooseColors</a></span><span class="op">(</span>n_colors <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>,
    legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span>
<span class="va">p_pca_tsne2</span></code></pre></div>
<p><img src="single_basic_tutorial_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<p>To run UMAP in SRTpipeline we use the AddUMAP() function. Frist, we evaluate the two-dimensional UMAPs.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddUMAP.html">AddUMAP</a></span><span class="op">(</span><span class="va">SRTProj</span>, n_comp <span class="op">=</span> <span class="fl">2</span>, reduction <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span>
<span class="va">p_pca_umap2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EmbedPlot.html">EmbedPlot</a></span><span class="op">(</span><span class="va">SRTProj</span>, item <span class="op">=</span> <span class="st">"cluster"</span>, plotEmbeddings <span class="op">=</span> <span class="st">"UMAP"</span>, cols <span class="op">=</span> <span class="fu"><a href="../reference/chooseColors.html">chooseColors</a></span><span class="op">(</span>n_colors <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>,
    legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span>
<span class="va">p_pca_umap2</span></code></pre></div>
<p><img src="single_basic_tutorial_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<p>Next, we merge all plots into one figure.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span>
<span class="va">p_all</span> <span class="op">&lt;-</span> <span class="va">p_pca_tsne2</span> <span class="op">+</span> <span class="va">p_pca_umap2</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">1</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="va">p_all</span></code></pre></div>
<p><img src="single_basic_tutorial_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<p>To save the plot, we can use <code><a href="../reference/write_fig.html">write_fig()</a></code> function.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/write_fig.html">write_fig</a></span><span class="op">(</span><span class="va">p_all</span>, filename <span class="op">=</span> <span class="st">"SCMEB_plots.png"</span><span class="op">)</span></code></pre></div>
<p>Except for the embedding plots, SRTpipeline also provides a variaty of visualization functions. First, we visualize the spatial distribution of cluster labels that shows the layer structure.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## choose colors to function chooseColors</span>
<span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chooseColors.html">chooseColors</a></span><span class="op">(</span>n_colors <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>
<span class="fu"><a href="../reference/EachClusterSpaHeatMap.html">EachClusterSpaHeatMap</a></span><span class="op">(</span><span class="va">SRTProj</span>, cols <span class="op">=</span> <span class="va">cols</span>, legend.position <span class="op">=</span> <span class="st">"right"</span>, base_size <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></code></pre></div>
<p><img src="single_basic_tutorial_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># remove the border EachClusterSpaHeatMap(SRTProj, cols=cols,</span>
<span class="co"># legend.position='right',base_size=16, border_col='white')</span></code></pre></div>
<p>We plotted the heatmap of Pearson’s correlation coefcients of the PCA embeddings among the detected domains shows the good separation of the estimated embeddings across domains and the correlations between deeper layers were high, e.g., there were high correlations between domain 2 and 3, while correlations among the separated layers were low, i.e., domain 1 and 4.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/CCHeatMap.html">CCHeatMap</a></span><span class="op">(</span><span class="va">SRTProj</span>, reduction <span class="op">=</span> <span class="st">"PCA"</span>, grp_color <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></code></pre></div>
<p><img src="single_basic_tutorial_files/figure-html/unnamed-chunk-25-1.png" width="700"> After adding the quantities for data visualization, the <code>SRTProject</code> object will have more information in the downstream analyses information. Now, we print this <code>SRTProject</code> object to check it. We observed two components added in the slot <code>plotEmbeddings</code> (Embeddings for plotting): <code>tSNE</code>, <code>tSNE3</code>, <code>UMAP</code> and <code>UMAP3</code>.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span></code></pre></div>
<pre><code><span class="co">## class: SRTProject </span>
<span class="co">## outputPath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\DLPFC151672 </span>
<span class="co">## h5filePath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\DLPFC151672/DLPFC151672.h5 </span>
<span class="co">## ---------Datasets basic information-----------------</span>
<span class="co">## samples(1): ID151672</span>
<span class="co">## sampleColData names(3): species tissues NumOfSpots</span>
<span class="co">## cellMetaData names(101): orig.ident nCount_RNA ... r_gc.V7 batch</span>
<span class="co">## numberOfSpots(1): 4015</span>
<span class="co">## ---------Downstream analyses information-----------------</span>
<span class="co">## Variable features:  2000 </span>
<span class="co">## Low-dimensional embeddings(1): PCA</span>
<span class="co">## Inferred cluster labels: Yes</span>
<span class="co">## Embedding for plotting(2): tSNE UMAP</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="deg-analysis">DEG analysis<a class="anchor" aria-label="anchor" href="#deg-analysis"></a>
</h2>
<p>To do downstream analyses, we require to get the count data from h5file. The function <code><a href="../reference/getGeneSpotData.html">getGeneSpotData()</a></code> access the gene-by-spot matrix and other data in <code>SRTProj</code> and then return a <code>SpatialExperiment</code> object, including two assays: counts, logcounts; rowData: from <code>SRTProj@geneMetaData</code>; colData: from <code>SRTProj@cellMetaData</code>, <code>SRTProj@clusters</code> and sample_id; reducedDim: from <code>SRTProj@reductions</code>.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGeneSpotData.html">getGeneSpotData</a></span><span class="op">(</span><span class="va">SRTProj</span><span class="op">)</span>
<span class="va">spe</span></code></pre></div>
<pre><code><span class="co">## class: SpatialExperiment </span>
<span class="co">## dim: 33538 4015 </span>
<span class="co">## metadata(0):</span>
<span class="co">## assays(2): counts logcounts</span>
<span class="co">## rownames(33538): ENSG00000243485 Fam138a ... ENSG00000277475</span>
<span class="co">##   ENSG00000268674</span>
<span class="co">## rowData names(3): isVGs#combine isHVGs#ID151672 rankHVGs#ID151672</span>
<span class="co">## colnames(4015): ID151672#AAACAAGTATCTCCCA-1 ID151672#AAACACCAATAACTGC-1</span>
<span class="co">##   ... ID151672#TTGTTTCCATACAACT-1 ID151672#TTGTTTGTGTAAATTC-1</span>
<span class="co">## colData names(103): orig.ident nCount_RNA ... clusters sample_id</span>
<span class="co">## reducedDimNames(4): PCA tSNE UMAP Coord</span>
<span class="co">## mainExpName: NULL</span>
<span class="co">## altExpNames(0):</span>
<span class="co">## spatialData names(0) :</span>
<span class="co">## spatialCoords names(2) : V1 V2</span>
<span class="co">## imgData names(0):</span></code></pre>
<p>After obtain the spatial cluster labels using a clustering model, we can perform differentially expression analysis. First, we detect the DE genes for cluster 3.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dat_deg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindDEGs.html">FindDEGs</a></span><span class="op">(</span><span class="va">spe</span>, cluster1 <span class="op">=</span> <span class="st">"3"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">dat_deg</span>, <span class="va">avg_log2FC</span> <span class="op">&gt;</span> <span class="fl">0.25</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 37 rows and 5 columns</span>
<span class="co">##                p_val avg_log2FC     pct.1     pct.2    p_val_adj</span>
<span class="co">##            &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;    &lt;numeric&gt;</span>
<span class="co">## Scgb2a2 2.93535e-140   0.834946     0.912     0.679 9.84457e-136</span>
<span class="co">## Mbp     3.55409e-110   0.322528     0.879     0.956 1.19197e-105</span>
<span class="co">## Plp1     3.71102e-72   0.397909     0.806     0.851  1.24460e-67</span>
<span class="co">## Gfap     5.91316e-62   0.492689     0.731     0.746  1.98316e-57</span>
<span class="co">## Scgb1d2  7.57499e-62   0.780682     0.655     0.497  2.54050e-57</span>
<span class="co">## ...              ...        ...       ...       ...          ...</span>
<span class="co">## Ccnd1     0.00364754   0.277977     0.157     0.128            1</span>
<span class="co">## Xbp1      0.00591564   0.307411     0.276     0.287            1</span>
<span class="co">## Tf        0.01093150   0.264046     0.328     0.378            1</span>
<span class="co">## Ifi27     0.01216829   0.284195     0.315     0.360            1</span>
<span class="co">## Neat1     0.02254569   0.270545     0.228     0.230            1</span></code></pre>
<p>We perform differential expression analysis for all clusters by using <code>FindAllMarkers()</code> function, then the DE genes’ information is saved in a data.frame object <code>dat_degs</code>.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dat_degs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindAllDEGs.html">FindAllDEGs</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span>
<span class="va">dat_degs</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 1679 rows and 7 columns</span>
<span class="co">##                p_val avg_log2FC     pct.1     pct.2   p_val_adj  cluster</span>
<span class="co">##            &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;factor&gt;</span>
<span class="co">## Diras2   4.73249e-76   0.483967     0.806     0.437 1.58718e-71        1</span>
<span class="co">## Themis   1.14220e-75   0.270493     0.218     0.030 3.83072e-71        1</span>
<span class="co">## Islr     2.89086e-69   0.304061     0.263     0.050 9.69538e-65        1</span>
<span class="co">## Krt17    5.45010e-69   0.372018     0.381     0.105 1.82785e-64        1</span>
<span class="co">## Nr4a2    2.43694e-68   0.280097     0.210     0.031 8.17300e-64        1</span>
<span class="co">## ...              ...        ...       ...       ...         ...      ...</span>
<span class="co">## Sncg.2   4.43041e-26   0.257604     0.829     0.582 1.48587e-21        5</span>
<span class="co">## Stx1a.1  2.36973e-25   0.255262     0.696     0.464 7.94760e-21        5</span>
<span class="co">## Hpca.1   4.13199e-25   0.255304     0.691     0.471 1.38579e-20        5</span>
<span class="co">## Cartpt.4 6.43882e-25  -0.328711     0.089     0.233 2.15945e-20        5</span>
<span class="co">## Hpcal1.2 5.78848e-15  -0.256518     0.275     0.365 1.94134e-10        5</span>
<span class="co">##                 gene</span>
<span class="co">##          &lt;character&gt;</span>
<span class="co">## Diras2        Diras2</span>
<span class="co">## Themis        Themis</span>
<span class="co">## Islr            Islr</span>
<span class="co">## Krt17          Krt17</span>
<span class="co">## Nr4a2          Nr4a2</span>
<span class="co">## ...              ...</span>
<span class="co">## Sncg.2          Sncg</span>
<span class="co">## Stx1a.1        Stx1a</span>
<span class="co">## Hpca.1          Hpca</span>
<span class="co">## Cartpt.4      Cartpt</span>
<span class="co">## Hpcal1.2      Hpcal1</span></code></pre>
<p>We identify the significant DE genes by two criteria: (a) adjustd p-value less than 0.01 and (b) average log fold change greater than 0.4.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">degs_sig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">dat_degs</span>, <span class="va">p_val_adj</span> <span class="op">&lt;</span> <span class="fl">0.01</span> <span class="op">&amp;</span> <span class="va">avg_log2FC</span> <span class="op">&gt;</span> <span class="fl">0.25</span><span class="op">)</span>
<span class="va">degs_sig</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 99 rows and 7 columns</span>
<span class="co">##               p_val avg_log2FC     pct.1     pct.2   p_val_adj  cluster</span>
<span class="co">##           &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;factor&gt;</span>
<span class="co">## Diras2  4.73249e-76   0.483967     0.806     0.437 1.58718e-71        1</span>
<span class="co">## Themis  1.14220e-75   0.270493     0.218     0.030 3.83072e-71        1</span>
<span class="co">## Islr    2.89086e-69   0.304061     0.263     0.050 9.69538e-65        1</span>
<span class="co">## Krt17   5.45010e-69   0.372018     0.381     0.105 1.82785e-64        1</span>
<span class="co">## Nr4a2   2.43694e-68   0.280097     0.210     0.031 8.17300e-64        1</span>
<span class="co">## ...             ...        ...       ...       ...         ...      ...</span>
<span class="co">## Mgll.1  1.08554e-27   0.260290     0.627     0.387 3.64067e-23        5</span>
<span class="co">## Ssbp3.1 1.64065e-27   0.258359     0.575     0.346 5.50240e-23        5</span>
<span class="co">## Sncg.2  4.43041e-26   0.257604     0.829     0.582 1.48587e-21        5</span>
<span class="co">## Stx1a.1 2.36973e-25   0.255262     0.696     0.464 7.94760e-21        5</span>
<span class="co">## Hpca.1  4.13199e-25   0.255304     0.691     0.471 1.38579e-20        5</span>
<span class="co">##                gene</span>
<span class="co">##         &lt;character&gt;</span>
<span class="co">## Diras2       Diras2</span>
<span class="co">## Themis       Themis</span>
<span class="co">## Islr           Islr</span>
<span class="co">## Krt17         Krt17</span>
<span class="co">## Nr4a2         Nr4a2</span>
<span class="co">## ...             ...</span>
<span class="co">## Mgll.1         Mgll</span>
<span class="co">## Ssbp3.1       Ssbp3</span>
<span class="co">## Sncg.2         Sncg</span>
<span class="co">## Stx1a.1       Stx1a</span>
<span class="co">## Hpca.1         Hpca</span></code></pre>
<p>In the following, we perform gene set enrichment analysis for the DE genes of each Domain identified by DR-SC model using R package <code>gprofiler2</code>.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">gprofiler2</span><span class="op">)</span>
<span class="va">termList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span> <span class="op">{</span>
    <span class="co"># k &lt;- 1</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"k = "</span>, <span class="va">k</span>, <span class="st">"\n"</span><span class="op">)</span>
    <span class="va">dat_degs_sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">degs_sig</span>, <span class="va">cluster</span> <span class="op">==</span> <span class="va">k</span><span class="op">)</span>

    <span class="va">que1</span> <span class="op">&lt;-</span> <span class="va">dat_degs_sub</span><span class="op">$</span><span class="va">gene</span>
    <span class="va">gostres</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gprofiler2/man/gost.html" class="external-link">gost</a></span><span class="op">(</span>query <span class="op">=</span> <span class="va">que1</span>, organism <span class="op">=</span> <span class="st">"hsapiens"</span>, correction_method <span class="op">=</span> <span class="st">"fdr"</span><span class="op">)</span>
    <span class="va">termList</span><span class="op">[[</span><span class="va">k</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">gostres</span>
<span class="op">}</span></code></pre></div>
<pre><code><span class="co">## k =  1 </span>
<span class="co">## k =  2 </span>
<span class="co">## k =  3 </span>
<span class="co">## k =  4 </span>
<span class="co">## k =  5</span></code></pre>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">termList</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">result</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##     query significant     p_value term_size query_size intersection_size</span>
<span class="co">## 1 query_1        TRUE 0.005374524         3          5                 1</span>
<span class="co">## 2 query_1        TRUE 0.005374524         3          5                 1</span>
<span class="co">## 3 query_1        TRUE 0.005374524         3          5                 1</span>
<span class="co">## 4 query_1        TRUE 0.005374524         2          5                 1</span>
<span class="co">## 5 query_1        TRUE 0.005374524         3          5                 1</span>
<span class="co">## 6 query_1        TRUE 0.005374524         3          5                 1</span>
<span class="co">##   precision    recall    term_id source                               term_name</span>
<span class="co">## 1       0.2 0.3333333 CORUM:1615  CORUM     G protein complex (BTK, GNG1, GNG2)</span>
<span class="co">## 2       0.2 0.3333333 CORUM:1539  CORUM  G protein complex (GNG2, GNB2L1, RAF1)</span>
<span class="co">## 3       0.2 0.3333333 CORUM:1612  CORUM                  GNAS-GNB1-GNG2 complex</span>
<span class="co">## 4       0.2 0.5000000 CORUM:1614  CORUM    G protein complex (MCF2, GNB1, GNG2)</span>
<span class="co">## 5       0.2 0.3333333 CORUM:1617  CORUM G protein complex (CACNA1A, GNB1, GNG2)</span>
<span class="co">## 6       0.2 0.3333333 CORUM:7148  CORUM                GNAS-L-GNB2-GNG2 complex</span>
<span class="co">##   effective_domain_size source_order       parents</span>
<span class="co">## 1                  3385          789 CORUM:0000000</span>
<span class="co">## 2                  3385          776 CORUM:0000000</span>
<span class="co">## 3                  3385          787 CORUM:0000000</span>
<span class="co">## 4                  3385          788 CORUM:0000000</span>
<span class="co">## 5                  3385          790 CORUM:0000000</span>
<span class="co">## 6                  3385         2637 CORUM:0000000</span></code></pre>
<p>To understand the functions of the identified spatial domains by DR-SC model, we compare the top significant biological process (BP) pathways in GO database for the DE genes from Domain 1 and 2. Here, we only show to visualize the significant BP pathways and users can explore the other databases such as KEGG and HPA.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Most commonly used databases</span>
<span class="va">source_set</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GO:BP"</span>, <span class="st">"GO:CC"</span>, <span class="st">"GO:MF"</span>, <span class="st">"KEGG"</span>, <span class="st">"HPA"</span><span class="op">)</span>
<span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"steelblue3"</span>, <span class="st">"goldenrod"</span>, <span class="st">"brown3"</span>, <span class="st">"#f98866"</span>, <span class="st">"#CE6DBD"</span><span class="op">)</span>
<span class="co">## Here, we show GO:BP</span>
<span class="va">source1</span> <span class="op">&lt;-</span> <span class="st">"GO:BP"</span>
<span class="va">ss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">source_set</span> <span class="op">==</span> <span class="va">source1</span><span class="op">)</span>
<span class="va">ntop</span> <span class="op">=</span> <span class="fl">5</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cols</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">source_set</span>
<span class="va">pList_enrich</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">ii</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span> <span class="op">{</span>
    <span class="co">## ii &lt;- 3</span>
    <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"ii="</span>, <span class="va">ii</span><span class="op">)</span>
    <span class="va">gostres2</span> <span class="op">&lt;-</span> <span class="va">termList</span><span class="op">[[</span><span class="va">ii</span><span class="op">]</span><span class="op">]</span>
    <span class="va">dat1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">gostres2</span><span class="op">$</span><span class="va">result</span>, <span class="va">term_size</span> <span class="op">&lt;</span> <span class="fl">500</span><span class="op">)</span>
    <span class="va">dat1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_top_pathway.html">get_top_pathway</a></span><span class="op">(</span><span class="va">dat1</span>, ntop <span class="op">=</span> <span class="va">ntop</span>, source_set <span class="op">=</span> <span class="va">source1</span><span class="op">)</span>
    <span class="va">dat1</span> <span class="op">&lt;-</span> <span class="va">dat1</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/stats/complete.cases.html" class="external-link">complete.cases</a></span><span class="op">(</span><span class="va">dat1</span><span class="op">)</span>, <span class="op">]</span>
    <span class="va">dat1</span><span class="op">$</span><span class="va">nlog10P</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">dat1</span><span class="op">$</span><span class="va">p_value</span><span class="op">)</span>

    <span class="va">pList_enrich</span><span class="op">[[</span><span class="va">ii</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/barPlot_enrich.html">barPlot_enrich</a></span><span class="op">(</span><span class="va">dat1</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">dat1</span><span class="op">$</span><span class="va">nlog10P</span><span class="op">)</span>, <span class="op">]</span>, source <span class="op">=</span> <span class="st">"source"</span>, <span class="st">"term_name"</span>,
        <span class="st">"nlog10P"</span>, cols <span class="op">=</span> <span class="va">cols</span><span class="op">[</span><span class="va">source_set</span><span class="op">[</span><span class="va">ss</span><span class="op">]</span><span class="op">]</span>, base_size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ylab</span><span class="op">(</span><span class="st">"-log10(p-adj)"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">xlab</span><span class="op">(</span><span class="st">"Biological terms"</span><span class="op">)</span> <span class="op">+</span>
        <span class="fu">ggtitle</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Domain"</span>, <span class="va">ii</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>

<span class="fu"><a href="../reference/drawFigs.html">drawFigs</a></span><span class="op">(</span><span class="va">pList_enrich</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">]</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span>, common.legend <span class="op">=</span> <span class="cn">T</span>, align <span class="op">=</span> <span class="st">"hv"</span><span class="op">)</span></code></pre></div>
<p><img src="single_basic_tutorial_files/figure-html/unnamed-chunk-32-1.png" width="700"></p>
<p>We take out the top DE genes for each cluster for visualization.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>
<span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">3</span>
<span class="va">dat_degs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="va">as.data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html" class="external-link">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n</span>, wt <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">topGene</span>
<span class="va">topGene</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 15 x 7</span></span>
<span class="co">## <span style="color: #949494;"># Groups:   cluster [5]</span></span>
<span class="co">##        p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene   </span>
<span class="co">##        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> 4.73<span style="color: #949494;">e</span><span style="color: #BB0000;">- 76</span>      0.484 0.806 0.437 1.59<span style="color: #949494;">e</span><span style="color: #BB0000;">- 71</span> 1       Diras2 </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> 2.26<span style="color: #949494;">e</span><span style="color: #BB0000;">- 60</span>      0.404 0.524 0.205 7.57<span style="color: #949494;">e</span><span style="color: #BB0000;">- 56</span> 1       Mmd    </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> 2.04<span style="color: #949494;">e</span><span style="color: #BB0000;">- 53</span>      0.400 0.764 0.449 6.84<span style="color: #949494;">e</span><span style="color: #BB0000;">- 49</span> 1       Map2k1 </span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> 8.45<span style="color: #949494;">e</span><span style="color: #BB0000;">-103</span>      0.520 0.95  0.637 2.83<span style="color: #949494;">e</span><span style="color: #BB0000;">- 98</span> 2       Scgb2a2</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> 5.19<span style="color: #949494;">e</span><span style="color: #BB0000;">- 87</span>      0.499 0.778 0.441 1.74<span style="color: #949494;">e</span><span style="color: #BB0000;">- 82</span> 2       Scgb1d2</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> 7.44<span style="color: #949494;">e</span><span style="color: #BB0000;">- 35</span>      0.340 0.48  0.306 2.49<span style="color: #949494;">e</span><span style="color: #BB0000;">- 30</span> 2       Hpcal1 </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> 2.94<span style="color: #949494;">e</span><span style="color: #BB0000;">-140</span>      0.835 0.912 0.679 9.84<span style="color: #949494;">e</span><span style="color: #BB0000;">-136</span> 3       Scgb2a2</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> 7.57<span style="color: #949494;">e</span><span style="color: #BB0000;">- 62</span>      0.781 0.655 0.497 2.54<span style="color: #949494;">e</span><span style="color: #BB0000;">- 57</span> 3       Scgb1d2</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> 3.68<span style="color: #949494;">e</span><span style="color: #BB0000;">- 53</span>      0.795 0.517 0.303 1.23<span style="color: #949494;">e</span><span style="color: #BB0000;">- 48</span> 3       Krt8   </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> 7.53<span style="color: #949494;">e</span><span style="color: #BB0000;">- 58</span>      0.328 0.793 0.594 2.53<span style="color: #949494;">e</span><span style="color: #BB0000;">- 53</span> 4       Hopx   </span>
<span class="co">## <span style="color: #BCBCBC;">11</span> 8.59<span style="color: #949494;">e</span><span style="color: #BB0000;">- 46</span>      0.341 0.343 0.147 2.88<span style="color: #949494;">e</span><span style="color: #BB0000;">- 41</span> 4       Cartpt </span>
<span class="co">## <span style="color: #BCBCBC;">12</span> 4.11<span style="color: #949494;">e</span><span style="color: #BB0000;">- 45</span>      0.309 0.703 0.53  1.38<span style="color: #949494;">e</span><span style="color: #BB0000;">- 40</span> 4       Nefm   </span>
<span class="co">## <span style="color: #BCBCBC;">13</span> 7.52<span style="color: #949494;">e</span><span style="color: #BB0000;">-172</span>      0.686 0.723 0.241 2.52<span style="color: #949494;">e</span><span style="color: #BB0000;">-167</span> 5       Pcp4   </span>
<span class="co">## <span style="color: #BCBCBC;">14</span> 2.06<span style="color: #949494;">e</span><span style="color: #BB0000;">- 50</span>      0.359 0.627 0.338 6.91<span style="color: #949494;">e</span><span style="color: #BB0000;">- 46</span> 5       Fam3c  </span>
<span class="co">## <span style="color: #BCBCBC;">15</span> 8.08<span style="color: #949494;">e</span><span style="color: #BB0000;">- 45</span>      0.333 0.79  0.503 2.71<span style="color: #949494;">e</span><span style="color: #BB0000;">- 40</span> 5       Cox7a1</span></code></pre>
<p>We visualize the DE genes for each cluster group by gene-by-cell heatmap using the <code><a href="../reference/GCHeatMap.html">GCHeatMap()</a></code> function.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cols_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chooseColors.html">chooseColors</a></span><span class="op">(</span>n_colors <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>

<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EachGCHeatMap.html">EachGCHeatMap</a></span><span class="op">(</span><span class="va">spe</span>, features <span class="op">=</span> <span class="va">topGene</span><span class="op">$</span><span class="va">gene</span>, grp_color <span class="op">=</span> <span class="va">cols_cluster</span>, y_text_size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>
<span class="va">p1</span></code></pre></div>
<p><img src="single_basic_tutorial_files/figure-html/unnamed-chunk-34-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="trajectory-inference">Trajectory inference<a class="anchor" aria-label="anchor" href="#trajectory-inference"></a>
</h2>
<p>Next, we performed trajectory inference using the PCA embeddings and domain labels estimated by SC-MEB model. The <code><a href="../reference/EmbedPlot.html">EmbedPlot()</a></code> function can be used to visualize the inferred pseudotime on a specified embedding. If there is only one data batch, the function <code><a href="../reference/EachEmbedPlot.html">EachEmbedPlot()</a></code> plays the same role as function <code><a href="../reference/EmbedPlot.html">EmbedPlot()</a></code>.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddTrajectory.html">AddTrajectory</a></span><span class="op">(</span><span class="va">spe</span>, reduction <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span>
<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EmbedPlot.html">EmbedPlot</a></span><span class="op">(</span><span class="va">spe</span>, reduction <span class="op">=</span> <span class="st">"PCA"</span>, colour_by <span class="op">=</span> <span class="st">"PT"</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EmbedPlot.html">EmbedPlot</a></span><span class="op">(</span><span class="va">spe</span>, reduction <span class="op">=</span> <span class="st">"Coord"</span>, colour_by <span class="op">=</span> <span class="st">"PT"</span><span class="op">)</span>
<span class="fu"><a href="../reference/drawFigs.html">drawFigs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span><span class="op">)</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, common.legend <span class="op">=</span> <span class="cn">TRUE</span>, legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></code></pre></div>
<p><img src="single_basic_tutorial_files/figure-html/unnamed-chunk-35-1.png" width="700"></p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># EachEmbedPlot(spe, reduction = 'PCA', colour_by='TrajPT') save(spe, SRTProj,</span>
<span class="co"># file=paste0(SRTProj@projectMetadata$outputPath,'/SRTProj.rds')) load('F:/Research</span>
<span class="co"># paper/IntegrateDRcluster/AnalysisCode/SRTpipeline/vignettes/DLPFC151672/SRTProj.rds')</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="other-downstream-analyses">Other downstream analyses<a class="anchor" aria-label="anchor" href="#other-downstream-analyses"></a>
</h2>
<details><p><summary><strong>Session Info</strong></summary></p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.1.2 (2021-11-01)</span>
<span class="co">## Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">## Running under: Windows 10 x64 (build 22621)</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">## [1] LC_COLLATE=Chinese (Simplified)_China.936 </span>
<span class="co">## [2] LC_CTYPE=Chinese (Simplified)_China.936   </span>
<span class="co">## [3] LC_MONETARY=Chinese (Simplified)_China.936</span>
<span class="co">## [4] LC_NUMERIC=C                              </span>
<span class="co">## [5] LC_TIME=Chinese (Simplified)_China.936    </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">## [8] methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] scater_1.25.1               scuttle_1.4.0              </span>
<span class="co">##  [3] slingshot_2.2.0             TrajectoryUtils_1.2.0      </span>
<span class="co">##  [5] princurve_2.1.6             gprofiler2_0.2.1           </span>
<span class="co">##  [7] SpatialExperiment_1.4.0     SingleCellExperiment_1.16.0</span>
<span class="co">##  [9] SummarizedExperiment_1.24.0 GenomicRanges_1.46.1       </span>
<span class="co">## [11] GenomeInfoDb_1.30.1         MatrixGenerics_1.6.0       </span>
<span class="co">## [13] matrixStats_0.62.0          patchwork_1.1.1            </span>
<span class="co">## [15] colorspace_2.0-3            SC.MEB_1.1                 </span>
<span class="co">## [17] mclust_5.4.10               irlba_2.3.5                </span>
<span class="co">## [19] DR.SC_3.1                   spatstat.geom_2.4-0        </span>
<span class="co">## [21] spatstat.data_3.0-0         ggplot2_3.3.6              </span>
<span class="co">## [23] dplyr_1.0.9                 Matrix_1.4-0               </span>
<span class="co">## [25] sp_1.5-0                    SeuratObject_4.1.0         </span>
<span class="co">## [27] Seurat_4.1.1                hdf5r_1.3.5                </span>
<span class="co">## [29] ff_4.0.7                    bit_4.0.4                  </span>
<span class="co">## [31] rhdf5_2.38.0                org.Hs.eg.db_3.14.0        </span>
<span class="co">## [33] AnnotationDbi_1.56.2        IRanges_2.28.0             </span>
<span class="co">## [35] S4Vectors_0.32.3            Biobase_2.54.0             </span>
<span class="co">## [37] BiocGenerics_0.40.0         SRTpipeline_0.1.1          </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] scattermore_0.8           ggthemes_4.2.4           </span>
<span class="co">##   [3] R.methodsS3_1.8.1         GiRaF_1.0.1              </span>
<span class="co">##   [5] ragg_1.2.2                tidyr_1.2.0              </span>
<span class="co">##   [7] bit64_4.0.5               knitr_1.37               </span>
<span class="co">##   [9] R.utils_2.11.0            DelayedArray_0.20.0      </span>
<span class="co">##  [11] data.table_1.14.2         rpart_4.1.16             </span>
<span class="co">##  [13] KEGGREST_1.34.0           RCurl_1.98-1.6           </span>
<span class="co">##  [15] generics_0.1.2            ScaledMatrix_1.2.0       </span>
<span class="co">##  [17] cowplot_1.1.1             RSQLite_2.2.10           </span>
<span class="co">##  [19] RANN_2.6.1                future_1.26.1            </span>
<span class="co">##  [21] httpuv_1.6.5              assertthat_0.2.1         </span>
<span class="co">##  [23] viridis_0.6.2             xfun_0.29                </span>
<span class="co">##  [25] jquerylib_0.1.4           evaluate_0.15            </span>
<span class="co">##  [27] promises_1.2.0.1          fansi_1.0.3              </span>
<span class="co">##  [29] igraph_1.3.5              DBI_1.1.2                </span>
<span class="co">##  [31] htmlwidgets_1.5.4         purrr_0.3.4              </span>
<span class="co">##  [33] ellipsis_0.3.2            RSpectra_0.16-1          </span>
<span class="co">##  [35] ggpubr_0.4.0              backports_1.4.1          </span>
<span class="co">##  [37] deldir_1.0-6              sparseMatrixStats_1.6.0  </span>
<span class="co">##  [39] vctrs_0.4.1               ROCR_1.0-11              </span>
<span class="co">##  [41] abind_1.4-5               cachem_1.0.6             </span>
<span class="co">##  [43] withr_2.5.0               PRECAST_1.4              </span>
<span class="co">##  [45] progressr_0.10.1          sctransform_0.3.3        </span>
<span class="co">##  [47] goftest_1.2-3             cluster_2.1.2            </span>
<span class="co">##  [49] lazyeval_0.2.2            crayon_1.5.1             </span>
<span class="co">##  [51] edgeR_3.36.0              pkgconfig_2.0.3          </span>
<span class="co">##  [53] labeling_0.4.2            nlme_3.1-155             </span>
<span class="co">##  [55] vipor_0.4.5               rlang_1.0.2              </span>
<span class="co">##  [57] globals_0.15.0            lifecycle_1.0.1          </span>
<span class="co">##  [59] miniUI_0.1.1.1            rsvd_1.0.5               </span>
<span class="co">##  [61] rprojroot_2.0.3           polyclip_1.10-0          </span>
<span class="co">##  [63] lmtest_0.9-40             carData_3.0-5            </span>
<span class="co">##  [65] Rhdf5lib_1.16.0           zoo_1.8-10               </span>
<span class="co">##  [67] beeswarm_0.4.0            ggridges_0.5.3           </span>
<span class="co">##  [69] rjson_0.2.21              png_0.1-7                </span>
<span class="co">##  [71] viridisLite_0.4.0         bitops_1.0-7             </span>
<span class="co">##  [73] R.oo_1.24.0               KernSmooth_2.23-20       </span>
<span class="co">##  [75] rhdf5filters_1.6.0        Biostrings_2.62.0        </span>
<span class="co">##  [77] blob_1.2.2                DelayedMatrixStats_1.16.0</span>
<span class="co">##  [79] stringr_1.4.0             parallelly_1.32.0        </span>
<span class="co">##  [81] spatstat.random_2.2-0     rstatix_0.7.0            </span>
<span class="co">##  [83] ggsignif_0.6.3            beachmat_2.10.0          </span>
<span class="co">##  [85] scales_1.2.0              memoise_2.0.1            </span>
<span class="co">##  [87] magrittr_2.0.3            plyr_1.8.7               </span>
<span class="co">##  [89] ica_1.0-2                 zlibbioc_1.40.0          </span>
<span class="co">##  [91] compiler_4.1.2            dqrng_0.3.0              </span>
<span class="co">##  [93] RColorBrewer_1.1-3        fitdistrplus_1.1-8       </span>
<span class="co">##  [95] cli_3.2.0                 XVector_0.34.0           </span>
<span class="co">##  [97] listenv_0.8.0             pbapply_1.5-0            </span>
<span class="co">##  [99] formatR_1.11              MASS_7.3-55              </span>
<span class="co">## [101] mgcv_1.8-39               tidyselect_1.1.2         </span>
<span class="co">## [103] stringi_1.7.6             textshaping_0.3.6        </span>
<span class="co">## [105] highr_0.9                 yaml_2.3.6               </span>
<span class="co">## [107] locfit_1.5-9.4            BiocSingular_1.10.0      </span>
<span class="co">## [109] ggrepel_0.9.1             grid_4.1.2               </span>
<span class="co">## [111] sass_0.4.1                tools_4.1.2              </span>
<span class="co">## [113] future.apply_1.9.0        rstudioapi_0.13          </span>
<span class="co">## [115] gridExtra_2.3             farver_2.1.0             </span>
<span class="co">## [117] Rtsne_0.16                DropletUtils_1.14.2      </span>
<span class="co">## [119] digest_0.6.29             rgeos_0.5-9              </span>
<span class="co">## [121] FNN_1.1.3.1               shiny_1.7.1              </span>
<span class="co">## [123] Rcpp_1.0.10               car_3.0-12               </span>
<span class="co">## [125] broom_0.7.12              later_1.3.0              </span>
<span class="co">## [127] RcppAnnoy_0.0.19          httr_1.4.3               </span>
<span class="co">## [129] fs_1.5.2                  tensor_1.5               </span>
<span class="co">## [131] reticulate_1.25           splines_4.1.2            </span>
<span class="co">## [133] uwot_0.1.11               spatstat.utils_3.0-1     </span>
<span class="co">## [135] pkgdown_2.0.6             plotly_4.10.0            </span>
<span class="co">## [137] systemfonts_1.0.4         xtable_1.8-4             </span>
<span class="co">## [139] jsonlite_1.8.0            R6_2.5.1                 </span>
<span class="co">## [141] pillar_1.7.0              htmltools_0.5.2          </span>
<span class="co">## [143] mime_0.12                 glue_1.6.2               </span>
<span class="co">## [145] fastmap_1.1.0             BiocParallel_1.28.3      </span>
<span class="co">## [147] BiocNeighbors_1.12.0      codetools_0.2-18         </span>
<span class="co">## [149] utf8_1.2.2                lattice_0.20-45          </span>
<span class="co">## [151] bslib_0.3.1               spatstat.sparse_2.1-1    </span>
<span class="co">## [153] tibble_3.1.7              ggbeeswarm_0.6.0         </span>
<span class="co">## [155] leiden_0.4.2              gtools_3.9.2.2           </span>
<span class="co">## [157] magick_2.7.3              limma_3.50.1             </span>
<span class="co">## [159] survival_3.2-13           CompQuadForm_1.4.3       </span>
<span class="co">## [161] rmarkdown_2.11            desc_1.4.0               </span>
<span class="co">## [163] munsell_0.5.0             GenomeInfoDbData_1.2.7   </span>
<span class="co">## [165] HDF5Array_1.22.1          reshape2_1.4.4           </span>
<span class="co">## [167] gtable_0.3.0              spatstat.core_2.4-4</span></code></pre>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Wei Liu, Yi Yang, Xu Liao, Xiao Zhang, Jin Liu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
