<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SpatialAnno for MOB • SRTpipeline</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="SpatialAnno for MOB">
<meta property="og:description" content="SRTpipeline">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SRTpipeline</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../articles/Installation.html">
    <span class="fas fa-book"></span>
     
    Install
  </a>
</li>
<li>
  <a href="../articles/get_started.html">
    <span class="fas fa-book"></span>
     
    Get Started
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fas fa-book"></span>
     
    Single Data Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/single_basic_tutorial.html">Basic tutorial</a>
    </li>
    <li>
      <a href="../articles/dlpfc_tutorial.html">Embedding and clustering introduction</a>
    </li>
    <li>
      <a href="../articles/single_Embryo_SC_tutorial.html">SRT data clustering</a>
    </li>
    <li>
      <a href="../articles/single_Embryo_DRSC_tutorial.html">SRT data embedding and clustering</a>
    </li>
    <li>
      <a href="../articles/single_MOB_Anno_tutorial.html">SRT data embedding and annotation</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fas fa-book"></span>
     
    Multiple Data Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/multibatch_basic_tutorial.html">Basic tutorial</a>
    </li>
    <li>
      <a href="../articles/integration_introduction.html">Basic tutorial v2</a>
    </li>
    <li>
      <a href="../articles/multibatch_brain12_tutorial.html">SRT data clustering and integration</a>
    </li>
    <li>
      <a href="../articles/multibatch_HCC4_tutorial.html">SRT data embedding, clustering and integration example 1</a>
    </li>
    <li>
      <a href="../articles/multibatch_liver8_tutorial.html">SRT data embedding, clustering and integration example 2</a>
    </li>
    <li>
      <a href="../articles/multibatch_MOB16_tutorial.html">SRT data embedding, clustering and integration example 3</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/faq.html">
    <span class="fa fa-question-circle-o"></span>
     
    FAQ
  </a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>SpatialAnno for MOB</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/feiyoung/SRTpipeline/blob/HEAD/vignettes/single_MOB_Anno_tutorial.Rmd"><code>vignettes/single_MOB_Anno_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>single_MOB_Anno_tutorial.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This tutorial demonstrates how to use <code>SRTpipeline</code> (&gt;=0.1.0) to automatically annotate a spatially-resolved transcriptomics (SRT) data. The analytical pipelines are similar to the <code>SRTpipeline</code> workflow in DR-SC model <a href="./index.html">vegenettes</a>. We emphases how to use <code>SpatialAnno</code> model to achieve spatial embedding and annotation and its followed applications. This tutorial will cover the following tasks, which we believe will be common for many spatial analyses:</p>
<ul>
<li>Normalization</li>
<li>Joint dimension reduction and annotation</li>
<li>Visualization</li>
<li>DEG analysis</li>
<li>Spatial trajectory inference</li>
</ul>
<p>First, we load <code>SRTpipeline</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feiyoung/SRTpipeline">SRTpipeline</a></span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dataset">Dataset<a class="anchor" aria-label="anchor" href="#dataset"></a>
</h2>
<p>We obtained the mouse olfactory bulb (OB) ST data from the spatial transcriptomics research website (<a href="https://www.spatialresearch.org/" class="external-link uri">https://www.spatialresearch.org/</a>). This data consists of gene expression levels in form of read counts which are collected for a number of spatial locations. We followed to focus on the mouse OB section 12, which contains 16,034 genes and 282 spatial locations. The gene expression of mouse OB section 12 and ground truth are both stored in the R package <code>SpatialAnno</code>.</p>
<div class="section level3">
<h3 id="load-the-mob-dataset">load the MOB dataset<a class="anchor" aria-label="anchor" href="#load-the-mob-dataset"></a>
</h3>
<p>First, we load the MOB dataset and extract the position of each spot from spot name. The four anatomic layers manually annotated based on H&amp;E staining was taken as ground truth.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SpatialAnno</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/find.package.html" class="external-link">path.package</a></span><span class="op">(</span><span class="st">"SpatialAnno"</span><span class="op">)</span>, <span class="st">"/extdata/Rep12_MOB_count_matrix-1.RData"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">MOB_raw</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## 6 x 6 sparse Matrix of class "dgCMatrix"</span>
<span class="co">##         16.918x16.996 18.017x17.034 20.075x17.059 18.979x17.065 21.937x16.967</span>
<span class="co">## Plekha1             5             1             .             .             1</span>
<span class="co">## Cnih3               1             1             1             .             .</span>
<span class="co">## Gpc5                1             .             .             .             .</span>
<span class="co">## Sirt3               2             1             .             .             .</span>
<span class="co">## Rab8a               5             1             .             .             1</span>
<span class="co">## Nadk                1             1             .             1             1</span>
<span class="co">##         20.899x17.015</span>
<span class="co">## Plekha1             .</span>
<span class="co">## Cnih3               1</span>
<span class="co">## Gpc5                1</span>
<span class="co">## Sirt3               .</span>
<span class="co">## Rab8a               .</span>
<span class="co">## Nadk                .</span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pos</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind.data.frame</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">MOB_raw</span><span class="op">)</span>,split<span class="op">=</span><span class="st">"x"</span><span class="op">)</span>,<span class="st">"["</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span>,
                       y<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">MOB_raw</span><span class="op">)</span>,split<span class="op">=</span><span class="st">"x"</span><span class="op">)</span>,<span class="st">"["</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pos</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##        x      y</span>
<span class="co">## 1 16.918 16.996</span>
<span class="co">## 2 18.017 17.034</span>
<span class="co">## 3 20.075 17.059</span>
<span class="co">## 4 18.979 17.065</span>
<span class="co">## 5 21.937 16.967</span>
<span class="co">## 6 20.899 17.015</span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/find.package.html" class="external-link">path.package</a></span><span class="op">(</span><span class="st">"SpatialAnno"</span><span class="op">)</span>, <span class="st">"/extdata/Rep12_MOB_manual_annotation.RData"</span><span class="op">)</span><span class="op">)</span>
<span class="va">y2</span> <span class="op">=</span> <span class="va">Rep12_MOB_manual_annotation</span>
<span class="va">y2</span><span class="op">[</span><span class="va">y2</span><span class="op">==</span><span class="st">"GCL"</span><span class="op">]</span> <span class="op">=</span> <span class="st">"GC"</span>
<span class="va">y2</span><span class="op">[</span><span class="va">y2</span><span class="op">==</span><span class="st">"MCL"</span><span class="op">]</span> <span class="op">=</span> <span class="st">"M/TC"</span>
<span class="va">y2</span><span class="op">[</span><span class="va">y2</span><span class="op">==</span><span class="st">"ONL"</span><span class="op">]</span> <span class="op">=</span> <span class="st">"OSNs"</span>
<span class="va">y2</span><span class="op">[</span><span class="va">y2</span><span class="op">==</span><span class="st">"GL"</span><span class="op">]</span> <span class="op">=</span> <span class="st">"PGC"</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">y2</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## y2</span>
<span class="co">##      GC    M/TC    OSNs     PGC Unknown </span>
<span class="co">##      68      75      55      80       4</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="pre-processing-workflow">Pre-processing workflow<a class="anchor" aria-label="anchor" href="#pre-processing-workflow"></a>
</h2>
<div class="section level3">
<h3 id="create-a-srtproject-object">Create a SRTProject object<a class="anchor" aria-label="anchor" href="#create-a-srtproject-object"></a>
</h3>
<p>First, we show how to create a SRTProject object step by step.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## create count matrix list: note each component has a name, i.e., `MOB`.</span>
<span class="va">cntList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>MOB<span class="op">=</span><span class="va">MOB_raw</span><span class="op">)</span>

<span class="co">## create spatial coordinate matrix </span>
<span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">pos</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">MOB_raw</span><span class="op">)</span>
<span class="va">coordList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">pos</span><span class="op">)</span>

<span class="co">## create metadata list</span>
<span class="va">meta.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>manual_annotation<span class="op">=</span><span class="va">y2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">meta.data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">MOB_raw</span><span class="op">)</span>
<span class="va">metadataList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">meta.data</span><span class="op">)</span>

<span class="co">## create meta data for each data batches. Here we only have one data batch.</span>
<span class="va">sampleMetadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>species<span class="op">=</span><span class="st">"Mouse"</span>, tissues<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Olfactory bulb'</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">sampleMetadata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cntList</span><span class="op">)</span>

<span class="co">## Name of this project</span>
<span class="va">projectName</span> <span class="op">&lt;-</span> <span class="st">"MOB"</span>

<span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">MOB_raw</span><span class="op">)</span></code></pre></div>
<p>Next, we start creating SRTProject object. We can print the basic information of this object, including three parts. The first part have the class of this object, outputPath of data that require to output, h5filePath that save the memory-cusuming data (i.e., count, logcount, …). The second part is about the datasets basic information, such as how many data batches(sample) and the data names, sample meta data (sampleColData) and meta data for each spot (cellMetaData). The last part is about downstream analyses information that is empty when this object created.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CreateSRTProject.html">CreateSRTProject</a></span><span class="op">(</span><span class="va">cntList</span>, <span class="va">coordList</span>, projectName <span class="op">=</span><span class="va">projectName</span>, cellMetaDataList<span class="op">=</span><span class="va">metadataList</span>,
    sampleMetadata<span class="op">=</span><span class="va">sampleMetadata</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Loading required package: rhdf5</span></code></pre>
<pre><code><span class="co">## Loading required package: S4Vectors</span></code></pre>
<pre><code><span class="co">## Loading required package: stats4</span></code></pre>
<pre><code><span class="co">## Loading required package: BiocGenerics</span></code></pre>
<pre><code><span class="co">## </span>
<span class="co">## Attaching package: 'BiocGenerics'</span></code></pre>
<pre><code><span class="co">## The following objects are masked from 'package:stats':</span>
<span class="co">## </span>
<span class="co">##     IQR, mad, sd, var, xtabs</span></code></pre>
<pre><code><span class="co">## The following objects are masked from 'package:base':</span>
<span class="co">## </span>
<span class="co">##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,</span>
<span class="co">##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span>
<span class="co">##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,</span>
<span class="co">##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span>
<span class="co">##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,</span>
<span class="co">##     union, unique, unsplit, which.max, which.min</span></code></pre>
<pre><code><span class="co">## </span>
<span class="co">## Attaching package: 'S4Vectors'</span></code></pre>
<pre><code><span class="co">## The following objects are masked from 'package:base':</span>
<span class="co">## </span>
<span class="co">##     expand.grid, I, unname</span></code></pre>
<pre><code><span class="co">## Warning: package 'ff' was built under R version 4.1.3</span></code></pre>
<pre><code><span class="co">## ****Remove the unshared genes for all batches...</span></code></pre>
<pre><code><span class="co">## ****Filter out 0 genes for 1 this data batch.</span></code></pre>
<pre><code><span class="co">## ****To differ the spots across data batch,</span>
<span class="co">## ****revise the name of each spot in each data batch by adding the name of each data batch</span></code></pre>
<pre><code><span class="co">## ****Filter out 0 spots for 1 data batches, respectively</span></code></pre>
<pre><code><span class="co">## 2023-02-14 09:26:43 : **** Data check and preprocessing, 0.027 mins elapsed.</span></code></pre>
<pre><code><span class="co">## Loading required package: hdf5r</span></code></pre>
<pre><code><span class="co">## </span>
<span class="co">## Attaching package: 'hdf5r'</span></code></pre>
<pre><code><span class="co">## The following object is masked from 'package:S4Vectors':</span>
<span class="co">## </span>
<span class="co">##     values</span></code></pre>
<pre><code><span class="co">## The following object is masked from 'package:rhdf5':</span>
<span class="co">## </span>
<span class="co">##     h5version</span></code></pre>
<pre><code><span class="co">## You created a large dataset with compression and chunking.</span>
<span class="co">## The chunk size is equal to the dataset dimensions.</span>
<span class="co">## If you want to read subsets of the dataset, you should testsmaller chunk sizes to improve read times.</span>
<span class="co">## You created a large dataset with compression and chunking.</span>
<span class="co">## The chunk size is equal to the dataset dimensions.</span>
<span class="co">## If you want to read subsets of the dataset, you should testsmaller chunk sizes to improve read times.</span></code></pre>
<pre><code><span class="co">## 2023-02-14 09:26:45 : **** Write count and meta information into MOB.h5 file, 0.024 mins elapsed.</span></code></pre>
<pre><code><span class="co">## Getting Sample Metadata...</span></code></pre>
<pre><code><span class="co">## Getting Cell Metadata...</span></code></pre>
<pre><code><span class="co">## Getting Spatial Coordinates...</span></code></pre>
<pre><code><span class="co">## 2023-02-14 09:26:45 : **** Read meta information from MOB.h5 file, 0.001 mins elapsed.</span></code></pre>
<pre><code><span class="co">## Initializing SRTProject...</span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span></code></pre></div>
<pre><code><span class="co">##  .oooooo..o ooooooooo.   ooooooooooooo             o8o                       oooo   o8o             </span>
<span class="co">## d8P'    `Y8 `888   `Y88. 8'   888   `8             `''                       `888   `''                      </span>
<span class="co">## Y88bo.       888   .d88'      888      oo.ooooo.  oooo  oo.ooooo.   .ooooo.   888  oooo  ooo. .oo.    .ooooo.</span>
<span class="co">## `'Y8888o.    888ooo88P'       888       888' `88b `888   888' `88b d88' `88b  888  `888  `888P'Y88b  d88' `88b</span>
<span class="co">## ``Y88b888`   88b.             888       888   888  888   888   888 888ooo888  888   888   888   888  888ooo888    </span>
<span class="co">## oo     .d8P  888 `88b.        888       888   888  888   888   888 888    .o  888   888   888   888  888    .o</span>
<span class="co">## 8''88888P'  o888o  o888o     o888o      888bod8P' o888o  888bod8P' `Y8bod8P' o888o o888o o888o o888o `Y8bod8P' </span>
<span class="co">##                                         888              888                                                   </span>
<span class="co">##                                        o888o            o888o</span></code></pre>
<pre><code><span class="co">## class: SRTProject </span>
<span class="co">## outputPath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\MOB </span>
<span class="co">## h5filePath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\MOB/MOB.h5 </span>
<span class="co">## ---------Datasets basic information-----------------</span>
<span class="co">## samples(1): MOB</span>
<span class="co">## sampleColData names(3): species tissues NumOfSpots</span>
<span class="co">## cellMetaData names(2): manual_annotation batch</span>
<span class="co">## numberOfSpots(1): 282</span>
<span class="co">## ---------Downstream analyses information-----------------</span>
<span class="co">## Low-dimensional embeddings(0):</span>
<span class="co">## Inferred cluster labels: No</span>
<span class="co">## Embedding for plotting(0):</span></code></pre>
<p>Then we read the cell-type-specific marker information and construct a cell type marker matrix. Top four DEGs of the scRNA-seq data from Gene Expression Omnibus (GEO;accession number GSE121891) are selected as markers based on log-fold change.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">marker</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="st">"list"</span>, <span class="fl">5</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">marker</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GC"</span>, <span class="st">"PGC"</span>, <span class="st">"M/TC"</span>, <span class="st">"OSNs"</span>, <span class="st">"EPL-IN"</span><span class="op">)</span>
<span class="va">marker</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gria2"</span>, <span class="st">"Meis2"</span>, <span class="st">"Prkca"</span>, <span class="st">"Penk"</span><span class="op">)</span>
<span class="va">marker</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Nppa"</span>, <span class="st">"Nrsn1"</span>, <span class="st">"Nxph1"</span>, <span class="st">"Th"</span><span class="op">)</span>
<span class="va">marker</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cdhr1"</span>, <span class="st">"Slc17a7"</span>, <span class="st">"Olfm1"</span>, <span class="st">"Reln"</span><span class="op">)</span>
<span class="va">marker</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gng13"</span>, <span class="st">"S100a5"</span>, <span class="st">"Omp"</span>, <span class="st">"Fam213b"</span><span class="op">)</span>
<span class="va">marker</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Kit"</span>, <span class="st">"Thy1"</span>,  <span class="st">"Dner"</span>, <span class="st">"Spock2"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">marker</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## $GC</span>
<span class="co">## [1] "Gria2" "Meis2" "Prkca" "Penk" </span>
<span class="co">## </span>
<span class="co">## $PGC</span>
<span class="co">## [1] "Nppa"  "Nrsn1" "Nxph1" "Th"   </span>
<span class="co">## </span>
<span class="co">## $`M/TC`</span>
<span class="co">## [1] "Cdhr1"   "Slc17a7" "Olfm1"   "Reln"   </span>
<span class="co">## </span>
<span class="co">## $OSNs</span>
<span class="co">## [1] "Gng13"   "S100a5"  "Omp"     "Fam213b"</span>
<span class="co">## </span>
<span class="co">## $`EPL-IN`</span>
<span class="co">## [1] "Kit"    "Thy1"   "Dner"   "Spock2"</span></code></pre>
</div>
<div class="section level3">
<h3 id="normalizing-the-data">Normalizing the data<a class="anchor" aria-label="anchor" href="#normalizing-the-data"></a>
</h3>
<p>After creating the SRTProject, the next step is to normalize the data. To save RAM memory, normalized values are stored in disk as a <code>h5file</code>.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeSRT.html">normalizeSRT</a></span><span class="op">(</span><span class="va">SRTProj</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Loading required package: Seurat</span></code></pre>
<pre><code><span class="co">## Warning: package 'Seurat' was built under R version 4.1.3</span></code></pre>
<pre><code><span class="co">## Attaching SeuratObject</span></code></pre>
<pre><code><span class="co">## Attaching sp</span></code></pre>
<pre><code><span class="co">## Loading required package: Matrix</span></code></pre>
<pre><code><span class="co">## </span>
<span class="co">## Attaching package: 'Matrix'</span></code></pre>
<pre><code><span class="co">## The following object is masked from 'package:S4Vectors':</span>
<span class="co">## </span>
<span class="co">##     expand</span></code></pre>
<pre><code><span class="co">## You created a large dataset with compression and chunking.</span>
<span class="co">## The chunk size is equal to the dataset dimensions.</span>
<span class="co">## If you want to read subsets of the dataset, you should testsmaller chunk sizes to improve read times.</span>
<span class="co">## You created a large dataset with compression and chunking.</span>
<span class="co">## The chunk size is equal to the dataset dimensions.</span>
<span class="co">## If you want to read subsets of the dataset, you should testsmaller chunk sizes to improve read times.</span></code></pre>
<pre><code><span class="co">## 2023-02-14 09:26:48 : ****** Finish normalization for sample 1, 0.043 mins elapsed.</span></code></pre>
<div class="section level4">
<h4 id="feature-selection">Feature selection<a class="anchor" aria-label="anchor" href="#feature-selection"></a>
</h4>
<p>We next select a subset of genes that exhibit high spot-to-spot variation in the dataset (i.e, they are highly expressed in some spots, and lowly expressed in others). It has been found that focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets <a href="https://www.nature.com/articles/nmeth.2645" class="external-link">here</a>. In annotaion task, we found the variable features by exluding the marker gene list. The <code>SpatialAnno</code> model will use the selected variable features plus the marker gene list in the moelling process.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Check the features names</span>
<span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">SRTProj</span><span class="op">@</span><span class="va">geneMetaData</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">##  [1] "Plekha1" "Cnih3"   "Gpc5"    "Sirt3"   "Rab8a"   "Nadk"    "Tiparp" </span>
<span class="co">##  [8] "Fbxo28"  "Rgl1"    "Ctsb"</span></code></pre>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectVariableFeatures.html">selectVariableFeatures</a></span><span class="op">(</span><span class="va">SRTProj</span>, nfeatures <span class="op">=</span> <span class="fl">2000</span>, exclude_features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">marker</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Find variable genes for sample MOB</span></code></pre>
<pre><code><span class="co">## 2023-02-14 09:26:52 : ****** Select HVGs(variable features) for each data batch using  vst method, 0.067 mins elapsed.</span></code></pre>
<pre><code><span class="co">## 2023-02-14 09:26:52 : ****** Prioritize HVGs based on the number of times they were selected in all data batches, 0 mins elapsed.</span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">SRTProj</span><span class="op">@</span><span class="va">geneMetaData</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 6 rows and 3 columns</span>
<span class="co">##         isVGs#combine isHVGs#MOB rankHVGs#MOB</span>
<span class="co">##             &lt;logical&gt;  &lt;logical&gt;    &lt;logical&gt;</span>
<span class="co">## Plekha1          TRUE       TRUE           NA</span>
<span class="co">## Cnih3           FALSE      FALSE           NA</span>
<span class="co">## Gpc5            FALSE      FALSE           NA</span>
<span class="co">## Sirt3           FALSE      FALSE           NA</span>
<span class="co">## Rab8a           FALSE      FALSE           NA</span>
<span class="co">## Nadk            FALSE      FALSE           NA</span></code></pre>
</div>
<div class="section level4">
<h4 id="calculate-the-adjcence-matrix">Calculate the adjcence matrix<a class="anchor" aria-label="anchor" href="#calculate-the-adjcence-matrix"></a>
</h4>
<p>Next, we define the neighbors for each spot by the spatial coordinates, which is represented by a sparse adjacency matrix (AdjMat) saved in the h5file. Because here the spatialcoordinates is not integer, we use general method to find the neighbors by setting the <code>platform = 'Other'</code>. We can check the saved sparse adjacency matrix using <code>h5ls()</code>.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">SRTProj</span><span class="op">@</span><span class="va">spatialCoords</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 6 rows and 2 columns</span>
<span class="co">##          V1        V2</span>
<span class="co">##   &lt;numeric&gt; &lt;numeric&gt;</span>
<span class="co">## 1    16.918    16.996</span>
<span class="co">## 2    18.017    17.034</span>
<span class="co">## 3    20.075    17.059</span>
<span class="co">## 4    18.979    17.065</span>
<span class="co">## 5    21.937    16.967</span>
<span class="co">## 6    20.899    17.015</span></code></pre>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Obtain adjacence matrix</span>
<span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddAdj.html">AddAdj</a></span><span class="op">(</span><span class="va">SRTProj</span>, platform <span class="op">=</span> <span class="st">"Other"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Find the adjacency matrix by bisection method...</span></code></pre>
<pre><code><span class="co">## Current radius is 10.87</span></code></pre>
<pre><code><span class="co">## Median of neighborhoods is 2</span></code></pre>
<pre><code><span class="co">## Current radius is 5.93</span></code></pre>
<pre><code><span class="co">## Median of neighborhoods is 184</span></code></pre>
<pre><code><span class="co">## Current radius is 3.47</span></code></pre>
<pre><code><span class="co">## Median of neighborhoods is 78.5</span></code></pre>
<pre><code><span class="co">## Current radius is 2.23</span></code></pre>
<pre><code><span class="co">## Median of neighborhoods is 33</span></code></pre>
<pre><code><span class="co">## Current radius is 1.62</span></code></pre>
<pre><code><span class="co">## Median of neighborhoods is 15</span></code></pre>
<pre><code><span class="co">## Current radius is 1.31</span></code></pre>
<pre><code><span class="co">## Median of neighborhoods is 8</span></code></pre>
<pre><code><span class="co">## Current radius is 1.31</span></code></pre>
<pre><code><span class="co">## Median of neighborhoods is 4</span></code></pre>
<pre><code><span class="co">## 2023-02-14 09:26:52 : ****** Write adjacency matrices to MOB.h5 file, 0.003 mins elapsed.</span></code></pre>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">h5ls</span><span class="op">(</span><span class="va">SRTProj</span><span class="op">@</span><span class="va">projectMetadata</span><span class="op">$</span><span class="va">h5filePath</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##               group           name       otype   dclass     dim</span>
<span class="co">## 0                 /         AdjMat   H5I_GROUP                 </span>
<span class="co">## 1           /AdjMat       AdjMat_1   H5I_GROUP                 </span>
<span class="co">## 2  /AdjMat/AdjMat_1            Dim H5I_DATASET  INTEGER       2</span>
<span class="co">## 3  /AdjMat/AdjMat_1           data H5I_DATASET    FLOAT    1110</span>
<span class="co">## 4  /AdjMat/AdjMat_1        indices H5I_DATASET  INTEGER    1110</span>
<span class="co">## 5  /AdjMat/AdjMat_1         indptr H5I_DATASET  INTEGER     283</span>
<span class="co">## 6                 /  SpatialCoords H5I_DATASET    FLOAT 282 x 2</span>
<span class="co">## 7                 /   cellMetaData H5I_DATASET   STRING     282</span>
<span class="co">## 8                 /          count   H5I_GROUP                 </span>
<span class="co">## 9            /count        count_1   H5I_GROUP                 </span>
<span class="co">## 10   /count/count_1            Dim H5I_DATASET  INTEGER       2</span>
<span class="co">## 11   /count/count_1      cellNames H5I_DATASET   STRING     282</span>
<span class="co">## 12   /count/count_1           data H5I_DATASET    FLOAT 1263280</span>
<span class="co">## 13   /count/count_1      geneNames H5I_DATASET   STRING   16034</span>
<span class="co">## 14   /count/count_1        indices H5I_DATASET  INTEGER 1263280</span>
<span class="co">## 15   /count/count_1         indptr H5I_DATASET  INTEGER     283</span>
<span class="co">## 16                /           data   H5I_GROUP                 </span>
<span class="co">## 17            /data         data_1   H5I_GROUP                 </span>
<span class="co">## 18     /data/data_1            Dim H5I_DATASET  INTEGER       2</span>
<span class="co">## 19     /data/data_1      cellNames H5I_DATASET   STRING     282</span>
<span class="co">## 20     /data/data_1           data H5I_DATASET    FLOAT 1263280</span>
<span class="co">## 21     /data/data_1      geneNames H5I_DATASET   STRING   16034</span>
<span class="co">## 22     /data/data_1        indices H5I_DATASET  INTEGER 1263280</span>
<span class="co">## 23     /data/data_1         indptr H5I_DATASET  INTEGER     283</span>
<span class="co">## 24                / sampleMetadata H5I_DATASET COMPOUND       1</span>
<span class="co">## 25                /    samplenames H5I_DATASET   STRING       1</span></code></pre>
</div>
</div>
</div>
<div class="section level2">
<h2 id="joint-dimension-reduction-and-spatial-annotation-using-spatialanno-model">Joint dimension reduction and spatial annotation using SpatialAnno model<a class="anchor" aria-label="anchor" href="#joint-dimension-reduction-and-spatial-annotation-using-spatialanno-model"></a>
</h2>
<p>We fit the SpatialAnno model by using the top 2000 HVGs. Users can also use the top SVGs since we found the performance of using SVGs or HVGs has no significant difference for the downstream analyses, such as dimension reduction and spatial annotation.</p>
<p>Afeter finishing fitting, the SRTProject <code>SRTProj</code> has two more downstream analyses information: the low-dimensional embeddings <code>SpatialAnno</code> and inferred cell types, and they are saved in <code>SRTProj@reductions$SpatialAnno</code> and <code>SRTProj@clusters</code>, respectively.</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">### Given K</span>
<span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Annotation_SpatialAnno.html">Annotation_SpatialAnno</a></span><span class="op">(</span><span class="va">SRTProj</span>, markers <span class="op">=</span> <span class="va">marker</span>, q<span class="op">=</span><span class="fl">15</span>, wpca_int<span class="op">=</span><span class="cn">F</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## iter = 2, loglik= -752044.469314, dloglik=0.999650 </span>
<span class="co">## iter = 3, loglik= -751469.479497, dloglik=0.000765 </span>
<span class="co">## iter = 4, loglik= -751217.140580, dloglik=0.000336 </span>
<span class="co">## iter = 5, loglik= -751151.965403, dloglik=0.000087 </span>
<span class="co">## iter = 6, loglik= -751113.862506, dloglik=0.000051 </span>
<span class="co">## iter = 7, loglik= -751086.893476, dloglik=0.000036 </span>
<span class="co">## iter = 8, loglik= -751062.423871, dloglik=0.000033 </span>
<span class="co">## iter = 9, loglik= -751043.503935, dloglik=0.000025 </span>
<span class="co">## iter = 10, loglik= -751029.662131, dloglik=0.000018 </span>
<span class="co">## iter = 11, loglik= -751025.199029, dloglik=0.000006</span></code></pre>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span></code></pre></div>
<pre><code><span class="co">##  .oooooo..o ooooooooo.   ooooooooooooo             o8o                       oooo   o8o             </span>
<span class="co">## d8P'    `Y8 `888   `Y88. 8'   888   `8             `''                       `888   `''                      </span>
<span class="co">## Y88bo.       888   .d88'      888      oo.ooooo.  oooo  oo.ooooo.   .ooooo.   888  oooo  ooo. .oo.    .ooooo.</span>
<span class="co">## `'Y8888o.    888ooo88P'       888       888' `88b `888   888' `88b d88' `88b  888  `888  `888P'Y88b  d88' `88b</span>
<span class="co">## ``Y88b888`   88b.             888       888   888  888   888   888 888ooo888  888   888   888   888  888ooo888    </span>
<span class="co">## oo     .d8P  888 `88b.        888       888   888  888   888   888 888    .o  888   888   888   888  888    .o</span>
<span class="co">## 8''88888P'  o888o  o888o     o888o      888bod8P' o888o  888bod8P' `Y8bod8P' o888o o888o o888o o888o `Y8bod8P' </span>
<span class="co">##                                         888              888                                                   </span>
<span class="co">##                                        o888o            o888o</span></code></pre>
<pre><code><span class="co">## class: SRTProject </span>
<span class="co">## outputPath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\MOB </span>
<span class="co">## h5filePath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\MOB/MOB.h5 </span>
<span class="co">## ---------Datasets basic information-----------------</span>
<span class="co">## samples(1): MOB</span>
<span class="co">## sampleColData names(3): species tissues NumOfSpots</span>
<span class="co">## cellMetaData names(2): manual_annotation batch</span>
<span class="co">## numberOfSpots(1): 282</span>
<span class="co">## ---------Downstream analyses information-----------------</span>
<span class="co">## Variable features:  2000 </span>
<span class="co">## Low-dimensional embeddings(1): SpatialAnno</span>
<span class="co">## Inferred cluster labels: Yes</span>
<span class="co">## Embedding for plotting(0):</span></code></pre>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">SRTProj</span><span class="op">@</span><span class="va">clusters</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">##  EPL-IN      GC    M/TC    OSNs     PGC Unknown </span>
<span class="co">##      10      76      98      62      32       4</span></code></pre>
</div>
<div class="section level2">
<h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h2>
<p>Then, we visualize the spatial distribution of inferred cell types that shows the layer structure.</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cols_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="../reference/chooseColors.html">chooseColors</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">5</span><span class="op">)</span>, <span class="st">'gray'</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Loading required package: colorspace</span></code></pre>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/EachClusterSpaHeatMap.html">EachClusterSpaHeatMap</a></span><span class="op">(</span><span class="va">SRTProj</span>, cols<span class="op">=</span><span class="va">cols_cluster</span>, legend.position<span class="op">=</span><span class="st">'right'</span>, pt_size <span class="op">=</span><span class="fl">6</span> ,base_size<span class="op">=</span><span class="fl">16</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Loading required package: dplyr</span></code></pre>
<pre><code><span class="co">## Warning: package 'dplyr' was built under R version 4.1.3</span></code></pre>
<pre><code><span class="co">## </span>
<span class="co">## Attaching package: 'dplyr'</span></code></pre>
<pre><code><span class="co">## The following objects are masked from 'package:S4Vectors':</span>
<span class="co">## </span>
<span class="co">##     first, intersect, rename, setdiff, setequal, union</span></code></pre>
<pre><code><span class="co">## The following objects are masked from 'package:BiocGenerics':</span>
<span class="co">## </span>
<span class="co">##     combine, intersect, setdiff, union</span></code></pre>
<pre><code><span class="co">## The following objects are masked from 'package:stats':</span>
<span class="co">## </span>
<span class="co">##     filter, lag</span></code></pre>
<pre><code><span class="co">## The following objects are masked from 'package:base':</span>
<span class="co">## </span>
<span class="co">##     intersect, setdiff, setequal, union</span></code></pre>
<pre><code><span class="co">## Loading required package: ggplot2</span></code></pre>
<pre><code><span class="co">## Warning: package 'ggplot2' was built under R version 4.1.3</span></code></pre>
<p><img src="single_MOB_Anno_tutorial_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>Calculate the 3-dimensional tSNEs based on the extracted features from SpatialAnno.</p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddTSNE.html">AddTSNE</a></span><span class="op">(</span><span class="va">SRTProj</span>, n_comp <span class="op">=</span> <span class="fl">3</span>, reduction <span class="op">=</span> <span class="st">'SpatialAnno'</span><span class="op">)</span></code></pre></div>
<p>We visualize the 3-dimensional tSNEs on the spatial coordinates using RGB colors. We can see there are apparent spatial pattern for the embeddings. Users can also plot the 3-dimensional UMAP RGB plot similarly.</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/EachRGBSpaHeatMap.html">EachRGBSpaHeatMap</a></span><span class="op">(</span><span class="va">SRTProj</span>, plot_type <span class="op">=</span><span class="st">'tSNE'</span>, title_name <span class="op">=</span> <span class="st">"tSNE RGB plot: "</span>, pt_size<span class="op">=</span><span class="fl">6</span><span class="op">)</span></code></pre></div>
<p><img src="single_MOB_Anno_tutorial_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>Visualize the inferred cell types on the tSNE or UMAP.</p>
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddTSNE.html">AddTSNE</a></span><span class="op">(</span><span class="va">SRTProj</span>, n_comp <span class="op">=</span> <span class="fl">2</span>, reduction<span class="op">=</span><span class="st">'SpatialAnno'</span><span class="op">)</span>
<span class="va">p_tsne2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EmbedPlot.html">EmbedPlot</a></span><span class="op">(</span><span class="va">SRTProj</span>, item<span class="op">=</span><span class="st">'cluster'</span>, plotEmbeddings <span class="op">=</span> <span class="st">'tSNE'</span>, cols<span class="op">=</span><span class="va">cols_cluster</span>, 
                     pt_size<span class="op">=</span><span class="fl">3</span>, legend.position<span class="op">=</span><span class="st">'right'</span><span class="op">)</span>
<span class="va">p_tsne2</span></code></pre></div>
<p><img src="single_MOB_Anno_tutorial_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddUMAP.html">AddUMAP</a></span><span class="op">(</span><span class="va">SRTProj</span>, n_comp <span class="op">=</span> <span class="fl">2</span>, reduction<span class="op">=</span><span class="st">'SpatialAnno'</span><span class="op">)</span>
<span class="va">p_umap2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EmbedPlot.html">EmbedPlot</a></span><span class="op">(</span><span class="va">SRTProj</span>, item<span class="op">=</span><span class="st">'cluster'</span>, plotEmbeddings <span class="op">=</span> <span class="st">'UMAP'</span>, 
                     cols<span class="op">=</span><span class="va">cols_cluster</span>, pt_size<span class="op">=</span><span class="fl">3</span>, legend.position<span class="op">=</span><span class="st">'right'</span><span class="op">)</span>
<span class="va">p_umap2</span></code></pre></div>
<p><img src="single_MOB_Anno_tutorial_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="deg-analysis">DEG analysis<a class="anchor" aria-label="anchor" href="#deg-analysis"></a>
</h2>
<p>To do downstream analyses, we require to get the count data from h5file. The function <code><a href="../reference/getGeneSpotData.html">getGeneSpotData()</a></code> access the gene-by-spot count matrix and other data in <code>SRTProj</code> and then return a <code>SpatialExperiment</code> object.</p>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGeneSpotData.html">getGeneSpotData</a></span><span class="op">(</span><span class="va">SRTProj</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Loading required package: SpatialExperiment</span></code></pre>
<pre><code><span class="co">## Loading required package: SingleCellExperiment</span></code></pre>
<pre><code><span class="co">## Loading required package: SummarizedExperiment</span></code></pre>
<pre><code><span class="co">## Loading required package: MatrixGenerics</span></code></pre>
<pre><code><span class="co">## Loading required package: matrixStats</span></code></pre>
<pre><code><span class="co">## Warning: package 'matrixStats' was built under R version 4.1.3</span></code></pre>
<pre><code><span class="co">## </span>
<span class="co">## Attaching package: 'matrixStats'</span></code></pre>
<pre><code><span class="co">## The following object is masked from 'package:dplyr':</span>
<span class="co">## </span>
<span class="co">##     count</span></code></pre>
<pre><code><span class="co">## </span>
<span class="co">## Attaching package: 'MatrixGenerics'</span></code></pre>
<pre><code><span class="co">## The following objects are masked from 'package:matrixStats':</span>
<span class="co">## </span>
<span class="co">##     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span>
<span class="co">##     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span>
<span class="co">##     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span>
<span class="co">##     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span>
<span class="co">##     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span>
<span class="co">##     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span>
<span class="co">##     colWeightedMeans, colWeightedMedians, colWeightedSds,</span>
<span class="co">##     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span>
<span class="co">##     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span>
<span class="co">##     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span>
<span class="co">##     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span>
<span class="co">##     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span>
<span class="co">##     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span>
<span class="co">##     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span>
<span class="co">##     rowWeightedSds, rowWeightedVars</span></code></pre>
<pre><code><span class="co">## Loading required package: GenomicRanges</span></code></pre>
<pre><code><span class="co">## Loading required package: IRanges</span></code></pre>
<pre><code><span class="co">## </span>
<span class="co">## Attaching package: 'IRanges'</span></code></pre>
<pre><code><span class="co">## The following objects are masked from 'package:dplyr':</span>
<span class="co">## </span>
<span class="co">##     collapse, desc, slice</span></code></pre>
<pre><code><span class="co">## The following object is masked from 'package:sp':</span>
<span class="co">## </span>
<span class="co">##     %over%</span></code></pre>
<pre><code><span class="co">## The following object is masked from 'package:grDevices':</span>
<span class="co">## </span>
<span class="co">##     windows</span></code></pre>
<pre><code><span class="co">## Loading required package: GenomeInfoDb</span></code></pre>
<pre><code><span class="co">## Loading required package: Biobase</span></code></pre>
<pre><code><span class="co">## Welcome to Bioconductor</span>
<span class="co">## </span>
<span class="co">##     Vignettes contain introductory material; view with</span>
<span class="co">##     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="co">##     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></code></pre>
<pre><code><span class="co">## </span>
<span class="co">## Attaching package: 'Biobase'</span></code></pre>
<pre><code><span class="co">## The following object is masked from 'package:MatrixGenerics':</span>
<span class="co">## </span>
<span class="co">##     rowMedians</span></code></pre>
<pre><code><span class="co">## The following objects are masked from 'package:matrixStats':</span>
<span class="co">## </span>
<span class="co">##     anyMissing, rowMedians</span></code></pre>
<pre><code><span class="co">## </span>
<span class="co">## Attaching package: 'SummarizedExperiment'</span></code></pre>
<pre><code><span class="co">## The following object is masked from 'package:SeuratObject':</span>
<span class="co">## </span>
<span class="co">##     Assays</span></code></pre>
<pre><code><span class="co">## The following object is masked from 'package:Seurat':</span>
<span class="co">## </span>
<span class="co">##     Assays</span></code></pre>
<div class="sourceCode" id="cb129"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">spe</span></code></pre></div>
<pre><code><span class="co">## class: SpatialExperiment </span>
<span class="co">## dim: 16034 282 </span>
<span class="co">## metadata(0):</span>
<span class="co">## assays(2): counts logcounts</span>
<span class="co">## rownames(16034): Plekha1 Cnih3 ... St14 Gdf5</span>
<span class="co">## rowData names(3): isVGs.combine isHVGs.MOB rankHVGs.MOB</span>
<span class="co">## colnames(282): MOB#16.918x16.996 MOB#18.017x17.034 ...</span>
<span class="co">##   MOB#25.134x28.934 MOB#29.961x18.97</span>
<span class="co">## colData names(4): manual_annotation batch clusters sample_id</span>
<span class="co">## reducedDimNames(5): SpatialAnno tSNE3 tSNE UMAP Coord</span>
<span class="co">## mainExpName: NULL</span>
<span class="co">## altExpNames(0):</span>
<span class="co">## spatialData names(0) :</span>
<span class="co">## spatialCoords names(2) : V1 V2</span>
<span class="co">## imgData names(0):</span></code></pre>
<p>We perform differential expression analysis for all clusters by using <code><a href="../reference/FindAllDEGs.html">FindAllDEGs()</a></code> function, then the DE genes’ information is saved in a <code>DataFrame</code> object <code>dat_degs</code>.</p>
<div class="sourceCode" id="cb131"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dat_degs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindAllDEGs.html">FindAllDEGs</a></span><span class="op">(</span><span class="va">spe</span>, use_cluster <span class="op">=</span> <span class="st">'clusters'</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## 2023-02-14 09:27:10 : ****** Start finding marker genes...</span></code></pre>
<pre><code><span class="co">## Calculating cluster EPL-IN</span></code></pre>
<pre><code><span class="co">## Calculating cluster GC</span></code></pre>
<pre><code><span class="co">## Calculating cluster M/TC</span></code></pre>
<pre><code><span class="co">## Calculating cluster OSNs</span></code></pre>
<pre><code><span class="co">## Calculating cluster PGC</span></code></pre>
<pre><code><span class="co">## Calculating cluster Unknown</span></code></pre>
<pre><code><span class="co">## 2023-02-14 09:27:46 : ****** Finish finding marker genes, 0.603 mins elapsed.</span></code></pre>
<div class="sourceCode" id="cb140"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dat_degs</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 3347 rows and 7 columns</span>
<span class="co">##               p_val avg_log2FC     pct.1     pct.2   p_val_adj  cluster</span>
<span class="co">##           &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;factor&gt;</span>
<span class="co">## Tspan8  1.92140e-20   0.413781       0.4     0.004 3.08077e-16   EPL-IN</span>
<span class="co">## Clec3b  7.67177e-16   0.485016       0.5     0.018 1.23009e-11   EPL-IN</span>
<span class="co">## Chek1   1.64141e-13   0.252318       0.2     0.000 2.63184e-09   EPL-IN</span>
<span class="co">## Fgfbp1  8.65072e-12   0.567707       0.6     0.051 1.38706e-07   EPL-IN</span>
<span class="co">## Smpdl3b 4.86295e-10   0.295957       0.3     0.011 7.79726e-06   EPL-IN</span>
<span class="co">## ...             ...        ...       ...       ...         ...      ...</span>
<span class="co">## Hectd1   0.00994409  -0.821458         0     0.745           1  Unknown</span>
<span class="co">## Fgfr1    0.00994409  -0.825079         0     0.745           1  Unknown</span>
<span class="co">## Zfp704   0.00994409  -0.828357         0     0.745           1  Unknown</span>
<span class="co">## Cnot7    0.00994409  -0.844428         0     0.745           1  Unknown</span>
<span class="co">## Dio2.3   0.00994409  -0.866538         0     0.745           1  Unknown</span>
<span class="co">##                gene</span>
<span class="co">##         &lt;character&gt;</span>
<span class="co">## Tspan8       Tspan8</span>
<span class="co">## Clec3b       Clec3b</span>
<span class="co">## Chek1         Chek1</span>
<span class="co">## Fgfbp1       Fgfbp1</span>
<span class="co">## Smpdl3b     Smpdl3b</span>
<span class="co">## ...             ...</span>
<span class="co">## Hectd1       Hectd1</span>
<span class="co">## Fgfr1         Fgfr1</span>
<span class="co">## Zfp704       Zfp704</span>
<span class="co">## Cnot7         Cnot7</span>
<span class="co">## Dio2.3         Dio2</span></code></pre>
<p>We identify the significant DE genes by two criteria: (a) adjustd p-value less than 0.01 and (b) average log fold change greater than 0.4.</p>
<div class="sourceCode" id="cb142"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">degs_sig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">dat_degs</span>, <span class="va">p_val_adj</span> <span class="op">&lt;</span> <span class="fl">0.01</span> <span class="op">&amp;</span> <span class="va">avg_log2FC</span><span class="op">&gt;</span><span class="fl">0.5</span><span class="op">)</span>
<span class="va">degs_sig</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 118 rows and 7 columns</span>
<span class="co">##                  p_val avg_log2FC     pct.1     pct.2   p_val_adj  cluster</span>
<span class="co">##              &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;factor&gt;</span>
<span class="co">## Fgfbp1     8.65072e-12   0.567707       0.6     0.051 1.38706e-07   EPL-IN</span>
<span class="co">## Fmod       5.88749e-09   0.850051       0.9     0.217 9.44001e-05   EPL-IN</span>
<span class="co">## Prss56     2.14120e-08   0.618173       0.8     0.151 3.43320e-04   EPL-IN</span>
<span class="co">## Aebp1      2.17723e-08   0.801152       0.9     0.250 3.49096e-04   EPL-IN</span>
<span class="co">## Myof       2.97383e-08   0.517158       0.6     0.081 4.76824e-04   EPL-IN</span>
<span class="co">## ...                ...        ...       ...       ...         ...      ...</span>
<span class="co">## Nxph1.1    2.09285e-14   0.613144     0.719     0.172 3.35567e-10      PGC</span>
<span class="co">## Pip5k1b    2.38307e-14   0.535098     0.656     0.120 3.82101e-10      PGC</span>
<span class="co">## AI593442.1 4.78346e-13   0.686870     0.906     0.412 7.66980e-09      PGC</span>
<span class="co">## Pde5a.1    3.60833e-11   0.583596     0.812     0.312 5.78560e-07      PGC</span>
<span class="co">## Apold1.1   1.57576e-09   0.575961     0.938     0.548 2.52658e-05      PGC</span>
<span class="co">##                   gene</span>
<span class="co">##            &lt;character&gt;</span>
<span class="co">## Fgfbp1          Fgfbp1</span>
<span class="co">## Fmod              Fmod</span>
<span class="co">## Prss56          Prss56</span>
<span class="co">## Aebp1            Aebp1</span>
<span class="co">## Myof              Myof</span>
<span class="co">## ...                ...</span>
<span class="co">## Nxph1.1          Nxph1</span>
<span class="co">## Pip5k1b        Pip5k1b</span>
<span class="co">## AI593442.1    AI593442</span>
<span class="co">## Pde5a.1          Pde5a</span>
<span class="co">## Apold1.1        Apold1</span></code></pre>
<p>We visualize the DE genes for each cell type group by gene-by-cell heatmap using the <code><a href="../reference/GCHeatMap.html">GCHeatMap()</a></code> function.</p>
<div class="sourceCode" id="cb144"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>
<span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">5</span>

<span class="va">dat_degs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">as.data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html" class="external-link">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n</span>, wt <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">topGene</span>

<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GCHeatMap.html">GCHeatMap</a></span><span class="op">(</span><span class="va">spe</span>, features<span class="op">=</span><span class="va">topGene</span><span class="op">$</span><span class="va">gene</span>, grp_color<span class="op">=</span><span class="va">cols_cluster</span> ,y_text_size<span class="op">=</span><span class="fl">6</span>, ncol.legend<span class="op">=</span><span class="fl">1</span><span class="op">)</span>
<span class="va">p1</span></code></pre></div>
<p><img src="single_MOB_Anno_tutorial_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="other-downstream-analyses">Other downstream analyses<a class="anchor" aria-label="anchor" href="#other-downstream-analyses"></a>
</h2>
<details><p><summary><strong>Session Info</strong></summary></p>
<div class="sourceCode" id="cb145"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.1.2 (2021-11-01)</span>
<span class="co">## Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">## Running under: Windows 10 x64 (build 22621)</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">## [1] LC_COLLATE=Chinese (Simplified)_China.936 </span>
<span class="co">## [2] LC_CTYPE=Chinese (Simplified)_China.936   </span>
<span class="co">## [3] LC_MONETARY=Chinese (Simplified)_China.936</span>
<span class="co">## [4] LC_NUMERIC=C                              </span>
<span class="co">## [5] LC_TIME=Chinese (Simplified)_China.936    </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">## [8] base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] SpatialExperiment_1.4.0     SingleCellExperiment_1.16.0</span>
<span class="co">##  [3] SummarizedExperiment_1.24.0 Biobase_2.54.0             </span>
<span class="co">##  [5] GenomicRanges_1.46.1        GenomeInfoDb_1.30.1        </span>
<span class="co">##  [7] IRanges_2.28.0              MatrixGenerics_1.6.0       </span>
<span class="co">##  [9] matrixStats_0.62.0          ggplot2_3.3.6              </span>
<span class="co">## [11] dplyr_1.0.9                 colorspace_2.0-3           </span>
<span class="co">## [13] Matrix_1.4-0                sp_1.5-0                   </span>
<span class="co">## [15] SeuratObject_4.1.0          Seurat_4.1.1               </span>
<span class="co">## [17] hdf5r_1.3.5                 ff_4.0.7                   </span>
<span class="co">## [19] bit_4.0.4                   S4Vectors_0.32.3           </span>
<span class="co">## [21] BiocGenerics_0.40.0         rhdf5_2.38.0               </span>
<span class="co">## [23] SpatialAnno_1.0.0           mclust_5.4.10              </span>
<span class="co">## [25] SRTpipeline_0.1.1          </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] scattermore_0.8           ggthemes_4.2.4           </span>
<span class="co">##   [3] R.methodsS3_1.8.1         GiRaF_1.0.1              </span>
<span class="co">##   [5] ragg_1.2.2                tidyr_1.2.0              </span>
<span class="co">##   [7] bit64_4.0.5               knitr_1.37               </span>
<span class="co">##   [9] irlba_2.3.5               DelayedArray_0.20.0      </span>
<span class="co">##  [11] R.utils_2.11.0            data.table_1.14.2        </span>
<span class="co">##  [13] rpart_4.1.16              RCurl_1.98-1.6           </span>
<span class="co">##  [15] generics_0.1.2            ScaledMatrix_1.2.0       </span>
<span class="co">##  [17] cowplot_1.1.1             RANN_2.6.1               </span>
<span class="co">##  [19] future_1.26.1             spatstat.data_3.0-0      </span>
<span class="co">##  [21] httpuv_1.6.5              assertthat_0.2.1         </span>
<span class="co">##  [23] viridis_0.6.2             xfun_0.29                </span>
<span class="co">##  [25] jquerylib_0.1.4           evaluate_0.15            </span>
<span class="co">##  [27] promises_1.2.0.1          fansi_1.0.3              </span>
<span class="co">##  [29] igraph_1.3.5              DBI_1.1.2                </span>
<span class="co">##  [31] htmlwidgets_1.5.4         spatstat.geom_2.4-0      </span>
<span class="co">##  [33] purrr_0.3.4               ellipsis_0.3.2           </span>
<span class="co">##  [35] RSpectra_0.16-1           ggpubr_0.4.0             </span>
<span class="co">##  [37] backports_1.4.1           DR.SC_3.1                </span>
<span class="co">##  [39] deldir_1.0-6              sparseMatrixStats_1.6.0  </span>
<span class="co">##  [41] vctrs_0.4.1               ROCR_1.0-11              </span>
<span class="co">##  [43] abind_1.4-5               cachem_1.0.6             </span>
<span class="co">##  [45] withr_2.5.0               PRECAST_1.4              </span>
<span class="co">##  [47] progressr_0.10.1          sctransform_0.3.3        </span>
<span class="co">##  [49] goftest_1.2-3             cluster_2.1.2            </span>
<span class="co">##  [51] lazyeval_0.2.2            crayon_1.5.1             </span>
<span class="co">##  [53] edgeR_3.36.0              pkgconfig_2.0.3          </span>
<span class="co">##  [55] labeling_0.4.2            nlme_3.1-155             </span>
<span class="co">##  [57] vipor_0.4.5               rlang_1.0.2              </span>
<span class="co">##  [59] globals_0.15.0            lifecycle_1.0.1          </span>
<span class="co">##  [61] miniUI_0.1.1.1            rsvd_1.0.5               </span>
<span class="co">##  [63] rprojroot_2.0.3           polyclip_1.10-0          </span>
<span class="co">##  [65] lmtest_0.9-40             carData_3.0-5            </span>
<span class="co">##  [67] Rhdf5lib_1.16.0           zoo_1.8-10               </span>
<span class="co">##  [69] beeswarm_0.4.0            ggridges_0.5.3           </span>
<span class="co">##  [71] rjson_0.2.21              png_0.1-7                </span>
<span class="co">##  [73] viridisLite_0.4.0         bitops_1.0-7             </span>
<span class="co">##  [75] R.oo_1.24.0               KernSmooth_2.23-20       </span>
<span class="co">##  [77] rhdf5filters_1.6.0        DelayedMatrixStats_1.16.0</span>
<span class="co">##  [79] stringr_1.4.0             parallelly_1.32.0        </span>
<span class="co">##  [81] spatstat.random_2.2-0     rstatix_0.7.0            </span>
<span class="co">##  [83] ggsignif_0.6.3            beachmat_2.10.0          </span>
<span class="co">##  [85] scales_1.2.0              memoise_2.0.1            </span>
<span class="co">##  [87] magrittr_2.0.3            plyr_1.8.7               </span>
<span class="co">##  [89] ica_1.0-2                 zlibbioc_1.40.0          </span>
<span class="co">##  [91] compiler_4.1.2            dqrng_0.3.0              </span>
<span class="co">##  [93] scSorter_0.0.2            RColorBrewer_1.1-3       </span>
<span class="co">##  [95] fitdistrplus_1.1-8        cli_3.2.0                </span>
<span class="co">##  [97] XVector_0.34.0            listenv_0.8.0            </span>
<span class="co">##  [99] patchwork_1.1.1           pbapply_1.5-0            </span>
<span class="co">## [101] MASS_7.3-55               mgcv_1.8-39              </span>
<span class="co">## [103] tidyselect_1.1.2          stringi_1.7.6            </span>
<span class="co">## [105] textshaping_0.3.6         highr_0.9                </span>
<span class="co">## [107] yaml_2.3.6                BiocSingular_1.10.0      </span>
<span class="co">## [109] locfit_1.5-9.4            ggrepel_0.9.1            </span>
<span class="co">## [111] grid_4.1.2                sass_0.4.1               </span>
<span class="co">## [113] tools_4.1.2               future.apply_1.9.0       </span>
<span class="co">## [115] parallel_4.1.2            rstudioapi_0.13          </span>
<span class="co">## [117] gridExtra_2.3             farver_2.1.0             </span>
<span class="co">## [119] Rtsne_0.16                DropletUtils_1.14.2      </span>
<span class="co">## [121] digest_0.6.29             rgeos_0.5-9              </span>
<span class="co">## [123] FNN_1.1.3.1               shiny_1.7.1              </span>
<span class="co">## [125] Rcpp_1.0.10               car_3.0-12               </span>
<span class="co">## [127] broom_0.7.12              scuttle_1.4.0            </span>
<span class="co">## [129] later_1.3.0               RcppAnnoy_0.0.19         </span>
<span class="co">## [131] httr_1.4.3                fs_1.5.2                 </span>
<span class="co">## [133] tensor_1.5                reticulate_1.25          </span>
<span class="co">## [135] splines_4.1.2             uwot_0.1.11              </span>
<span class="co">## [137] spatstat.utils_3.0-1      pkgdown_2.0.6            </span>
<span class="co">## [139] scater_1.25.1             plotly_4.10.0            </span>
<span class="co">## [141] systemfonts_1.0.4         xtable_1.8-4             </span>
<span class="co">## [143] jsonlite_1.8.0            R6_2.5.1                 </span>
<span class="co">## [145] pillar_1.7.0              htmltools_0.5.2          </span>
<span class="co">## [147] mime_0.12                 glue_1.6.2               </span>
<span class="co">## [149] fastmap_1.1.0             BiocParallel_1.28.3      </span>
<span class="co">## [151] BiocNeighbors_1.12.0      codetools_0.2-18         </span>
<span class="co">## [153] utf8_1.2.2                lattice_0.20-45          </span>
<span class="co">## [155] bslib_0.3.1               spatstat.sparse_2.1-1    </span>
<span class="co">## [157] tibble_3.1.7              ggbeeswarm_0.6.0         </span>
<span class="co">## [159] leiden_0.4.2              gtools_3.9.2.2           </span>
<span class="co">## [161] magick_2.7.3              survival_3.2-13          </span>
<span class="co">## [163] limma_3.50.1              CompQuadForm_1.4.3       </span>
<span class="co">## [165] rmarkdown_2.11            desc_1.4.0               </span>
<span class="co">## [167] munsell_0.5.0             GenomeInfoDbData_1.2.7   </span>
<span class="co">## [169] HDF5Array_1.22.1          reshape2_1.4.4           </span>
<span class="co">## [171] gtable_0.3.0              spatstat.core_2.4-4</span></code></pre>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Wei Liu, Yi Yang, Xu Liao, Xiao Zhang, Jin Liu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
