<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SRT data clustering • SRTpipeline</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="SRT data clustering">
<meta property="og:description" content="SRTpipeline">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SRTpipeline</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../articles/Installation.html">
    <span class="fas fa-book"></span>
     
    Install
  </a>
</li>
<li>
  <a href="../articles/get_started.html">
    <span class="fas fa-book"></span>
     
    Get Started
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fas fa-book"></span>
     
    Single Data Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/single_basic_tutorial.html">Basic tutorial</a>
    </li>
    <li>
      <a href="../articles/dlpfc_tutorial.html">Embedding and clustering introduction</a>
    </li>
    <li>
      <a href="../articles/single_Embryo_SC_tutorial.html">SRT data clustering</a>
    </li>
    <li>
      <a href="../articles/single_Embryo_DRSC_tutorial.html">SRT data embedding and clustering</a>
    </li>
    <li>
      <a href="../articles/single_MOB_Anno_tutorial.html">SRT data embedding and annotation</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fas fa-book"></span>
     
    Multiple Data Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/multibatch_basic_tutorial.html">Basic tutorial</a>
    </li>
    <li>
      <a href="../articles/integration_introduction.html">Basic tutorial v2</a>
    </li>
    <li>
      <a href="../articles/multibatch_brain12_tutorial.html">SRT data clustering and integration</a>
    </li>
    <li>
      <a href="../articles/multibatch_HCC4_tutorial.html">SRT data embedding, clustering and integration example 1</a>
    </li>
    <li>
      <a href="../articles/multibatch_liver8_tutorial.html">SRT data embedding, clustering and integration example 2</a>
    </li>
    <li>
      <a href="../articles/multibatch_MOB16_tutorial.html">SRT data embedding, clustering and integration example 3</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/faq.html">
    <span class="fa fa-question-circle-o"></span>
     
    FAQ
  </a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>SRT data clustering</h1>
            
            <h4 data-toc-skip class="date">Compiled: 二月 14, 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/feiyoung/SRTpipeline/blob/HEAD/vignettes/single_Embryo_SC_tutorial.Rmd"><code>vignettes/single_Embryo_SC_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>single_Embryo_SC_tutorial.Rmd</code></div>

    </div>

    
    
<hr>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This tutorial demonstrates how to use <code>SRTpipeline</code> (&gt;=0.1.0) to analyze a spatially-resolved transcriptomics (SRT) data from seqFISH platform. The analytical pipelines are similar to the <code>SRTpipeline</code> workflow for Visium SRT analysis. We emphases how to use SC-MEB model for spatial clustering and its followed applications. This tutorial will cover the following tasks, which we believe will be common for many spatial analyses:</p>
<ul>
<li>Normalization</li>
<li>Dimension reduction</li>
<li>Spatial clustering</li>
<li>Detecting spatially-variable features</li>
<li>DEG analysis</li>
<li>Spatial trajectory inference</li>
<li>Visualization</li>
</ul>
<p>First, we load <code>SRTpipeline</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feiyoung/SRTpipeline">SRTpipeline</a></span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dataset">Dataset<a class="anchor" aria-label="anchor" href="#dataset"></a>
</h2>
<p>For this tutorial, we will introduce how to create a <code>SRTProject</code> object with single SRT sample using SRTpipeline that includes an introduction to common analytical workflows for a single data batch. Here, we will be taking a spatial transcriptomics dataset for mouse embryo as an example. There are 23194 spots and 351 genes that were sequenced on the seqFISH platform. Our preprocessed data can be downloaded <a href="https://github.com/feiyoung/PRECAST/blob/main/vignettes_data/seu_em3.rda?raw=true" class="external-link">here</a>.</p>
<p>We download the data to the current working path for the followed analysis by the following command:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">githubURL</span> <span class="op">&lt;-</span> <span class="st">"https://github.com/feiyoung/PRECAST/blob/main/vignettes_data/Embryo3_seqFISH.rds?raw=true"</span>
<span class="co"># download.file(githubURL,'Embryo3_seqFISH.rds',mode='wb')</span></code></pre></div>
<p>Then load to R. The data is saved in a Seurat object named <code>Embryo3_seqFISH.rds</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"Embryo3_seqFISH.rds"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="pre-processing-workflow">Pre-processing workflow<a class="anchor" aria-label="anchor" href="#pre-processing-workflow"></a>
</h2>
<div class="section level3">
<h3 id="create-a-srtproject-object">Create a SRTProject object<a class="anchor" aria-label="anchor" href="#create-a-srtproject-object"></a>
</h3>
<p>First, we show how to create a SRTProject object step by step.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## create count matrix list: note each component has a name, i.e., `embryo3`.</span>
<span class="va">count_matrix</span> <span class="op">&lt;-</span> <span class="va">seu_em3</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">counts</span>
<span class="va">cntList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>embryo3 <span class="op">=</span> <span class="va">count_matrix</span><span class="op">)</span>

<span class="co">## create spatial coordinate matrix</span>
<span class="va">coordList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">seu_em3</span><span class="op">$</span><span class="va">x_global</span>, <span class="va">seu_em3</span><span class="op">$</span><span class="va">y_global</span><span class="op">)</span><span class="op">)</span>

<span class="co">## create metadata list</span>
<span class="va">meta.data</span> <span class="op">&lt;-</span> <span class="va">seu_em3</span><span class="op">@</span><span class="va">meta.data</span>
<span class="va">metadataList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">meta.data</span><span class="op">)</span>

<span class="co">## create meta data for each data batches. Here we only have one data batch.</span>
<span class="va">sampleMetadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">"Mouse"</span>, tissues <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Embryo"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">sampleMetadata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cntList</span><span class="op">)</span>

<span class="co">## Name of this project</span>
<span class="va">projectName</span> <span class="op">&lt;-</span> <span class="st">"Embryo3"</span>

<span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">seu_em3</span><span class="op">)</span></code></pre></div>
<p>Next, we start creating SRTProject object. We can print the basic information of this object, including three parts. The first part have the class of this object, outputPath of data that require to output, h5filePath that save the memory-cusuming data (i.e., count, logcount, …). The second part is about the datasets basic information, such as how many data batches(sample) and the data names, sample meta data (sampleColData) and meta data for each spot (cellMetaData). The last part is about downstream analyses information that is empty when this object created.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CreateSRTProject.html">CreateSRTProject</a></span><span class="op">(</span><span class="va">cntList</span>, <span class="va">coordList</span>, projectName <span class="op">=</span> <span class="va">projectName</span>, <span class="va">metadataList</span>, <span class="va">sampleMetadata</span><span class="op">)</span>
<span class="va">SRTProj</span></code></pre></div>
<pre><code><span class="co">## class: SRTProject </span>
<span class="co">## outputPath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\Embryo3 </span>
<span class="co">## h5filePath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\Embryo3/Embryo3.h5 </span>
<span class="co">## ---------Datasets basic information-----------------</span>
<span class="co">## samples(1): embryo3</span>
<span class="co">## sampleColData names(3): species tissues NumOfSpots</span>
<span class="co">## cellMetaData names(18): orig.ident nCount_RNA ...</span>
<span class="co">##   celltype_mapped_refined batch</span>
<span class="co">## numberOfSpots(1): 23194</span>
<span class="co">## ---------Downstream analyses information-----------------</span>
<span class="co">## Low-dimensional embeddings(0):</span>
<span class="co">## Inferred cluster labels: No</span>
<span class="co">## Embedding for plotting(0):</span></code></pre>
</div>
<div class="section level3">
<h3 id="normalizing-the-data">Normalizing the data<a class="anchor" aria-label="anchor" href="#normalizing-the-data"></a>
</h3>
<p>After creating the SRTProject, the next step is to normalize the data. To save RAM memory, normalized values are stored in disk as a <code>h5file</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeSRT.html">normalizeSRT</a></span><span class="op">(</span><span class="va">SRTProj</span>, force <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></code></pre></div>
<p>Because seqFISH technology is based on target-genes, we do not select genes but use all the targeted genes by the argument <code>use_custom_features</code>. Then the custom features will be regarded as variable features for the followed analyses.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">geneNames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">SRTProj</span><span class="op">@</span><span class="va">geneMetaData</span><span class="op">)</span>
<span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectVariableFeatures.html">selectVariableFeatures</a></span><span class="op">(</span><span class="va">SRTProj</span>, use_custom_features <span class="op">=</span> <span class="va">geneNames</span><span class="op">)</span>
<span class="va">SRTProj</span></code></pre></div>
<pre><code><span class="co">## class: SRTProject </span>
<span class="co">## outputPath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\Embryo3 </span>
<span class="co">## h5filePath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\Embryo3/Embryo3.h5 </span>
<span class="co">## ---------Datasets basic information-----------------</span>
<span class="co">## samples(1): embryo3</span>
<span class="co">## sampleColData names(3): species tissues NumOfSpots</span>
<span class="co">## cellMetaData names(18): orig.ident nCount_RNA ...</span>
<span class="co">##   celltype_mapped_refined batch</span>
<span class="co">## numberOfSpots(1): 23194</span>
<span class="co">## ---------Downstream analyses information-----------------</span>
<span class="co">## Variable features:  351 </span>
<span class="co">## Low-dimensional embeddings(0):</span>
<span class="co">## Inferred cluster labels: No</span>
<span class="co">## Embedding for plotting(0):</span></code></pre>
<p>Calculate the adjcence matrix</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Obtain adjacence matrix</span>
<span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddAdj.html">AddAdj</a></span><span class="op">(</span><span class="va">SRTProj</span>, platform <span class="op">=</span> <span class="st">"Other"</span>, radius.upper <span class="op">=</span> <span class="fl">12</span><span class="op">)</span></code></pre></div>
<p>Visualize</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">selectFeatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">SRTProj</span><span class="op">@</span><span class="va">geneMetaData</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="fu"><a href="../reference/EachExprSpaHeatMap.html">EachExprSpaHeatMap</a></span><span class="op">(</span><span class="va">SRTProj</span>, features <span class="op">=</span> <span class="va">selectFeatures</span>, title_name <span class="op">=</span> <span class="cn">T</span>, quantVec <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.9</span>, <span class="fl">0.9</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="single_Embryo_SC_tutorial_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="dimension-reduction">Dimension reduction<a class="anchor" aria-label="anchor" href="#dimension-reduction"></a>
</h2>
<p>PCA is the most popular dimension reduction technique in single cell RNA sequencing (scRNA-seq) data and SRT　data because of its simplicity, computational efficiency, and relatively comparable performance. SRTpipeline provided three versions of PCA: standard PCA (PCA), approximated PCA (APCA), and weighted PCA (WPCA) by the function <code>AddPCA</code> with default version as APCA for fastest computation. The 15-dimensional PCs are extracted by default, and users can use their own values.</p>
<p>After running <code>AddPCA</code>, we would see the output of <code>SRTProj</code> includes <code>PCA</code> in the <code>Low-dimensional embeddings</code> field.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## APCA</span>
<span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddPCA.html">AddPCA</a></span><span class="op">(</span><span class="va">SRTProj</span>, n_comp <span class="op">=</span> <span class="fl">15</span><span class="op">)</span>
<span class="co"># accurate PCA</span>

<span class="co"># SRTProj &lt;- AddPCA(SRTProj, Method='PCA')</span>

<span class="co"># weighted PCA</span>

<span class="co"># SRTProj &lt;- AddPCA(SRTProj, Method='WPCA')</span>
<span class="va">SRTProj</span></code></pre></div>
<pre><code><span class="co">## class: SRTProject </span>
<span class="co">## outputPath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\Embryo3 </span>
<span class="co">## h5filePath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\Embryo3/Embryo3.h5 </span>
<span class="co">## ---------Datasets basic information-----------------</span>
<span class="co">## samples(1): embryo3</span>
<span class="co">## sampleColData names(3): species tissues NumOfSpots</span>
<span class="co">## cellMetaData names(18): orig.ident nCount_RNA ...</span>
<span class="co">##   celltype_mapped_refined batch</span>
<span class="co">## numberOfSpots(1): 23194</span>
<span class="co">## ---------Downstream analyses information-----------------</span>
<span class="co">## Variable features:  351 </span>
<span class="co">## Low-dimensional embeddings(1): PCA</span>
<span class="co">## Inferred cluster labels: No</span>
<span class="co">## Embedding for plotting(0):</span></code></pre>
</div>
<div class="section level2">
<h2 id="clustering-using-sc-meb">Clustering using SC-MEB<a class="anchor" aria-label="anchor" href="#clustering-using-sc-meb"></a>
</h2>
<p>Some SRT clustering methods use Markov random field to model clusters of spots, such as SC-MEB and DR-SC. These approaches work extremely well for data with spatial smoothness and are standard practice in SRT data. For this reason, SRTpipeline uses existing state-of-the-art clustering methods from SRT packages for clustering.</p>
<p>We have had the most success using the Markov random field model implemented by SC-MEB. In SRTpipeline, SC-MEB clustering is performed using the <code><a href="../reference/Cluster_SCMEB.html">Cluster_SCMEB()</a></code> function.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Cluster_SCMEB.html">Cluster_SCMEB</a></span><span class="op">(</span><span class="va">SRTProj</span>, K <span class="op">=</span> <span class="fl">25</span>, reduction <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## diff Energy = 65.622244</span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span></code></pre></div>
<pre><code><span class="co">## class: SRTProject </span>
<span class="co">## outputPath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\Embryo3 </span>
<span class="co">## h5filePath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\Embryo3/Embryo3.h5 </span>
<span class="co">## ---------Datasets basic information-----------------</span>
<span class="co">## samples(1): embryo3</span>
<span class="co">## sampleColData names(3): species tissues NumOfSpots</span>
<span class="co">## cellMetaData names(18): orig.ident nCount_RNA ...</span>
<span class="co">##   celltype_mapped_refined batch</span>
<span class="co">## numberOfSpots(1): 23194</span>
<span class="co">## ---------Downstream analyses information-----------------</span>
<span class="co">## Variable features:  351 </span>
<span class="co">## Low-dimensional embeddings(1): PCA</span>
<span class="co">## Inferred cluster labels: Yes</span>
<span class="co">## Embedding for plotting(0):</span></code></pre>
<div class="section level3">
<h3 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h3>
<p>To run tSNE in SRTpipeline we use the AddTSNE() function.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddTSNE.html">AddTSNE</a></span><span class="op">(</span><span class="va">SRTProj</span>, n_comp <span class="op">=</span> <span class="fl">2</span>, reduction <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span>
<span class="va">cols_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="../reference/chooseColors.html">chooseColors</a></span><span class="op">(</span>palettes_name <span class="op">=</span> <span class="st">"Blink 23"</span>, n_colors <span class="op">=</span> <span class="fl">23</span><span class="op">)</span>, <span class="st">"red4"</span>, <span class="st">"green4"</span><span class="op">)</span>
<span class="va">p_pca_tsne2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EmbedPlot.html">EmbedPlot</a></span><span class="op">(</span><span class="va">SRTProj</span>, item <span class="op">=</span> <span class="st">"cluster"</span>, plotEmbeddings <span class="op">=</span> <span class="st">"tSNE"</span>, cols <span class="op">=</span> <span class="va">cols_cluster</span>,
    legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></code></pre></div>
<p>To run UMAP in SRTpipeline we use the AddUMAP() function.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddUMAP.html">AddUMAP</a></span><span class="op">(</span><span class="va">SRTProj</span>, n_comp <span class="op">=</span> <span class="fl">2</span>, reduction <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span>
<span class="va">p_pca_umap2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EmbedPlot.html">EmbedPlot</a></span><span class="op">(</span><span class="va">SRTProj</span>, item <span class="op">=</span> <span class="st">"cluster"</span>, plotEmbeddings <span class="op">=</span> <span class="st">"UMAP"</span>, cols <span class="op">=</span> <span class="va">cols_cluster</span>,
    legend.position <span class="op">=</span> <span class="st">"right"</span>, axis_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"UMAP1"</span>, <span class="st">"UMAP2"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Next, we merge all plots into one figure.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/drawFigs.html">drawFigs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p_pca_tsne2</span>, <span class="va">p_pca_umap2</span><span class="op">)</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, common.legend <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">p_all</span></code></pre></div>
<p><img src="single_Embryo_SC_tutorial_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>To save the plot, we can use <code><a href="../reference/write_fig.html">write_fig()</a></code> function.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/write_fig.html">write_fig</a></span><span class="op">(</span><span class="va">p_all</span>, filename <span class="op">=</span> <span class="st">"SCMEB_plots.png"</span><span class="op">)</span></code></pre></div>
<p>Except for the embedding plots, SRTpipeline also provides a variaty of visualization functions. First, we visualize the spatial distribution of cluster labels that shows the layer structure.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## choose colors to function chooseColors</span>
<span class="fu"><a href="../reference/EachClusterSpaHeatMap.html">EachClusterSpaHeatMap</a></span><span class="op">(</span><span class="va">SRTProj</span>, cols <span class="op">=</span> <span class="va">cols_cluster</span>, legend.position <span class="op">=</span> <span class="st">"right"</span>, base_size <span class="op">=</span> <span class="fl">16</span>, pt_size <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></code></pre></div>
<p><img src="single_Embryo_SC_tutorial_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># remove the border EachClusterSpaHeatMap(SRTProj, cols=cols,</span>
<span class="co"># legend.position='right',base_size=16, border_col='white')</span></code></pre></div>
<p>We plotted the heatmap of Pearson’s correlation coefcients of the PCA embeddings among the detected domains shows the good separation of the estimated embeddings across domains and the correlations between deeper layers were high, e.g., there were high correlations between domain 2 and 3, while correlations among the separated layers were low, i.e., domain 1 and 4.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/CCHeatMap.html">CCHeatMap</a></span><span class="op">(</span><span class="va">SRTProj</span>, reduction <span class="op">=</span> <span class="st">"PCA"</span>, grp_color <span class="op">=</span> <span class="va">cols_cluster</span>, ncol.legend <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></code></pre></div>
<p><img src="single_Embryo_SC_tutorial_files/figure-html/unnamed-chunk-16-1.png" width="700"> After adding the quantities for data visualization, the <code>SRTProject</code> object will have more information in the downstream analyses information. Now, we print this <code>SRTProject</code> object to check it. We observed two components added in the slot <code>plotEmbeddings</code> (Embeddings for plotting): <code>tSNE</code> and <code>UMAP</code>.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRTProj</span></code></pre></div>
<pre><code><span class="co">## class: SRTProject </span>
<span class="co">## outputPath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\Embryo3 </span>
<span class="co">## h5filePath: F:\Research paper\IntegrateDRcluster\AnalysisCode\SRTpipeline\vignettes\Embryo3/Embryo3.h5 </span>
<span class="co">## ---------Datasets basic information-----------------</span>
<span class="co">## samples(1): embryo3</span>
<span class="co">## sampleColData names(3): species tissues NumOfSpots</span>
<span class="co">## cellMetaData names(18): orig.ident nCount_RNA ...</span>
<span class="co">##   celltype_mapped_refined batch</span>
<span class="co">## numberOfSpots(1): 23194</span>
<span class="co">## ---------Downstream analyses information-----------------</span>
<span class="co">## Variable features:  351 </span>
<span class="co">## Low-dimensional embeddings(1): PCA</span>
<span class="co">## Inferred cluster labels: Yes</span>
<span class="co">## Embedding for plotting(2): tSNE UMAP</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="deg-analysis">DEG analysis<a class="anchor" aria-label="anchor" href="#deg-analysis"></a>
</h2>
<p>To do downstream analyses, we require to get the count data from h5file. The function <code><a href="../reference/getGeneSpotData.html">getGeneSpotData()</a></code> access the gene-by-spot count matrix and other data in <code>SRTProj</code> and then return a <code>SpatialExperiment</code> object, including two assays: counts, logcounts; rowData: from <code>SRTProj@geneMetaData</code>; colData: from <code>SRTProj@cellMetaData</code>, <code>SRTProj@clusters</code> and sample_id; reducedDim: from <code>SRTProj@reductions</code>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGeneSpotData.html">getGeneSpotData</a></span><span class="op">(</span><span class="va">SRTProj</span><span class="op">)</span>
<span class="va">spe</span></code></pre></div>
<pre><code><span class="co">## class: SpatialExperiment </span>
<span class="co">## dim: 351 23194 </span>
<span class="co">## metadata(0):</span>
<span class="co">## assays(2): counts logcounts</span>
<span class="co">## rownames(351): Abcc4 Acp5 ... Zfp57 Zic3</span>
<span class="co">## rowData names(1): isVGs.combine</span>
<span class="co">## colnames(23194): embryo3#embryo3_Pos0_cell10_z2</span>
<span class="co">##   embryo3#embryo3_Pos0_cell10_z5 ... embryo3#embryo3_Pos39_cell98_z5</span>
<span class="co">##   embryo3#embryo3_Pos39_cell99_z2</span>
<span class="co">## colData names(20): orig.ident nCount_RNA ... clusters sample_id</span>
<span class="co">## reducedDimNames(4): PCA tSNE UMAP Coord</span>
<span class="co">## mainExpName: NULL</span>
<span class="co">## altExpNames(0):</span>
<span class="co">## spatialData names(0) :</span>
<span class="co">## spatialCoords names(2) : V1 V2</span>
<span class="co">## imgData names(0):</span></code></pre>
<p>We perform differential expression analysis for all clusters by using <code><a href="../reference/FindAllDEGs.html">FindAllDEGs()</a></code> function, then the DE genes’ information is saved in a <code>DataFrame</code> object <code>dat_degs</code>.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dat_degs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindAllDEGs.html">FindAllDEGs</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span>
<span class="va">dat_degs</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 4031 rows and 7 columns</span>
<span class="co">##               p_val avg_log2FC     pct.1     pct.2   p_val_adj  cluster</span>
<span class="co">##           &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;factor&gt;</span>
<span class="co">## Sox4    9.98183e-60  -1.136041     0.165     0.672 3.50362e-57        1</span>
<span class="co">## Marcks  9.93938e-55  -1.108176     0.142     0.606 3.48872e-52        1</span>
<span class="co">## Setd1a  1.09492e-46  -1.001061     0.136     0.569 3.84318e-44        1</span>
<span class="co">## Kmt2b   1.66724e-46  -0.971486     0.169     0.632 5.85201e-44        1</span>
<span class="co">## F2r     3.07166e-45  -1.023345     0.078     0.452 1.07815e-42        1</span>
<span class="co">## ...             ...        ...       ...       ...         ...      ...</span>
<span class="co">## Gypa.4  1.42749e-05  -0.256357     0.071     0.131  0.00501048       25</span>
<span class="co">## Isl1.21 1.57008e-05  -0.264269     0.128     0.200  0.00551099       25</span>
<span class="co">## Bmp2.12 1.87106e-05  -0.262016     0.106     0.173  0.00656742       25</span>
<span class="co">## T.12    5.11573e-05  -0.254265     0.113     0.178  0.01795622       25</span>
<span class="co">## Fgf15.5 5.60192e-05  -0.252738     0.072     0.128  0.01966273       25</span>
<span class="co">##                gene</span>
<span class="co">##         &lt;character&gt;</span>
<span class="co">## Sox4           Sox4</span>
<span class="co">## Marcks       Marcks</span>
<span class="co">## Setd1a       Setd1a</span>
<span class="co">## Kmt2b         Kmt2b</span>
<span class="co">## F2r             F2r</span>
<span class="co">## ...             ...</span>
<span class="co">## Gypa.4         Gypa</span>
<span class="co">## Isl1.21        Isl1</span>
<span class="co">## Bmp2.12        Bmp2</span>
<span class="co">## T.12              T</span>
<span class="co">## Fgf15.5       Fgf15</span></code></pre>
<p>We identify the significant DE genes by two criteria: (a) adjustd p-value less than 0.01 and (b) average log fold change greater than 0.4.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">degs_sig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">dat_degs</span>, <span class="va">p_val_adj</span> <span class="op">&lt;</span> <span class="fl">0.01</span> <span class="op">&amp;</span> <span class="va">avg_log2FC</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">degs_sig</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 203 rows and 7 columns</span>
<span class="co">##                  p_val avg_log2FC     pct.1     pct.2    p_val_adj  cluster</span>
<span class="co">##              &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;    &lt;numeric&gt; &lt;factor&gt;</span>
<span class="co">## Foxa1.1              0    1.66291     0.767     0.131            0        2</span>
<span class="co">## Cdh1.1               0    1.62270     0.750     0.121            0        2</span>
<span class="co">## Shh.1                0    1.59177     0.639     0.097            0        2</span>
<span class="co">## Cldn4.1              0    1.57513     0.788     0.157            0        2</span>
<span class="co">## Foxa2.1              0    1.40669     0.624     0.112            0        2</span>
<span class="co">## ...                ...        ...       ...       ...          ...      ...</span>
<span class="co">## Pax8.5     7.00383e-97    1.06645     0.407     0.109  2.45835e-94       24</span>
<span class="co">## Gata3.16   1.86058e-82    1.01690     0.492     0.187  6.53064e-80       24</span>
<span class="co">## Prrx1.15   0.00000e+00    1.39438     0.810     0.235  0.00000e+00       25</span>
<span class="co">## Tfap2a.15 2.08217e-248    1.30171     0.653     0.164 7.30841e-246       25</span>
<span class="co">## Msx1.19   8.48764e-138    1.02382     0.601     0.212 2.97916e-135       25</span>
<span class="co">##                  gene</span>
<span class="co">##           &lt;character&gt;</span>
<span class="co">## Foxa1.1         Foxa1</span>
<span class="co">## Cdh1.1           Cdh1</span>
<span class="co">## Shh.1             Shh</span>
<span class="co">## Cldn4.1         Cldn4</span>
<span class="co">## Foxa2.1         Foxa2</span>
<span class="co">## ...               ...</span>
<span class="co">## Pax8.5           Pax8</span>
<span class="co">## Gata3.16        Gata3</span>
<span class="co">## Prrx1.15        Prrx1</span>
<span class="co">## Tfap2a.15      Tfap2a</span>
<span class="co">## Msx1.19          Msx1</span></code></pre>
<p>In the following, we perform gene set enrichment analysis for the DE genes of each Domain identified by SC-MEB model using R package <code>gprofiler2</code>.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">gprofiler2</span><span class="op">)</span>
<span class="va">termList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">25</span><span class="op">)</span> <span class="op">{</span>
    <span class="co"># k &lt;- 1</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"k = "</span>, <span class="va">k</span>, <span class="st">"\n"</span><span class="op">)</span>
    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">degs_sig</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="va">k</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">{</span>
        <span class="va">dat_degs_sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">degs_sig</span>, <span class="va">cluster</span> <span class="op">==</span> <span class="va">k</span><span class="op">)</span>
        <span class="va">que1</span> <span class="op">&lt;-</span> <span class="va">dat_degs_sub</span><span class="op">$</span><span class="va">gene</span>
        <span class="va">gostres</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gprofiler2/man/gost.html" class="external-link">gost</a></span><span class="op">(</span>query <span class="op">=</span> <span class="va">que1</span>, organism <span class="op">=</span> <span class="st">"hsapiens"</span>, correction_method <span class="op">=</span> <span class="st">"fdr"</span><span class="op">)</span>
        <span class="va">termList</span><span class="op">[[</span><span class="va">k</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">gostres</span>
    <span class="op">}</span>

<span class="op">}</span></code></pre></div>
<pre><code><span class="co">## k =  1 </span>
<span class="co">## k =  2 </span>
<span class="co">## k =  3 </span>
<span class="co">## k =  4 </span>
<span class="co">## k =  5 </span>
<span class="co">## k =  6 </span>
<span class="co">## k =  7 </span>
<span class="co">## k =  8 </span>
<span class="co">## k =  9 </span>
<span class="co">## k =  10 </span>
<span class="co">## k =  11 </span>
<span class="co">## k =  12 </span>
<span class="co">## k =  13 </span>
<span class="co">## k =  14 </span>
<span class="co">## k =  15 </span>
<span class="co">## k =  16 </span>
<span class="co">## k =  17 </span>
<span class="co">## k =  18 </span>
<span class="co">## k =  19 </span>
<span class="co">## k =  20 </span>
<span class="co">## k =  21 </span>
<span class="co">## k =  22 </span>
<span class="co">## k =  23 </span>
<span class="co">## k =  24 </span>
<span class="co">## k =  25</span></code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">termList</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">result</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##     query significant      p_value term_size query_size intersection_size</span>
<span class="co">## 1 query_1        TRUE 4.066137e-05       946         10                 7</span>
<span class="co">## 2 query_1        TRUE 5.426093e-05        21         10                 3</span>
<span class="co">## 3 query_1        TRUE 6.671020e-05         2         10                 2</span>
<span class="co">## 4 query_1        TRUE 6.671020e-05        28         10                 3</span>
<span class="co">## 5 query_1        TRUE 1.760578e-04      1521         10                 7</span>
<span class="co">## 6 query_1        TRUE 1.760578e-04        43         10                 3</span>
<span class="co">##   precision      recall    term_id source</span>
<span class="co">## 1       0.7 0.007399577 GO:0098609  GO:BP</span>
<span class="co">## 2       0.3 0.142857143 GO:2000047  GO:BP</span>
<span class="co">## 3       0.2 1.000000000 GO:0060738  GO:BP</span>
<span class="co">## 4       0.3 0.107142857 GO:0044331  GO:BP</span>
<span class="co">## 5       0.7 0.004602235 GO:0007155  GO:BP</span>
<span class="co">## 6       0.3 0.069767442 GO:0071542  GO:BP</span>
<span class="co">##                                                                 term_name</span>
<span class="co">## 1                                                      cell-cell adhesion</span>
<span class="co">## 2                   regulation of cell-cell adhesion mediated by cadherin</span>
<span class="co">## 3 epithelial-mesenchymal signaling involved in prostate gland development</span>
<span class="co">## 4                                 cell-cell adhesion mediated by cadherin</span>
<span class="co">## 5                                                           cell adhesion</span>
<span class="co">## 6                                     dopaminergic neuron differentiation</span>
<span class="co">##   effective_domain_size source_order                            parents</span>
<span class="co">## 1                 21092        21975                         GO:0007155</span>
<span class="co">## 2                 21092        29255             GO:0022407, GO:0044331</span>
<span class="co">## 3                 21092        17471 GO:0022414, GO:0030850, GO:0060684</span>
<span class="co">## 4                 21092        12705                         GO:0098609</span>
<span class="co">## 5                 21092         3051                         GO:0009987</span>
<span class="co">## 6                 21092        19530                         GO:0030182</span></code></pre>
<p>To understand the functions of the identified spatial domains by DR-SC model, we compare the top significant biological process (BP) pathways in GO database for the DE genes from Domain 1 and 2. Here, we only show to visualize the significant BP pathways and users can explore the other databases such as KEGG and HPA.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Most commonly used databases</span>
<span class="va">source_set</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GO:BP"</span>, <span class="st">"GO:CC"</span>, <span class="st">"GO:MF"</span>, <span class="st">"KEGG"</span>, <span class="st">"HPA"</span><span class="op">)</span>
<span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"steelblue3"</span>, <span class="st">"goldenrod"</span>, <span class="st">"brown3"</span>, <span class="st">"#f98866"</span>, <span class="st">"#CE6DBD"</span><span class="op">)</span>
<span class="co">## Here, we show GO:BP</span>
<span class="va">source1</span> <span class="op">&lt;-</span> <span class="st">"GO:BP"</span>
<span class="va">ss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">source_set</span> <span class="op">==</span> <span class="va">source1</span><span class="op">)</span>
<span class="va">ntop</span> <span class="op">=</span> <span class="fl">5</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cols</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">source_set</span>
<span class="va">pList_enrich</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">ii</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">25</span><span class="op">)</span> <span class="op">{</span>
    <span class="co">## ii &lt;- 4</span>
    <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"ii="</span>, <span class="va">ii</span><span class="op">)</span>
    <span class="va">dat</span> <span class="op">&lt;-</span> <span class="va">termList</span><span class="op">[[</span><span class="va">ii</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">result</span>
    <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
        <span class="va">dat1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">dat</span>, <span class="va">term_size</span> <span class="op">&lt;</span> <span class="fl">500</span><span class="op">)</span>
        <span class="va">dat1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_top_pathway.html">get_top_pathway</a></span><span class="op">(</span><span class="va">dat1</span>, ntop <span class="op">=</span> <span class="va">ntop</span>, source_set <span class="op">=</span> <span class="va">source1</span><span class="op">)</span>
        <span class="va">dat1</span> <span class="op">&lt;-</span> <span class="va">dat1</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/stats/complete.cases.html" class="external-link">complete.cases</a></span><span class="op">(</span><span class="va">dat1</span><span class="op">)</span>, <span class="op">]</span>
        <span class="va">dat1</span><span class="op">$</span><span class="va">nlog10P</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">dat1</span><span class="op">$</span><span class="va">p_value</span><span class="op">)</span>

        <span class="va">pList_enrich</span><span class="op">[[</span><span class="va">ii</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/barPlot_enrich.html">barPlot_enrich</a></span><span class="op">(</span><span class="va">dat1</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">dat1</span><span class="op">$</span><span class="va">nlog10P</span><span class="op">)</span>, <span class="op">]</span>, source <span class="op">=</span> <span class="st">"source"</span>, <span class="st">"term_name"</span>,
            <span class="st">"nlog10P"</span>, cols <span class="op">=</span> <span class="va">cols</span><span class="op">[</span><span class="va">source_set</span><span class="op">[</span><span class="va">ss</span><span class="op">]</span><span class="op">]</span>, base_size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ylab</span><span class="op">(</span><span class="st">"-log10(p-adj)"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">xlab</span><span class="op">(</span><span class="st">"Biological terms"</span><span class="op">)</span> <span class="op">+</span>
            <span class="fu">ggtitle</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Domain"</span>, <span class="va">ii</span><span class="op">)</span><span class="op">)</span>
    <span class="op">}</span>

<span class="op">}</span>

<span class="fu"><a href="../reference/drawFigs.html">drawFigs</a></span><span class="op">(</span><span class="va">pList_enrich</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span>, common.legend <span class="op">=</span> <span class="cn">T</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span>, align <span class="op">=</span> <span class="st">"hv"</span><span class="op">)</span></code></pre></div>
<p><img src="single_Embryo_SC_tutorial_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<p>We take out the top DE genes for each cluster for visualization.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>
<span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">3</span>
<span class="va">dat_degs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="va">as.data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html" class="external-link">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n</span>, wt <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">topGene</span>
<span class="va">topGene</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 75 x 7</span></span>
<span class="co">## <span style="color: #949494;"># Groups:   cluster [25]</span></span>
<span class="co">##          p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene </span>
<span class="co">##          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> 0.000<span style="text-decoration: underline;">000</span>117     -<span style="color: #BB0000;">0.285</span> 0.027 0.103 0.000<span style="text-decoration: underline;">041</span>1 1       Wisp1</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> 0.000<span style="text-decoration: underline;">000</span>160     -<span style="color: #BB0000;">0.282</span> 0.031 0.108 0.000<span style="text-decoration: underline;">056</span>2 1       Tdo2 </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> 0.000<span style="text-decoration: underline;">000</span>223     -<span style="color: #BB0000;">0.283</span> 0.027 0.1   0.000<span style="text-decoration: underline;">078</span>1 1       Mixl1</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> 0                1.66  0.767 0.131 0         2       Foxa1</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> 0                1.62  0.75  0.121 0         2       Cdh1 </span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> 0                1.59  0.639 0.097 0         2       Shh  </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> 0                1.43  0.802 0.208 0         3       Cdh5 </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> 0                1.35  0.835 0.263 0         3       Plvap</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> 0                1.31  0.68  0.172 0         3       Cd34 </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> 0                1.42  0.908 0.271 0         4       Foxf1</span>
<span class="co">## <span style="color: #949494;"># ... with 65 more rows</span></span></code></pre>
<p>We visualize the DE genes for each cluster group by gene-by-cell heatmap using the <code><a href="../reference/GCHeatMap.html">GCHeatMap()</a></code> function.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EachGCHeatMap.html">EachGCHeatMap</a></span><span class="op">(</span><span class="va">spe</span>, features <span class="op">=</span> <span class="va">topGene</span><span class="op">$</span><span class="va">gene</span>, grp_color <span class="op">=</span> <span class="va">cols_cluster</span>, y_text_size <span class="op">=</span> <span class="fl">6</span>, ncol.legend <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>
<span class="va">p1</span></code></pre></div>
<p><img src="single_Embryo_SC_tutorial_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="trajectory-inference">Trajectory inference<a class="anchor" aria-label="anchor" href="#trajectory-inference"></a>
</h2>
<p>Next, we performed trajectory inference using the PCA embeddings and domain labels estimated by SC-MEB model. The <code><a href="../reference/EmbedPlot.html">EmbedPlot()</a></code> function can be used to visualize the inferred pseudotime on a specified embedding. If there is only one data batch, the function <code><a href="../reference/EachEmbedPlot.html">EachEmbedPlot()</a></code> plays the same role as function <code><a href="../reference/EmbedPlot.html">EmbedPlot()</a></code>.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddTrajectory.html">AddTrajectory</a></span><span class="op">(</span><span class="va">spe</span>, reduction <span class="op">=</span> <span class="st">"PCA"</span>, name <span class="op">=</span> <span class="st">"PT"</span><span class="op">)</span>
<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EmbedPlot.html">EmbedPlot</a></span><span class="op">(</span><span class="va">spe</span>, reduction <span class="op">=</span> <span class="st">"PCA"</span>, colour_by <span class="op">=</span> <span class="st">"PT"</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EmbedPlot.html">EmbedPlot</a></span><span class="op">(</span><span class="va">spe</span>, reduction <span class="op">=</span> <span class="st">"Coord"</span>, colour_by <span class="op">=</span> <span class="st">"PT"</span><span class="op">)</span>
<span class="fu"><a href="../reference/drawFigs.html">drawFigs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span><span class="op">)</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, common.legend <span class="op">=</span> <span class="cn">TRUE</span>, legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></code></pre></div>
<p><img src="single_Embryo_SC_tutorial_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># EachEmbedPlot(spe, reduction = 'PCA', colour_by='TrajPT') save(spe, SRTProj,</span>
<span class="co"># file=paste0(SRTProj@projectMetadata$outputPath,'/SRTProj.rds')) load('F:/Research</span>
<span class="co"># paper/IntegrateDRcluster/AnalysisCode/SRTpipeline/vignettes/Embryo3/SRTProj.rds')</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="other-downstream-analyses">Other downstream analyses<a class="anchor" aria-label="anchor" href="#other-downstream-analyses"></a>
</h2>
<details><p><summary><strong>Session Info</strong></summary></p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.1.2 (2021-11-01)</span>
<span class="co">## Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">## Running under: Windows 10 x64 (build 22621)</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">## [1] LC_COLLATE=Chinese (Simplified)_China.936 </span>
<span class="co">## [2] LC_CTYPE=Chinese (Simplified)_China.936   </span>
<span class="co">## [3] LC_MONETARY=Chinese (Simplified)_China.936</span>
<span class="co">## [4] LC_NUMERIC=C                              </span>
<span class="co">## [5] LC_TIME=Chinese (Simplified)_China.936    </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">## [8] methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] scater_1.25.1               scuttle_1.4.0              </span>
<span class="co">##  [3] slingshot_2.2.0             TrajectoryUtils_1.2.0      </span>
<span class="co">##  [5] princurve_2.1.6             gprofiler2_0.2.1           </span>
<span class="co">##  [7] SpatialExperiment_1.4.0     SingleCellExperiment_1.16.0</span>
<span class="co">##  [9] SummarizedExperiment_1.24.0 Biobase_2.54.0             </span>
<span class="co">## [11] GenomicRanges_1.46.1        GenomeInfoDb_1.30.1        </span>
<span class="co">## [13] IRanges_2.28.0              MatrixGenerics_1.6.0       </span>
<span class="co">## [15] matrixStats_0.62.0          colorspace_2.0-3           </span>
<span class="co">## [17] SC.MEB_1.1                  mclust_5.4.10              </span>
<span class="co">## [19] irlba_2.3.5                 DR.SC_3.1                  </span>
<span class="co">## [21] spatstat.geom_2.4-0         spatstat.data_3.0-0        </span>
<span class="co">## [23] ggplot2_3.3.6               dplyr_1.0.9                </span>
<span class="co">## [25] Matrix_1.4-0                sp_1.5-0                   </span>
<span class="co">## [27] SeuratObject_4.1.0          Seurat_4.1.1               </span>
<span class="co">## [29] hdf5r_1.3.5                 ff_4.0.7                   </span>
<span class="co">## [31] bit_4.0.4                   S4Vectors_0.32.3           </span>
<span class="co">## [33] BiocGenerics_0.40.0         rhdf5_2.38.0               </span>
<span class="co">## [35] SRTpipeline_0.1.1          </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] scattermore_0.8           ggthemes_4.2.4           </span>
<span class="co">##   [3] R.methodsS3_1.8.1         GiRaF_1.0.1              </span>
<span class="co">##   [5] ragg_1.2.2                tidyr_1.2.0              </span>
<span class="co">##   [7] bit64_4.0.5               knitr_1.37               </span>
<span class="co">##   [9] DelayedArray_0.20.0       R.utils_2.11.0           </span>
<span class="co">##  [11] data.table_1.14.2         rpart_4.1.16             </span>
<span class="co">##  [13] RCurl_1.98-1.6            generics_0.1.2           </span>
<span class="co">##  [15] ScaledMatrix_1.2.0        cowplot_1.1.1            </span>
<span class="co">##  [17] RANN_2.6.1                future_1.26.1            </span>
<span class="co">##  [19] httpuv_1.6.5              assertthat_0.2.1         </span>
<span class="co">##  [21] viridis_0.6.2             xfun_0.29                </span>
<span class="co">##  [23] jquerylib_0.1.4           evaluate_0.15            </span>
<span class="co">##  [25] promises_1.2.0.1          fansi_1.0.3              </span>
<span class="co">##  [27] igraph_1.3.5              DBI_1.1.2                </span>
<span class="co">##  [29] htmlwidgets_1.5.4         purrr_0.3.4              </span>
<span class="co">##  [31] ellipsis_0.3.2            RSpectra_0.16-1          </span>
<span class="co">##  [33] ggpubr_0.4.0              backports_1.4.1          </span>
<span class="co">##  [35] deldir_1.0-6              sparseMatrixStats_1.6.0  </span>
<span class="co">##  [37] vctrs_0.4.1               ROCR_1.0-11              </span>
<span class="co">##  [39] abind_1.4-5               cachem_1.0.6             </span>
<span class="co">##  [41] withr_2.5.0               PRECAST_1.4              </span>
<span class="co">##  [43] progressr_0.10.1          sctransform_0.3.3        </span>
<span class="co">##  [45] goftest_1.2-3             cluster_2.1.2            </span>
<span class="co">##  [47] lazyeval_0.2.2            crayon_1.5.1             </span>
<span class="co">##  [49] edgeR_3.36.0              pkgconfig_2.0.3          </span>
<span class="co">##  [51] labeling_0.4.2            nlme_3.1-155             </span>
<span class="co">##  [53] vipor_0.4.5               rlang_1.0.2              </span>
<span class="co">##  [55] globals_0.15.0            lifecycle_1.0.1          </span>
<span class="co">##  [57] miniUI_0.1.1.1            rsvd_1.0.5               </span>
<span class="co">##  [59] rprojroot_2.0.3           polyclip_1.10-0          </span>
<span class="co">##  [61] lmtest_0.9-40             carData_3.0-5            </span>
<span class="co">##  [63] Rhdf5lib_1.16.0           zoo_1.8-10               </span>
<span class="co">##  [65] beeswarm_0.4.0            ggridges_0.5.3           </span>
<span class="co">##  [67] rjson_0.2.21              png_0.1-7                </span>
<span class="co">##  [69] viridisLite_0.4.0         bitops_1.0-7             </span>
<span class="co">##  [71] R.oo_1.24.0               KernSmooth_2.23-20       </span>
<span class="co">##  [73] rhdf5filters_1.6.0        DelayedMatrixStats_1.16.0</span>
<span class="co">##  [75] stringr_1.4.0             parallelly_1.32.0        </span>
<span class="co">##  [77] spatstat.random_2.2-0     rstatix_0.7.0            </span>
<span class="co">##  [79] ggsignif_0.6.3            beachmat_2.10.0          </span>
<span class="co">##  [81] scales_1.2.0              memoise_2.0.1            </span>
<span class="co">##  [83] magrittr_2.0.3            plyr_1.8.7               </span>
<span class="co">##  [85] ica_1.0-2                 zlibbioc_1.40.0          </span>
<span class="co">##  [87] compiler_4.1.2            dqrng_0.3.0              </span>
<span class="co">##  [89] RColorBrewer_1.1-3        fitdistrplus_1.1-8       </span>
<span class="co">##  [91] cli_3.2.0                 XVector_0.34.0           </span>
<span class="co">##  [93] listenv_0.8.0             patchwork_1.1.1          </span>
<span class="co">##  [95] pbapply_1.5-0             formatR_1.11             </span>
<span class="co">##  [97] MASS_7.3-55               mgcv_1.8-39              </span>
<span class="co">##  [99] tidyselect_1.1.2          stringi_1.7.6            </span>
<span class="co">## [101] textshaping_0.3.6         highr_0.9                </span>
<span class="co">## [103] yaml_2.3.6                BiocSingular_1.10.0      </span>
<span class="co">## [105] locfit_1.5-9.4            ggrepel_0.9.1            </span>
<span class="co">## [107] grid_4.1.2                sass_0.4.1               </span>
<span class="co">## [109] tools_4.1.2               future.apply_1.9.0       </span>
<span class="co">## [111] rstudioapi_0.13           gridExtra_2.3            </span>
<span class="co">## [113] farver_2.1.0              Rtsne_0.16               </span>
<span class="co">## [115] DropletUtils_1.14.2       digest_0.6.29            </span>
<span class="co">## [117] rgeos_0.5-9               shiny_1.7.1              </span>
<span class="co">## [119] Rcpp_1.0.10               car_3.0-12               </span>
<span class="co">## [121] broom_0.7.12              later_1.3.0              </span>
<span class="co">## [123] RcppAnnoy_0.0.19          httr_1.4.3               </span>
<span class="co">## [125] fs_1.5.2                  tensor_1.5               </span>
<span class="co">## [127] reticulate_1.25           splines_4.1.2            </span>
<span class="co">## [129] uwot_0.1.11               spatstat.utils_3.0-1     </span>
<span class="co">## [131] pkgdown_2.0.6             plotly_4.10.0            </span>
<span class="co">## [133] systemfonts_1.0.4         xtable_1.8-4             </span>
<span class="co">## [135] jsonlite_1.8.0            R6_2.5.1                 </span>
<span class="co">## [137] pillar_1.7.0              htmltools_0.5.2          </span>
<span class="co">## [139] mime_0.12                 glue_1.6.2               </span>
<span class="co">## [141] fastmap_1.1.0             BiocParallel_1.28.3      </span>
<span class="co">## [143] BiocNeighbors_1.12.0      codetools_0.2-18         </span>
<span class="co">## [145] utf8_1.2.2                lattice_0.20-45          </span>
<span class="co">## [147] bslib_0.3.1               spatstat.sparse_2.1-1    </span>
<span class="co">## [149] tibble_3.1.7              ggbeeswarm_0.6.0         </span>
<span class="co">## [151] leiden_0.4.2              gtools_3.9.2.2           </span>
<span class="co">## [153] magick_2.7.3              survival_3.2-13          </span>
<span class="co">## [155] limma_3.50.1              CompQuadForm_1.4.3       </span>
<span class="co">## [157] rmarkdown_2.11            desc_1.4.0               </span>
<span class="co">## [159] munsell_0.5.0             GenomeInfoDbData_1.2.7   </span>
<span class="co">## [161] HDF5Array_1.22.1          reshape2_1.4.4           </span>
<span class="co">## [163] gtable_0.3.0              spatstat.core_2.4-4</span></code></pre>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Wei Liu, Yi Yang, Xu Liao, Xiao Zhang, Jin Liu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
