<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ SRTpipeline</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SRTpipeline</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="../articles/Installation.html">
    <span class="fas fa-book"></span>
     
    Install
  </a>
</li>
<li>
  <a href="../articles/get_started.html">
    <span class="fas fa-book"></span>
     
    Get Started
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fas fa-book"></span>
     
    Single Data Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/single_basic_tutorial.html">Basic overview</a>
    </li>
    <li>
      <a href="../articles/dlpfc_tutorial.html">Analyze a human DLPFC section</a>
    </li>
    <li>
      <a href="../articles/single_Embryo_SC_tutorial.html">Analyze a mouse embryo section</a>
    </li>
    <li>
      <a href="../articles/single_Embryo_DRSC_tutorial.html">Analyze a mouse embryo section v2</a>
    </li>
    <li>
      <a href="../articles/single_MOB_Anno_tutorial.html">Analyze a mouse olfactory bulb section</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fas fa-book"></span>
     
    Multiple Data Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/multibatch_basic_tutorial.html">Basic overview</a>
    </li>
    <li>
      <a href="../articles/integration_introduction.html">Integration introduction</a>
    </li>
    <li>
      <a href="../articles/multibatch_brain12_tutorial.html">Integrate three human DLPFC slices</a>
    </li>
    <li>
      <a href="../articles/multibatch_HCC4_tutorial.html">Integrate four human HCC slices</a>
    </li>
    <li>
      <a href="../articles/multibatch_liver8_tutorial.html">Integrate eight mouse liver slices</a>
    </li>
    <li>
      <a href="../articles/multibatch_BS2_tutorial.html">Integrate two human breast cancer slices</a>
    </li>
    <li>
      <a href="../articles/integration_large_datasets.html">Tips for integrating large datasets</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/faq.html">
    <span class="fa fa-question-circle-o"></span>
     
    FAQ
  </a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Functions
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="simu_seuList.html">simu_seuList()</a></code> </p>
        </td>
        <td><p>SpatialExperiment object to Seurat object</p></td>
      </tr><tr><td>
          <p><code><a href="AddTSNE.html">AddTSNE()</a></code> </p>
        </td>
        <td><p>Add tSNE embeddings for a SRTProject object</p></td>
      </tr><tr><td>
          <p><code><a href="chooseColors.html">chooseColors()</a></code> </p>
        </td>
        <td><p>Choose color schema from a palette</p></td>
      </tr><tr><td>
          <p><code><a href="AddUMAP.html">AddUMAP()</a></code> </p>
        </td>
        <td><p>Add UMAP embeddings for a SRTProject object</p></td>
      </tr><tr><td>
          <p><code><a href="transferGeneNames.html">transferGeneNames()</a></code> </p>
        </td>
        <td><p>Transfer gene names from one fortmat to the other format</p></td>
      </tr><tr><td>
          <p><code><a href="filter_gene.html">filter_gene()</a></code> </p>
        </td>
        <td><p>Filter genes for a Seurat object</p></td>
      </tr><tr><td>
          <p><code><a href="filter_spot.html">filter_spot()</a></code> </p>
        </td>
        <td><p>Filter spots for a Seurat object</p></td>
      </tr><tr><td>
          <p><code><a href="get_top_pathway.html">get_top_pathway()</a></code> </p>
        </td>
        <td><p>Get top pathways from enrichment analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="barPlot_enrich.html">barPlot_enrich()</a></code> </p>
        </td>
        <td><p>Barplot for the enrichment terms</p></td>
      </tr><tr><td>
          <p><code><a href="RCTD_structure.html">RCTD_structure()</a></code> </p>
        </td>
        <td><p>Prepare the reference panel data for the RCTD deconvolution analysis</p></td>
      </tr><tr><td>
          <p><code><a href="RCTD_run.html">RCTD_run()</a></code> </p>
        </td>
        <td><p>Perform spatial deconvolution analysis using RCTD model</p></td>
      </tr><tr><td>
          <p><code><a href="model_set_SCMEB.html">model_set_SCMEB()</a></code> </p>
        </td>
        <td><p>Set parameters for SC-MEB model</p></td>
      </tr><tr><td>
          <p><code><a href="model_set_DRSC.html">model_set_DRSC()</a></code> </p>
        </td>
        <td><p>Set parameters for DR-SC model</p></td>
      </tr><tr><td>
          <p><code><a href="model_set_PRECAST.html">model_set_PRECAST()</a></code> </p>
        </td>
        <td><p>Set parameters for PRECAST model</p></td>
      </tr><tr><td>
          <p><code><a href="getIntegratedData.html">getIntegratedData()</a></code> </p>
        </td>
        <td><p>Get the batch-corrected integrated gene expresssions</p></td>
      </tr><tr><td>
          <p><code><a href="SRTProject-class.html">SRTProject-class</a></code> </p>
        </td>
        <td><p>SRTProject Class to store SRT datasets</p></td>
      </tr><tr><td>
          <p><code><a href="addProjectSummary.html">addProjectSummary()</a></code> </p>
        </td>
        <td><p>Add the summary data for a SRTProject object</p></td>
      </tr><tr><td>
          <p><code><a href="setSRTverbose.html">setSRTverbose()</a></code> </p>
        </td>
        <td><p>Determine whether output the information for all functions in SRTpipeline package.</p></td>
      </tr><tr><td>
          <p><code><a href="getSRTVerbose.html">getSRTVerbose()</a></code> </p>
        </td>
        <td><p>Get the output status</p></td>
      </tr><tr><td>
          <p><code><a href="setSRToutputPrefix.html">setSRToutputPrefix()</a></code> </p>
        </td>
        <td><p>Set the Prefix when output informtion</p></td>
      </tr><tr><td>
          <p><code><a href="getSRToutputPrefix.html">getSRToutputPrefix()</a></code> </p>
        </td>
        <td><p>Get the Prefix of output information</p></td>
      </tr><tr><td>
          <p><code><a href="CreateSRTProject.html">CreateSRTProject()</a></code> </p>
        </td>
        <td><p>Create a SRTProject object</p></td>
      </tr><tr><td>
          <p><code><a href="normalizeSRT.html">normalizeSRT()</a></code> </p>
        </td>
        <td><p>Normalize the expression count data</p></td>
      </tr><tr><td>
          <p><code><a href="selectVariableFeatures.html">selectVariableFeatures()</a></code> </p>
        </td>
        <td><p>Select variable genes</p></td>
      </tr><tr><td>
          <p><code><a href="AddAdj.html">AddAdj()</a></code> </p>
        </td>
        <td><p>Add adjacency matrix list for a SRTProject object</p></td>
      </tr><tr><td>
          <p><code><a href="CCHeatMap.html">CCHeatMap()</a></code> </p>
        </td>
        <td><p>Plot cluster-by-cluster correlation hetamap</p></td>
      </tr><tr><td>
          <p><code><a href="EmbedPlot.html">EmbedPlot()</a></code> </p>
        </td>
        <td><p>Two-dimensional embeddings plot</p></td>
      </tr><tr><td>
          <p><code><a href="model_set_iSCMEB.html">model_set_iSCMEB()</a></code> </p>
        </td>
        <td><p>Set parameters for iSC.MEB model</p></td>
      </tr><tr><td>
          <p><code><a href="FindDEGs.html">FindDEGs()</a></code> </p>
        </td>
        <td><p>Find differential expression genes
Find differential expression genes by comparing two clusters.</p></td>
      </tr><tr><td>
          <p><code><a href="FindAllDEGs.html">FindAllDEGs()</a></code> </p>
        </td>
        <td><p>Find differential expression genes for all clusters by comparing one cluster to the remaining clusters.</p></td>
      </tr><tr><td>
          <p><code><a href="AddTrajectory.html">AddTrajectory()</a></code> </p>
        </td>
        <td><p>Implement trajectory inference</p></td>
      </tr><tr><td>
          <p><code><a href="AddPCA.html">AddPCA()</a></code> </p>
        </td>
        <td><p>Add PCA embeddings for a SRTProject object</p></td>
      </tr><tr><td>
          <p><code><a href="cluster_metric.html">cluster_metric()</a></code> </p>
        </td>
        <td><p>Compute adjusted Rand index and normalized mutual information</p></td>
      </tr><tr><td>
          <p><code><a href="Integrate_PRECAST.html">Integrate_PRECAST()</a></code> </p>
        </td>
        <td><p>Joint embedding, clustering and alignment for SRT data by fitting a PRECAST model</p></td>
      </tr><tr><td>
          <p><code><a href="Cluster_DRSC.html">Cluster_DRSC()</a></code> </p>
        </td>
        <td><p>Joint embedding and clustering for SRT data by fitting an iSC.MEB model a DR-SC model</p></td>
      </tr><tr><td>
          <p><code><a href="Cluster_SCMEB.html">Cluster_SCMEB()</a></code> </p>
        </td>
        <td><p>Spatial clustering for SRT data by fitting a SC-MEB model</p></td>
      </tr><tr><td>
          <p><code><a href="Cluster_Louvain.html">Cluster_Louvain()</a></code> </p>
        </td>
        <td><p>Louvain clustering for a SRT data</p></td>
      </tr><tr><td>
          <p><code><a href="Annotation_SpatialAnno.html">Annotation_SpatialAnno()</a></code> </p>
        </td>
        <td><p>Make annotations for SRT data by fitting a SpatialAnno model</p></td>
      </tr><tr><td>
          <p><code><a href="TrajectEmbedPlot.html">TrajectEmbedPlot()</a></code> </p>
        </td>
        <td><p>Visualize the inferred pseudotime on  embeddings.</p></td>
      </tr><tr><td>
          <p><code><a href="write_fig.html">write_fig()</a></code> </p>
        </td>
        <td><p>Save a ggplot object to a PNG figure</p></td>
      </tr><tr><td>
          <p><code><a href="Integrate_iSCMEB.html">Integrate_iSCMEB()</a></code> </p>
        </td>
        <td><p>Joint clustering and alignment for SRT data by fitting an iSC.MEB model</p></td>
      </tr><tr><td>
          <p><code><a href="vec2list.html">vec2list()</a></code> </p>
        </td>
        <td><p>Transfer a vector to a list</p></td>
      </tr><tr><td>
          <p><code><a href="mat2list.html">mat2list()</a></code> </p>
        </td>
        <td><p>Transfer a matrix to a list</p></td>
      </tr><tr><td>
          <p><code><a href="matlist2mat.html">matlist2mat()</a></code> </p>
        </td>
        <td><p>Transfer matrix list to a matrix</p></td>
      </tr><tr><td>
          <p><code><a href="getGeneSpotData.html">getGeneSpotData()</a></code> </p>
        </td>
        <td><p>Get the expression matrix from a SRTProject object</p></td>
      </tr><tr><td>
          <p><code><a href="EachClusterSpaHeatMap.html">EachClusterSpaHeatMap()</a></code> </p>
        </td>
        <td><p>Plot spatial cluster heatmap for each data batch</p></td>
      </tr><tr><td>
          <p><code><a href="EachRGBSpaHeatMap.html">EachRGBSpaHeatMap()</a></code> </p>
        </td>
        <td><p>Plot spatial RGB heatmaps for a group of SRT data batches</p></td>
      </tr><tr><td>
          <p><code><a href="EachExprSpaHeatMap.html">EachExprSpaHeatMap()</a></code> </p>
        </td>
        <td><p>Plot spatial feature heatmaps for a group of SRT data batches</p></td>
      </tr><tr><td>
          <p><code><a href="EachGCHeatMap.html">EachGCHeatMap()</a></code> </p>
        </td>
        <td><p>Plot gene-by-cell heatmaps for a group of SRT data batches</p></td>
      </tr><tr><td>
          <p><code><a href="model_set_Louvain.html">model_set_Louvain()</a></code> </p>
        </td>
        <td><p>Set parameters for Louvain model</p></td>
      </tr><tr><td>
          <p><code><a href="model_set_SpatialAnno.html">model_set_SpatialAnno()</a></code> </p>
        </td>
        <td><p>Set parameters for SpatialAnno model</p></td>
      </tr><tr><td>
          <p><code><a href="topVGs.html">topVGs()</a></code> </p>
        </td>
        <td><p>Get the top variables genes</p></td>
      </tr><tr><td>
          <p><code><a href="AddClusters2MetaData.html">AddClusters2MetaData()</a></code> </p>
        </td>
        <td><p>Add inferred clusters to the <code>cellMetaData</code></p></td>
      </tr><tr><td>
          <p><code><a href="GCHeatMap.html">GCHeatMap()</a></code> </p>
        </td>
        <td><p>Plot gene-by-cell heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="EachEmbedPlot.html">EachEmbedPlot()</a></code> </p>
        </td>
        <td><p>Plot the embeddings for each SRT data</p></td>
      </tr><tr><td>
          <p><code><a href="drawFigs.html">drawFigs()</a></code> </p>
        </td>
        <td><p>Draw a figure using a group of ggplot objects</p></td>
      </tr><tr><td>
          <p><code><a href="spe2seurat.html">spe2seurat()</a></code> </p>
        </td>
        <td><p>SpatialExperiment object to Seurat object</p></td>
      </tr><tr><td>
          <p><code><a href="spe2seuratList.html">spe2seuratList()</a></code> </p>
        </td>
        <td><p>SpatialExperiment object to Seurat list object</p></td>
      </tr><tr><td>
          <p><code><a href="seu2seuList.html">seu2seuList()</a></code> </p>
        </td>
        <td><p>Seurat object to seuratList object</p></td>
      </tr><tr><td>
          <p><code><a href="sce2seurat.html">sce2seurat()</a></code> </p>
        </td>
        <td><p>SingleCellExperiment object to Seurat object</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Wei Liu, Yi Yang, Xu Liao, Xiao Zhang, Jin Liu.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

